// node-opcua-transport/_generated_/_auto_generated_HelloMessage.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:06.421Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/HelloMessage_schema").HelloMessage_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class HelloMessage
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function HelloMessage(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * The latest version of the OPC UA TCP protocol supported by the Client
      * @property protocolVersion
      * @type {UInt32}
      */
    self.protocolVersion = initialize_field(schema.fields[0], options.protocolVersion);

    /**
      * The largest message that the sender can receive.
      * @property receiveBufferSize
      * @type {UInt32}
      */
    self.receiveBufferSize = initialize_field(schema.fields[1], options.receiveBufferSize);

    /**
      * The largest message that the sender will send.
      * @property sendBufferSize
      * @type {UInt32}
      */
    self.sendBufferSize = initialize_field(schema.fields[2], options.sendBufferSize);

    /**
      * The maximum size for any response message.
      * @property maxMessageSize
      * @type {UInt32}
      */
    self.maxMessageSize = initialize_field(schema.fields[3], options.maxMessageSize);

    /**
      * The maximum number of chunks in any response message
      * @property maxChunkCount
      * @type {UInt32}
      */
    self.maxChunkCount = initialize_field(schema.fields[4], options.maxChunkCount);

    /**
      * The URL of the Endpoint which the Client wished to connect to.
      * @property endpointUrl
      * @type {UAString}
      */
    self.endpointUrl = initialize_field(schema.fields[5], options.endpointUrl);

   // Object.preventExtensions(self);
}
util.inherits(HelloMessage,BaseUAObject);
schema.id = generate_new_id();
HelloMessage.prototype.encodingDefaultBinary = makeExpandedNodeId(schema.id);
HelloMessage.prototype._schema = schema;

const encode_UInt32 = _defaultTypeMap.UInt32.encode;
const decode_UInt32 = _defaultTypeMap.UInt32.decode;
const encode_UAString = _defaultTypeMap.UAString.encode;
const decode_UAString = _defaultTypeMap.UAString.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
HelloMessage.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_UInt32(this.protocolVersion,stream);
    encode_UInt32(this.receiveBufferSize,stream);
    encode_UInt32(this.sendBufferSize,stream);
    encode_UInt32(this.maxMessageSize,stream);
    encode_UInt32(this.maxChunkCount,stream);
    encode_UAString(this.endpointUrl,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
HelloMessage.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.protocolVersion = decode_UInt32(stream);
    this.receiveBufferSize = decode_UInt32(stream);
    this.sendBufferSize = decode_UInt32(stream);
    this.maxMessageSize = decode_UInt32(stream);
    this.maxChunkCount = decode_UInt32(stream);
    this.endpointUrl = decode_UAString(stream);
};
HelloMessage.possibleFields = [
  "protocolVersion",
         "receiveBufferSize",
         "sendBufferSize",
         "maxMessageSize",
         "maxChunkCount",
         "endpointUrl"
];


exports.HelloMessage = HelloMessage;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("HelloMessage",HelloMessage);
