// node-opcua-service-subscription/_generated_/_auto_generated_TransferSubscriptionsRequest.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:34.700Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/TransferSubscriptionsRequest_schema").TransferSubscriptionsRequest_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var RequestHeader = getFactory("RequestHeader");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class TransferSubscriptionsRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} 
 * @param  [options.subscriptionIds] {IntegerId[]} 
 * @param  [options.sendInitialValues] {Boolean} 
 */
function TransferSubscriptionsRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestHeader =  null; /* new RequestHeader(null); */
        return ;
    }

    /**
      * 
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * 
      * @property subscriptionIds
      * @type {IntegerId[]}
      */
    self.subscriptionIds = initialize_field_array(schema.fields[1], options.subscriptionIds);

    /**
      * 
      * @property sendInitialValues
      * @type {Boolean}
      */
    self.sendInitialValues = initialize_field(schema.fields[2], options.sendInitialValues);

   // Object.preventExtensions(self);
}
util.inherits(TransferSubscriptionsRequest,BaseUAObject);
TransferSubscriptionsRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(841,0);
TransferSubscriptionsRequest.prototype.encodingDefaultXml = makeExpandedNodeId(840,0);
TransferSubscriptionsRequest.prototype._schema = schema;

var encode_IntegerId = _defaultTypeMap.IntegerId.encode;
var decode_IntegerId = _defaultTypeMap.IntegerId.decode;
var encode_Boolean = _defaultTypeMap.Boolean.encode;
var decode_Boolean = _defaultTypeMap.Boolean.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
TransferSubscriptionsRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encodeArray(this.subscriptionIds, stream, encode_IntegerId);
    encode_Boolean(this.sendInitialValues,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
TransferSubscriptionsRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.subscriptionIds = decodeArray(stream, decode_IntegerId);
    this.sendInitialValues = decode_Boolean(stream,options);
};
TransferSubscriptionsRequest.possibleFields = [
  "requestHeader",
         "subscriptionIds",
         "sendInitialValues"
];


exports.TransferSubscriptionsRequest = TransferSubscriptionsRequest;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("TransferSubscriptionsRequest",TransferSubscriptionsRequest);
