// node-opcua-service-subscription/_generated_/_auto_generated_MonitoringParameters.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:34.492Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/MonitoringParameters_schema").MonitoringParameters_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class MonitoringParameters
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.clientHandle] {IntegerId} 
 * @param  [options.samplingInterval] {Duration} 
 * @param  [options.filter] {ExtensionObject} 
 * @param  [options.queueSize] {Counter} 
 * @param  [options.discardOldest] {Boolean} 
 */
function MonitoringParameters(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property clientHandle
      * @type {IntegerId}
      */
    self.clientHandle = initialize_field(schema.fields[0], options.clientHandle);

    /**
      * 
      * @property samplingInterval
      * @type {Duration}
      */
    self.samplingInterval = initialize_field(schema.fields[1], options.samplingInterval);

    /**
      * 
      * @property filter
      * @type {ExtensionObject}
      */
    self.filter = initialize_field(schema.fields[2], options.filter);

    /**
      * 
      * @property queueSize
      * @type {Counter}
      */
    self.queueSize = initialize_field(schema.fields[3], options.queueSize);

    /**
      * 
      * @property discardOldest
      * @type {Boolean}
      */
    self.discardOldest = initialize_field(schema.fields[4], options.discardOldest);

   // Object.preventExtensions(self);
}
util.inherits(MonitoringParameters,BaseUAObject);
MonitoringParameters.prototype.encodingDefaultBinary = makeExpandedNodeId(742,0);
MonitoringParameters.prototype.encodingDefaultXml = makeExpandedNodeId(741,0);
MonitoringParameters.prototype._schema = schema;

var encode_IntegerId = _defaultTypeMap.IntegerId.encode;
var decode_IntegerId = _defaultTypeMap.IntegerId.decode;
var encode_Duration = _defaultTypeMap.Duration.encode;
var decode_Duration = _defaultTypeMap.Duration.decode;
var encode_ExtensionObject = _defaultTypeMap.ExtensionObject.encode;
var decode_ExtensionObject = _defaultTypeMap.ExtensionObject.decode;
var encode_Counter = _defaultTypeMap.Counter.encode;
var decode_Counter = _defaultTypeMap.Counter.decode;
var encode_Boolean = _defaultTypeMap.Boolean.encode;
var decode_Boolean = _defaultTypeMap.Boolean.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
MonitoringParameters.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_IntegerId(this.clientHandle,stream);
    encode_Duration(this.samplingInterval,stream);
    encode_ExtensionObject(this.filter,stream);
    encode_Counter(this.queueSize,stream);
    encode_Boolean(this.discardOldest,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
MonitoringParameters.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.clientHandle = decode_IntegerId(stream,options);
    this.samplingInterval = decode_Duration(stream,options);
    this.filter = decode_ExtensionObject(stream,options);
    this.queueSize = decode_Counter(stream,options);
    this.discardOldest = decode_Boolean(stream,options);
};
MonitoringParameters.possibleFields = [
  "clientHandle",
         "samplingInterval",
         "filter",
         "queueSize",
         "discardOldest"
];


exports.MonitoringParameters = MonitoringParameters;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("MonitoringParameters",MonitoringParameters);
