// node-opcua-service-subscription/_generated_/_auto_generated_StatusChangeNotification.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:34.673Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/StatusChangeNotification_schema").StatusChangeNotification_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var DiagnosticInfo = getFactory("DiagnosticInfo");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class StatusChangeNotification
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.statusCode] {StatusCode} The StatusCode that indicates the status change.
 * @param  [options.diagnosticInfo] {DiagnosticInfo} 
 */
function StatusChangeNotification(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.diagnosticInfo =  null; /* new DiagnosticInfo(null); */
        return ;
    }

    /**
      * The StatusCode that indicates the status change.
      * @property statusCode
      * @type {StatusCode}
      */
    self.statusCode = initialize_field(schema.fields[0], options.statusCode);

    /**
      * 
      * @property diagnosticInfo
      * @type {DiagnosticInfo}
      */
    self.diagnosticInfo =  new DiagnosticInfo( options.diagnosticInfo);

   // Object.preventExtensions(self);
}
util.inherits(StatusChangeNotification,BaseUAObject);
StatusChangeNotification.prototype.encodingDefaultBinary = makeExpandedNodeId(820,0);
StatusChangeNotification.prototype.encodingDefaultXml = makeExpandedNodeId(819,0);
StatusChangeNotification.prototype._schema = schema;

var encode_StatusCode = _defaultTypeMap.StatusCode.encode;
var decode_StatusCode = _defaultTypeMap.StatusCode.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
StatusChangeNotification.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_StatusCode(this.statusCode,stream);
   this.diagnosticInfo.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
StatusChangeNotification.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.statusCode = decode_StatusCode(stream,options);
    this.diagnosticInfo.decode(stream,options);
};
StatusChangeNotification.possibleFields = [
  "statusCode",
         "diagnosticInfo"
];


exports.StatusChangeNotification = StatusChangeNotification;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("StatusChangeNotification",StatusChangeNotification);
