// node-opcua-service-subscription/_generated_/_auto_generated_SetTriggeringRequest.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:34.679Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/SetTriggeringRequest_schema").SetTriggeringRequest_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var RequestHeader = getFactory("RequestHeader");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class SetTriggeringRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} A standard header included in all requests sent to a server.
 * @param  [options.subscriptionId] {IntegerId} 
 * @param  [options.triggeringItemId] {IntegerId} 
 * @param  [options.linksToAdd] {IntegerId[]} 
 * @param  [options.linksToRemove] {IntegerId[]} 
 */
function SetTriggeringRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestHeader =  null; /* new RequestHeader(null); */
        return ;
    }

    /**
      * A standard header included in all requests sent to a server.
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * 
      * @property subscriptionId
      * @type {IntegerId}
      */
    self.subscriptionId = initialize_field(schema.fields[1], options.subscriptionId);

    /**
      * 
      * @property triggeringItemId
      * @type {IntegerId}
      */
    self.triggeringItemId = initialize_field(schema.fields[2], options.triggeringItemId);

    /**
      * 
      * @property linksToAdd
      * @type {IntegerId[]}
      */
    self.linksToAdd = initialize_field_array(schema.fields[3], options.linksToAdd);

    /**
      * 
      * @property linksToRemove
      * @type {IntegerId[]}
      */
    self.linksToRemove = initialize_field_array(schema.fields[4], options.linksToRemove);

   // Object.preventExtensions(self);
}
util.inherits(SetTriggeringRequest,BaseUAObject);
SetTriggeringRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(775,0);
SetTriggeringRequest.prototype.encodingDefaultXml = makeExpandedNodeId(774,0);
SetTriggeringRequest.prototype._schema = schema;

var encode_IntegerId = _defaultTypeMap.IntegerId.encode;
var decode_IntegerId = _defaultTypeMap.IntegerId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
SetTriggeringRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encode_IntegerId(this.subscriptionId,stream);
    encode_IntegerId(this.triggeringItemId,stream);
    encodeArray(this.linksToAdd, stream, encode_IntegerId);
    encodeArray(this.linksToRemove, stream, encode_IntegerId);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
SetTriggeringRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.subscriptionId = decode_IntegerId(stream,options);
    this.triggeringItemId = decode_IntegerId(stream,options);
    this.linksToAdd = decodeArray(stream, decode_IntegerId);
    this.linksToRemove = decodeArray(stream, decode_IntegerId);
};
SetTriggeringRequest.possibleFields = [
  "requestHeader",
         "subscriptionId",
         "triggeringItemId",
         "linksToAdd",
         "linksToRemove"
];


exports.SetTriggeringRequest = SetTriggeringRequest;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("SetTriggeringRequest",SetTriggeringRequest);
