// node-opcua-service-subscription/_generated_/_auto_generated_CreateSubscriptionResponse.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:34.477Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/CreateSubscriptionResponse_schema").CreateSubscriptionResponse_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var ResponseHeader = getFactory("ResponseHeader");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class CreateSubscriptionResponse
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.responseHeader] {ResponseHeader} 
 * @param  [options.subscriptionId] {IntegerId} 
 * @param  [options.revisedPublishingInterval] {Duration} 
 * @param  [options.revisedLifetimeCount] {Counter} 
 * @param  [options.revisedMaxKeepAliveCount] {Counter} 
 */
function CreateSubscriptionResponse(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.responseHeader =  null; /* new ResponseHeader(null); */
        return ;
    }

    /**
      * 
      * @property responseHeader
      * @type {ResponseHeader}
      */
    self.responseHeader =  new ResponseHeader( options.responseHeader);

    /**
      * 
      * @property subscriptionId
      * @type {IntegerId}
      */
    self.subscriptionId = initialize_field(schema.fields[1], options.subscriptionId);

    /**
      * 
      * @property revisedPublishingInterval
      * @type {Duration}
      */
    self.revisedPublishingInterval = initialize_field(schema.fields[2], options.revisedPublishingInterval);

    /**
      * 
      * @property revisedLifetimeCount
      * @type {Counter}
      */
    self.revisedLifetimeCount = initialize_field(schema.fields[3], options.revisedLifetimeCount);

    /**
      * 
      * @property revisedMaxKeepAliveCount
      * @type {Counter}
      */
    self.revisedMaxKeepAliveCount = initialize_field(schema.fields[4], options.revisedMaxKeepAliveCount);

   // Object.preventExtensions(self);
}
util.inherits(CreateSubscriptionResponse,BaseUAObject);
CreateSubscriptionResponse.prototype.encodingDefaultBinary = makeExpandedNodeId(790,0);
CreateSubscriptionResponse.prototype.encodingDefaultXml = makeExpandedNodeId(789,0);
CreateSubscriptionResponse.prototype._schema = schema;

var encode_IntegerId = _defaultTypeMap.IntegerId.encode;
var decode_IntegerId = _defaultTypeMap.IntegerId.decode;
var encode_Duration = _defaultTypeMap.Duration.encode;
var decode_Duration = _defaultTypeMap.Duration.decode;
var encode_Counter = _defaultTypeMap.Counter.encode;
var decode_Counter = _defaultTypeMap.Counter.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
CreateSubscriptionResponse.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.responseHeader.encode(stream,options);
    encode_IntegerId(this.subscriptionId,stream);
    encode_Duration(this.revisedPublishingInterval,stream);
    encode_Counter(this.revisedLifetimeCount,stream);
    encode_Counter(this.revisedMaxKeepAliveCount,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
CreateSubscriptionResponse.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.responseHeader.decode(stream,options);
    this.subscriptionId = decode_IntegerId(stream,options);
    this.revisedPublishingInterval = decode_Duration(stream,options);
    this.revisedLifetimeCount = decode_Counter(stream,options);
    this.revisedMaxKeepAliveCount = decode_Counter(stream,options);
};
CreateSubscriptionResponse.possibleFields = [
  "responseHeader",
         "subscriptionId",
         "revisedPublishingInterval",
         "revisedLifetimeCount",
         "revisedMaxKeepAliveCount"
];


exports.CreateSubscriptionResponse = CreateSubscriptionResponse;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("CreateSubscriptionResponse",CreateSubscriptionResponse);
