// node-opcua-service-subscription/_generated_/_auto_generated_MonitoredItemModifyResult.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:34.620Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/MonitoredItemModifyResult_schema").MonitoredItemModifyResult_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class MonitoredItemModifyResult
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.statusCode] {StatusCode} 
 * @param  [options.revisedSamplingInterval] {Duration} 
 * @param  [options.revisedQueueSize] {Counter} 
 * @param  [options.filterResult] {ExtensionObject} 
 */
function MonitoredItemModifyResult(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property statusCode
      * @type {StatusCode}
      */
    self.statusCode = initialize_field(schema.fields[0], options.statusCode);

    /**
      * 
      * @property revisedSamplingInterval
      * @type {Duration}
      */
    self.revisedSamplingInterval = initialize_field(schema.fields[1], options.revisedSamplingInterval);

    /**
      * 
      * @property revisedQueueSize
      * @type {Counter}
      */
    self.revisedQueueSize = initialize_field(schema.fields[2], options.revisedQueueSize);

    /**
      * 
      * @property filterResult
      * @type {ExtensionObject}
      */
    self.filterResult = initialize_field(schema.fields[3], options.filterResult);

   // Object.preventExtensions(self);
}
util.inherits(MonitoredItemModifyResult,BaseUAObject);
MonitoredItemModifyResult.prototype.encodingDefaultBinary = makeExpandedNodeId(760,0);
MonitoredItemModifyResult.prototype.encodingDefaultXml = makeExpandedNodeId(759,0);
MonitoredItemModifyResult.prototype._schema = schema;

var encode_StatusCode = _defaultTypeMap.StatusCode.encode;
var decode_StatusCode = _defaultTypeMap.StatusCode.decode;
var encode_Duration = _defaultTypeMap.Duration.encode;
var decode_Duration = _defaultTypeMap.Duration.decode;
var encode_Counter = _defaultTypeMap.Counter.encode;
var decode_Counter = _defaultTypeMap.Counter.decode;
var encode_ExtensionObject = _defaultTypeMap.ExtensionObject.encode;
var decode_ExtensionObject = _defaultTypeMap.ExtensionObject.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
MonitoredItemModifyResult.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_StatusCode(this.statusCode,stream);
    encode_Duration(this.revisedSamplingInterval,stream);
    encode_Counter(this.revisedQueueSize,stream);
    encode_ExtensionObject(this.filterResult,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
MonitoredItemModifyResult.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.statusCode = decode_StatusCode(stream,options);
    this.revisedSamplingInterval = decode_Duration(stream,options);
    this.revisedQueueSize = decode_Counter(stream,options);
    this.filterResult = decode_ExtensionObject(stream,options);
};
MonitoredItemModifyResult.possibleFields = [
  "statusCode",
         "revisedSamplingInterval",
         "revisedQueueSize",
         "filterResult"
];


exports.MonitoredItemModifyResult = MonitoredItemModifyResult;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("MonitoredItemModifyResult",MonitoredItemModifyResult);
