// node-opcua-service-subscription/_generated_/_auto_generated_MonitoredItemModifyRequest.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:34.614Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/MonitoredItemModifyRequest_schema").MonitoredItemModifyRequest_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var MonitoringParameters = require("./_auto_generated_MonitoringParameters").MonitoringParameters;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class MonitoredItemModifyRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.monitoredItemId] {IntegerId} 
 * @param  [options.requestedParameters] {MonitoringParameters} 
 */
function MonitoredItemModifyRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestedParameters =  null; /* new MonitoringParameters(null); */
        return ;
    }

    /**
      * 
      * @property monitoredItemId
      * @type {IntegerId}
      */
    self.monitoredItemId = initialize_field(schema.fields[0], options.monitoredItemId);

    /**
      * 
      * @property requestedParameters
      * @type {MonitoringParameters}
      */
    self.requestedParameters =  new MonitoringParameters( options.requestedParameters);

   // Object.preventExtensions(self);
}
util.inherits(MonitoredItemModifyRequest,BaseUAObject);
MonitoredItemModifyRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(757,0);
MonitoredItemModifyRequest.prototype.encodingDefaultXml = makeExpandedNodeId(756,0);
MonitoredItemModifyRequest.prototype._schema = schema;

var encode_IntegerId = _defaultTypeMap.IntegerId.encode;
var decode_IntegerId = _defaultTypeMap.IntegerId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
MonitoredItemModifyRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_IntegerId(this.monitoredItemId,stream);
   this.requestedParameters.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
MonitoredItemModifyRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.monitoredItemId = decode_IntegerId(stream,options);
    this.requestedParameters.decode(stream,options);
};
MonitoredItemModifyRequest.possibleFields = [
  "monitoredItemId",
         "requestedParameters"
];


exports.MonitoredItemModifyRequest = MonitoredItemModifyRequest;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("MonitoredItemModifyRequest",MonitoredItemModifyRequest);
