// node-opcua-service-subscription/_generated_/_auto_generated_AggregateFilter.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:34.729Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/AggregateFilter_schema").AggregateFilter_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var AggregateConfiguration = getFactory("AggregateConfiguration");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class AggregateFilter
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.StartTime] {DateTime} 
 * @param  [options.AggregateType] {NodeId} 
 * @param  [options.ProcessingInterval] {Double} 
 * @param  [options.AggregateType] {AggregateConfiguration} 
 */
function AggregateFilter(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.AggregateType =  null; /* new AggregateConfiguration(null); */
        return ;
    }

    /**
      * 
      * @property StartTime
      * @type {DateTime}
      */
    self.StartTime = initialize_field(schema.fields[0], options.StartTime);

    /**
      * 
      * @property AggregateType
      * @type {NodeId}
      */
    self.AggregateType = initialize_field(schema.fields[1], options.AggregateType);

    /**
      * 
      * @property ProcessingInterval
      * @type {Double}
      */
    self.ProcessingInterval = initialize_field(schema.fields[2], options.ProcessingInterval);

    /**
      * 
      * @property AggregateType
      * @type {AggregateConfiguration}
      */
    self.AggregateType =  new AggregateConfiguration( options.AggregateType);

   // Object.preventExtensions(self);
}
util.inherits(AggregateFilter,BaseUAObject);
AggregateFilter.prototype.encodingDefaultBinary = makeExpandedNodeId(730,0);
AggregateFilter.prototype.encodingDefaultXml = makeExpandedNodeId(729,0);
AggregateFilter.prototype._schema = schema;

var encode_DateTime = _defaultTypeMap.DateTime.encode;
var decode_DateTime = _defaultTypeMap.DateTime.decode;
var encode_NodeId = _defaultTypeMap.NodeId.encode;
var decode_NodeId = _defaultTypeMap.NodeId.decode;
var encode_Double = _defaultTypeMap.Double.encode;
var decode_Double = _defaultTypeMap.Double.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
AggregateFilter.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_DateTime(this.StartTime,stream);
    encode_NodeId(this.AggregateType,stream);
    encode_Double(this.ProcessingInterval,stream);
   this.AggregateType.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
AggregateFilter.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.StartTime = decode_DateTime(stream,options);
    this.AggregateType = decode_NodeId(stream,options);
    this.ProcessingInterval = decode_Double(stream,options);
    this.AggregateType.decode(stream,options);
};
AggregateFilter.possibleFields = [
  "StartTime",
         "AggregateType",
         "ProcessingInterval",
         "AggregateType"
];


exports.AggregateFilter = AggregateFilter;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("AggregateFilter",AggregateFilter);
