// node-opcua-service-subscription/_generated_/_auto_generated_CreateSubscriptionRequest.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:34.454Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/CreateSubscriptionRequest_schema").CreateSubscriptionRequest_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var RequestHeader = getFactory("RequestHeader");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class CreateSubscriptionRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} 
 * @param  [options.requestedPublishingInterval] {Duration} 
 * @param  [options.requestedLifetimeCount] {Counter} 
 * @param  [options.requestedMaxKeepAliveCount] {Counter} 
 * @param  [options.maxNotificationsPerPublish] {Counter} 
 * @param  [options.publishingEnabled] {Boolean} 
 * @param  [options.priority] {Byte} 
 */
function CreateSubscriptionRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestHeader =  null; /* new RequestHeader(null); */
        return ;
    }

    /**
      * 
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * 
      * @property requestedPublishingInterval
      * @type {Duration}
      */
    self.requestedPublishingInterval = initialize_field(schema.fields[1], options.requestedPublishingInterval);

    /**
      * 
      * @property requestedLifetimeCount
      * @type {Counter}
      */
    self.requestedLifetimeCount = initialize_field(schema.fields[2], options.requestedLifetimeCount);

    /**
      * 
      * @property requestedMaxKeepAliveCount
      * @type {Counter}
      */
    self.requestedMaxKeepAliveCount = initialize_field(schema.fields[3], options.requestedMaxKeepAliveCount);

    /**
      * 
      * @property maxNotificationsPerPublish
      * @type {Counter}
      */
    self.maxNotificationsPerPublish = initialize_field(schema.fields[4], options.maxNotificationsPerPublish);

    /**
      * 
      * @property publishingEnabled
      * @type {Boolean}
      */
    self.publishingEnabled = initialize_field(schema.fields[5], options.publishingEnabled);

    /**
      * 
      * @property priority
      * @type {Byte}
      */
    self.priority = initialize_field(schema.fields[6], options.priority);

   // Object.preventExtensions(self);
}
util.inherits(CreateSubscriptionRequest,BaseUAObject);
CreateSubscriptionRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(787,0);
CreateSubscriptionRequest.prototype.encodingDefaultXml = makeExpandedNodeId(786,0);
CreateSubscriptionRequest.prototype._schema = schema;

var encode_Duration = _defaultTypeMap.Duration.encode;
var decode_Duration = _defaultTypeMap.Duration.decode;
var encode_Counter = _defaultTypeMap.Counter.encode;
var decode_Counter = _defaultTypeMap.Counter.decode;
var encode_Boolean = _defaultTypeMap.Boolean.encode;
var decode_Boolean = _defaultTypeMap.Boolean.decode;
var encode_Byte = _defaultTypeMap.Byte.encode;
var decode_Byte = _defaultTypeMap.Byte.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
CreateSubscriptionRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encode_Duration(this.requestedPublishingInterval,stream);
    encode_Counter(this.requestedLifetimeCount,stream);
    encode_Counter(this.requestedMaxKeepAliveCount,stream);
    encode_Counter(this.maxNotificationsPerPublish,stream);
    encode_Boolean(this.publishingEnabled,stream);
    encode_Byte(this.priority,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
CreateSubscriptionRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.requestedPublishingInterval = decode_Duration(stream,options);
    this.requestedLifetimeCount = decode_Counter(stream,options);
    this.requestedMaxKeepAliveCount = decode_Counter(stream,options);
    this.maxNotificationsPerPublish = decode_Counter(stream,options);
    this.publishingEnabled = decode_Boolean(stream,options);
    this.priority = decode_Byte(stream,options);
};
CreateSubscriptionRequest.possibleFields = [
  "requestHeader",
         "requestedPublishingInterval",
         "requestedLifetimeCount",
         "requestedMaxKeepAliveCount",
         "maxNotificationsPerPublish",
         "publishingEnabled",
         "priority"
];


exports.CreateSubscriptionRequest = CreateSubscriptionRequest;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("CreateSubscriptionRequest",CreateSubscriptionRequest);
