// node-opcua-service-subscription/_generated_/_auto_generated_ModifyMonitoredItemsRequest.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:34.626Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ModifyMonitoredItemsRequest_schema").ModifyMonitoredItemsRequest_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var RequestHeader = getFactory("RequestHeader");
var MonitoredItemModifyRequest = require("./_auto_generated_MonitoredItemModifyRequest").MonitoredItemModifyRequest;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ModifyMonitoredItemsRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} 
 * @param  [options.subscriptionId] {IntegerId} 
 * @param  [options.timestampsToReturn] {TimestampsToReturn} 
 * @param  [options.itemsToModify] {MonitoredItemModifyRequest[]} 
 */
function ModifyMonitoredItemsRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestHeader =  null; /* new RequestHeader(null); */
        self.itemsToModify =  null; /* null array */
        return ;
    }

    /**
      * 
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * 
      * @property subscriptionId
      * @type {IntegerId}
      */
    self.subscriptionId = initialize_field(schema.fields[1], options.subscriptionId);

    /**
      * 
      * @property timestampsToReturn
      * @type {TimestampsToReturn}
      */
    self.setTimestampsToReturn(initialize_field(schema.fields[2], options.timestampsToReturn));

    /**
      * 
      * @property itemsToModify
      * @type {MonitoredItemModifyRequest[]}
      */
    self.itemsToModify = [];
    if (options.itemsToModify) {
        assert(_.isArray(options.itemsToModify));
        self.itemsToModify = options.itemsToModify.map(function(e){ return new MonitoredItemModifyRequest(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(ModifyMonitoredItemsRequest,BaseUAObject);

//## Define Enumeration setters
ModifyMonitoredItemsRequest.prototype.setTimestampsToReturn = function(value) {
   var coercedValue = _enumerations.TimestampsToReturn.typedEnum.get(value);
   /* istanbul ignore next */
   if (coercedValue === undefined || coercedValue === null) {
      throw new Error("value cannot be coerced to TimestampsToReturn: " + value);
   }
   this.timestampsToReturn = coercedValue;
};
ModifyMonitoredItemsRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(763,0);
ModifyMonitoredItemsRequest.prototype.encodingDefaultXml = makeExpandedNodeId(762,0);
ModifyMonitoredItemsRequest.prototype._schema = schema;

var encode_IntegerId = _defaultTypeMap.IntegerId.encode;
var decode_IntegerId = _defaultTypeMap.IntegerId.decode;
var encode_TimestampsToReturn = _enumerations.TimestampsToReturn.encode;
var decode_TimestampsToReturn = _enumerations.TimestampsToReturn.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ModifyMonitoredItemsRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encode_IntegerId(this.subscriptionId,stream);
    encode_TimestampsToReturn(this.timestampsToReturn,stream);
    encodeArray(this.itemsToModify,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
ModifyMonitoredItemsRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.subscriptionId = decode_IntegerId(stream,options);
    this.timestampsToReturn = decode_TimestampsToReturn(stream,options);
    this.itemsToModify = decodeArray(stream, function(stream) { 
       var obj = new MonitoredItemModifyRequest(null);
       obj.decode(stream,options);
       return obj; 
    });
};
ModifyMonitoredItemsRequest.possibleFields = [
  "requestHeader",
         "subscriptionId",
         "timestampsToReturn",
         "itemsToModify"
];


exports.ModifyMonitoredItemsRequest = ModifyMonitoredItemsRequest;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ModifyMonitoredItemsRequest",ModifyMonitoredItemsRequest);
