// node-opcua-service-subscription/_generated_/_auto_generated_SubscriptionAcknowledgement.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:34.524Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/SubscriptionAcknowledgement_schema").SubscriptionAcknowledgement_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class SubscriptionAcknowledgement
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.subscriptionId] {IntegerId} 
 * @param  [options.sequenceNumber] {Counter} 
 */
function SubscriptionAcknowledgement(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property subscriptionId
      * @type {IntegerId}
      */
    self.subscriptionId = initialize_field(schema.fields[0], options.subscriptionId);

    /**
      * 
      * @property sequenceNumber
      * @type {Counter}
      */
    self.sequenceNumber = initialize_field(schema.fields[1], options.sequenceNumber);

   // Object.preventExtensions(self);
}
util.inherits(SubscriptionAcknowledgement,BaseUAObject);
SubscriptionAcknowledgement.prototype.encodingDefaultBinary = makeExpandedNodeId(823,0);
SubscriptionAcknowledgement.prototype.encodingDefaultXml = makeExpandedNodeId(822,0);
SubscriptionAcknowledgement.prototype._schema = schema;

var encode_IntegerId = _defaultTypeMap.IntegerId.encode;
var decode_IntegerId = _defaultTypeMap.IntegerId.decode;
var encode_Counter = _defaultTypeMap.Counter.encode;
var decode_Counter = _defaultTypeMap.Counter.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
SubscriptionAcknowledgement.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_IntegerId(this.subscriptionId,stream);
    encode_Counter(this.sequenceNumber,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
SubscriptionAcknowledgement.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.subscriptionId = decode_IntegerId(stream,options);
    this.sequenceNumber = decode_Counter(stream,options);
};
SubscriptionAcknowledgement.possibleFields = [
  "subscriptionId",
         "sequenceNumber"
];


exports.SubscriptionAcknowledgement = SubscriptionAcknowledgement;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("SubscriptionAcknowledgement",SubscriptionAcknowledgement);
