// node-opcua-service-subscription/_generated_/_auto_generated_SetTriggeringResponse.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:34.687Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/SetTriggeringResponse_schema").SetTriggeringResponse_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var ResponseHeader = getFactory("ResponseHeader");
var DiagnosticInfo = getFactory("DiagnosticInfo");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class SetTriggeringResponse
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.responseHeader] {ResponseHeader} 
 * @param  [options.addResults] {StatusCode[]} 
 * @param  [options.addDiagnosticInfos] {DiagnosticInfo[]} 
 * @param  [options.removeResults] {StatusCode[]} 
 * @param  [options.removeDiagnosticInfos] {DiagnosticInfo[]} 
 */
function SetTriggeringResponse(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.responseHeader =  null; /* new ResponseHeader(null); */
        self.addDiagnosticInfos =  null; /* null array */
        self.removeDiagnosticInfos =  null; /* null array */
        return ;
    }

    /**
      * 
      * @property responseHeader
      * @type {ResponseHeader}
      */
    self.responseHeader =  new ResponseHeader( options.responseHeader);

    /**
      * 
      * @property addResults
      * @type {StatusCode[]}
      */
    self.addResults = initialize_field_array(schema.fields[1], options.addResults);

    /**
      * 
      * @property addDiagnosticInfos
      * @type {DiagnosticInfo[]}
      */
    self.addDiagnosticInfos = [];
    if (options.addDiagnosticInfos) {
        assert(_.isArray(options.addDiagnosticInfos));
        self.addDiagnosticInfos = options.addDiagnosticInfos.map(function(e){ return new DiagnosticInfo(e); } );
    }

    /**
      * 
      * @property removeResults
      * @type {StatusCode[]}
      */
    self.removeResults = initialize_field_array(schema.fields[3], options.removeResults);

    /**
      * 
      * @property removeDiagnosticInfos
      * @type {DiagnosticInfo[]}
      */
    self.removeDiagnosticInfos = [];
    if (options.removeDiagnosticInfos) {
        assert(_.isArray(options.removeDiagnosticInfos));
        self.removeDiagnosticInfos = options.removeDiagnosticInfos.map(function(e){ return new DiagnosticInfo(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(SetTriggeringResponse,BaseUAObject);
SetTriggeringResponse.prototype.encodingDefaultBinary = makeExpandedNodeId(778,0);
SetTriggeringResponse.prototype.encodingDefaultXml = makeExpandedNodeId(777,0);
SetTriggeringResponse.prototype._schema = schema;

var encode_StatusCode = _defaultTypeMap.StatusCode.encode;
var decode_StatusCode = _defaultTypeMap.StatusCode.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
SetTriggeringResponse.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.responseHeader.encode(stream,options);
    encodeArray(this.addResults, stream, encode_StatusCode);
    encodeArray(this.addDiagnosticInfos,stream,function(obj,stream){ obj.encode(stream,options); }); 
    encodeArray(this.removeResults, stream, encode_StatusCode);
    encodeArray(this.removeDiagnosticInfos,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
SetTriggeringResponse.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.responseHeader.decode(stream,options);
    this.addResults = decodeArray(stream, decode_StatusCode);
    this.addDiagnosticInfos = decodeArray(stream, function(stream) { 
       var obj = new DiagnosticInfo(null);
       obj.decode(stream,options);
       return obj; 
    });
    this.removeResults = decodeArray(stream, decode_StatusCode);
    this.removeDiagnosticInfos = decodeArray(stream, function(stream) { 
       var obj = new DiagnosticInfo(null);
       obj.decode(stream,options);
       return obj; 
    });
};
SetTriggeringResponse.possibleFields = [
  "responseHeader",
         "addResults",
         "addDiagnosticInfos",
         "removeResults",
         "removeDiagnosticInfos"
];


exports.SetTriggeringResponse = SetTriggeringResponse;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("SetTriggeringResponse",SetTriggeringResponse);
