// node-opcua-data-access/_generated_/_auto_generated_EUInformation.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:07.915Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/EUInformation_schema").EUInformation_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const LocalizedText = getFactory("LocalizedText");
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class EUInformation
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function EUInformation(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.displayName =  null; /* new LocalizedText(null); */
        self.description =  null; /* new LocalizedText(null); */
        return ;
    }

    /**
      * Identifies the organization (company, standards organization) that defines the EUInformation.
      * @property namespaceUri
      * @type {String}
      * @default  http://www.opcfoundation.org/UA/units/un/cefact
      */
    self.namespaceUri = initialize_field(schema.fields[0], options.namespaceUri);

    /**
      * Identifier for programmatic evaluation. 1 is used if a unitId is not available.
      * @property unitId
      * @type {Int32}
      */
    self.unitId = initialize_field(schema.fields[1], options.unitId);

    /**
      * The displayName of the engineering  ( for instance 'm/s' )
      * @property displayName
      * @type {LocalizedText}
      */
    self.displayName =  new LocalizedText( options.displayName);

    /**
      * Contains the full name of the engineering unit such as hour or meter per second
      * @property description
      * @type {LocalizedText}
      */
    self.description =  new LocalizedText( options.description);

   // Object.preventExtensions(self);
}
util.inherits(EUInformation,BaseUAObject);
EUInformation.prototype.encodingDefaultBinary = makeExpandedNodeId(889,0);
EUInformation.prototype.encodingDefaultXml = makeExpandedNodeId(888,0);
EUInformation.prototype._schema = schema;

const encode_String = _defaultTypeMap.String.encode;
const decode_String = _defaultTypeMap.String.decode;
const encode_Int32 = _defaultTypeMap.Int32.encode;
const decode_Int32 = _defaultTypeMap.Int32.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
EUInformation.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_String(this.namespaceUri,stream);
    encode_Int32(this.unitId,stream);
   this.displayName.encode(stream,options);
   this.description.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
EUInformation.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.namespaceUri = decode_String(stream);
    this.unitId = decode_Int32(stream);
    this.displayName.decode(stream);
    this.description.decode(stream);
};
EUInformation.possibleFields = [
  "namespaceUri",
         "unitId",
         "displayName",
         "description"
];


exports.EUInformation = EUInformation;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("EUInformation",EUInformation);
