// node-opcua-service-history/_generated_/_auto_generated_ReadAtTimeDetails.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:16.719Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/ReadAtTimeDetails_schema").ReadAtTimeDetails_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;
const HistoryReadDetails = require("./_auto_generated_HistoryReadDetails").HistoryReadDetails;

/**
 * 
 * @class ReadAtTimeDetails
 * @constructor
 * @extends HistoryReadDetails
 * @param  options {Object}
 */
function ReadAtTimeDetails(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    HistoryReadDetails.call(this,options);
    if (options === null) { 
        HistoryReadDetails.call(this,options);
        return ;
    }

    /**
      * 
      * @property reqTimes
      * @type {DateTime[]}
      */
    self.reqTimes = initialize_field_array(schema.fields[0], options.reqTimes);

    /**
      * 
      * @property useSimpleBounds
      * @type {Boolean}
      */
    self.useSimpleBounds = initialize_field(schema.fields[1], options.useSimpleBounds);

   // Object.preventExtensions(self);
}
util.inherits(ReadAtTimeDetails,HistoryReadDetails);
ReadAtTimeDetails.prototype.encodingDefaultBinary = makeExpandedNodeId(655,0);
ReadAtTimeDetails.prototype.encodingDefaultXml = makeExpandedNodeId(654,0);
ReadAtTimeDetails.prototype._schema = schema;

const encode_DateTime = _defaultTypeMap.DateTime.encode;
const decode_DateTime = _defaultTypeMap.DateTime.decode;
const encode_Boolean = _defaultTypeMap.Boolean.encode;
const decode_Boolean = _defaultTypeMap.Boolean.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ReadAtTimeDetails.prototype.encode = function(stream,options) {
    // call base class implementation first
    HistoryReadDetails.prototype.encode.call(this,stream,options);
    encodeArray(this.reqTimes, stream, encode_DateTime);
    encode_Boolean(this.useSimpleBounds,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
ReadAtTimeDetails.prototype.decode = function(stream) {
    // call base class implementation first
    HistoryReadDetails.prototype.decode.call(this,stream);
    this.reqTimes = decodeArray(stream, decode_DateTime);
    this.useSimpleBounds = decode_Boolean(stream);
};
ReadAtTimeDetails.possibleFields = [
  "reqTimes",
         "useSimpleBounds"
];


exports.ReadAtTimeDetails = ReadAtTimeDetails;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ReadAtTimeDetails",ReadAtTimeDetails);
