// node-opcua-service-history/_generated_/_auto_generated_HistoryModifiedData.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:16.696Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/HistoryModifiedData_schema").HistoryModifiedData_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const DataValue = getFactory("DataValue");
const ModificationInfo = require("./_auto_generated_ModificationInfo").ModificationInfo;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class HistoryModifiedData
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function HistoryModifiedData(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.dataValues =  null; /* null array */
        self.modificationInfos =  null; /* null array */
        return ;
    }

    /**
      * An array of values of history data for the Node. The size of the array depends on the requested data parameters.
      * @property dataValues
      * @type {DataValue[]}
      */
    self.dataValues = [];
    if (options.dataValues) {
        assert(_.isArray(options.dataValues));
        self.dataValues = options.dataValues.map(function(e){ return new DataValue(e); } );
    }

    /**
      * 
      * @property modificationInfos
      * @type {ModificationInfo[]}
      */
    self.modificationInfos = [];
    if (options.modificationInfos) {
        assert(_.isArray(options.modificationInfos));
        self.modificationInfos = options.modificationInfos.map(function(e){ return new ModificationInfo(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(HistoryModifiedData,BaseUAObject);
HistoryModifiedData.prototype.encodingDefaultBinary = makeExpandedNodeId(11227,0);
HistoryModifiedData.prototype.encodingDefaultXml = makeExpandedNodeId(11219,0);
HistoryModifiedData.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
HistoryModifiedData.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encodeArray(this.dataValues,stream,function(obj,stream){ obj.encode(stream,options); }); 
    encodeArray(this.modificationInfos,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
HistoryModifiedData.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.dataValues = decodeArray(stream, function(stream) { 
       const obj = new DataValue(null);
       obj.decode(stream);
       return obj; 
    });
    this.modificationInfos = decodeArray(stream, function(stream) { 
       const obj = new ModificationInfo(null);
       obj.decode(stream);
       return obj; 
    });
};
HistoryModifiedData.possibleFields = [
  "dataValues",
         "modificationInfos"
];


exports.HistoryModifiedData = HistoryModifiedData;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("HistoryModifiedData",HistoryModifiedData);
