// node-opcua-service-history/_generated_/_auto_generated_ReadRawModifiedDetails.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:16.711Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/ReadRawModifiedDetails_schema").ReadRawModifiedDetails_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;
const HistoryReadDetails = require("./_auto_generated_HistoryReadDetails").HistoryReadDetails;

/**
 * 
 * @class ReadRawModifiedDetails
 * @constructor
 * @extends HistoryReadDetails
 * @param  options {Object}
 */
function ReadRawModifiedDetails(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    HistoryReadDetails.call(this,options);
    if (options === null) { 
        HistoryReadDetails.call(this,options);
        return ;
    }

    /**
      * 
      * @property isReadModified
      * @type {Boolean}
      */
    self.isReadModified = initialize_field(schema.fields[0], options.isReadModified);

    /**
      * 
      * @property startTime
      * @type {DateTime}
      */
    self.startTime = initialize_field(schema.fields[1], options.startTime);

    /**
      * 
      * @property endTime
      * @type {DateTime}
      */
    self.endTime = initialize_field(schema.fields[2], options.endTime);

    /**
      * 
      * @property numValuesPerNode
      * @type {UInt32}
      */
    self.numValuesPerNode = initialize_field(schema.fields[3], options.numValuesPerNode);

    /**
      * 
      * @property returnBounds
      * @type {Boolean}
      */
    self.returnBounds = initialize_field(schema.fields[4], options.returnBounds);

   // Object.preventExtensions(self);
}
util.inherits(ReadRawModifiedDetails,HistoryReadDetails);
ReadRawModifiedDetails.prototype.encodingDefaultBinary = makeExpandedNodeId(649,0);
ReadRawModifiedDetails.prototype.encodingDefaultXml = makeExpandedNodeId(648,0);
ReadRawModifiedDetails.prototype._schema = schema;

const encode_Boolean = _defaultTypeMap.Boolean.encode;
const decode_Boolean = _defaultTypeMap.Boolean.decode;
const encode_DateTime = _defaultTypeMap.DateTime.encode;
const decode_DateTime = _defaultTypeMap.DateTime.decode;
const encode_UInt32 = _defaultTypeMap.UInt32.encode;
const decode_UInt32 = _defaultTypeMap.UInt32.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ReadRawModifiedDetails.prototype.encode = function(stream,options) {
    // call base class implementation first
    HistoryReadDetails.prototype.encode.call(this,stream,options);
    encode_Boolean(this.isReadModified,stream);
    encode_DateTime(this.startTime,stream);
    encode_DateTime(this.endTime,stream);
    encode_UInt32(this.numValuesPerNode,stream);
    encode_Boolean(this.returnBounds,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
ReadRawModifiedDetails.prototype.decode = function(stream) {
    // call base class implementation first
    HistoryReadDetails.prototype.decode.call(this,stream);
    this.isReadModified = decode_Boolean(stream);
    this.startTime = decode_DateTime(stream);
    this.endTime = decode_DateTime(stream);
    this.numValuesPerNode = decode_UInt32(stream);
    this.returnBounds = decode_Boolean(stream);
};
ReadRawModifiedDetails.possibleFields = [
  "isReadModified",
         "startTime",
         "endTime",
         "numValuesPerNode",
         "returnBounds"
];


exports.ReadRawModifiedDetails = ReadRawModifiedDetails;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ReadRawModifiedDetails",ReadRawModifiedDetails);
