// node-opcua-service-history/_generated_/_auto_generated_ModificationInfo.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:16.690Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/ModificationInfo_schema").ModificationInfo_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ModificationInfo
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function ModificationInfo(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * The name of the user that made the modification. Support for this field is optional. A null shall be returned if it is not defined.
      * @property username
      * @type {String}
      */
    self.username = initialize_field(schema.fields[0], options.username);

    /**
      *  The time the modification was made. Support for this field is optional. A null shall be returned if it is not defined.
      * @property modificationTime
      * @type {UtcTime}
      */
    self.modificationTime = initialize_field(schema.fields[1], options.modificationTime);

    /**
      *  The modification type for the item.
      * @property updateType
      * @type {HistoryUpdateType}
      */
    self.setUpdateType(initialize_field(schema.fields[2], options.updateType));

   // Object.preventExtensions(self);
}
util.inherits(ModificationInfo,BaseUAObject);

//## Define Enumeration setters
ModificationInfo.prototype.setUpdateType = function(value) {
   const coercedValue = _enumerations.HistoryUpdateType.typedEnum.get(value);
   /* istanbul ignore next */
   if (coercedValue === undefined || coercedValue === null) {
      throw new Error("value cannot be coerced to HistoryUpdateType: " + value);
   }
   this.updateType = coercedValue;
};
ModificationInfo.prototype.encodingDefaultBinary = makeExpandedNodeId(11226,0);
ModificationInfo.prototype.encodingDefaultXml = makeExpandedNodeId(11218,0);
ModificationInfo.prototype._schema = schema;

const encode_String = _defaultTypeMap.String.encode;
const decode_String = _defaultTypeMap.String.decode;
const encode_UtcTime = _defaultTypeMap.UtcTime.encode;
const decode_UtcTime = _defaultTypeMap.UtcTime.decode;
const encode_HistoryUpdateType = _enumerations.HistoryUpdateType.encode;
const decode_HistoryUpdateType = _enumerations.HistoryUpdateType.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ModificationInfo.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_String(this.username,stream);
    encode_UtcTime(this.modificationTime,stream);
    encode_HistoryUpdateType(this.updateType,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
ModificationInfo.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.username = decode_String(stream);
    this.modificationTime = decode_UtcTime(stream);
    this.updateType = decode_HistoryUpdateType(stream);
};
ModificationInfo.possibleFields = [
  "username",
         "modificationTime",
         "updateType"
];


exports.ModificationInfo = ModificationInfo;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ModificationInfo",ModificationInfo);
