// node-opcua-service-history/_generated_/_auto_generated_ReadProcessedDetails.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:16.714Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/ReadProcessedDetails_schema").ReadProcessedDetails_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const AggregateConfiguration = require("./_auto_generated_AggregateConfiguration").AggregateConfiguration;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;
const HistoryReadDetails = require("./_auto_generated_HistoryReadDetails").HistoryReadDetails;

/**
 * 
 * @class ReadProcessedDetails
 * @constructor
 * @extends HistoryReadDetails
 * @param  options {Object}
 */
function ReadProcessedDetails(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    HistoryReadDetails.call(this,options);
    if (options === null) { 
        HistoryReadDetails.call(this,options);
        self.aggregateConfiguration =  null; /* new AggregateConfiguration(null); */
        return ;
    }

    /**
      * 
      * @property startTime
      * @type {DateTime}
      */
    self.startTime = initialize_field(schema.fields[0], options.startTime);

    /**
      * 
      * @property endTime
      * @type {DateTime}
      */
    self.endTime = initialize_field(schema.fields[1], options.endTime);

    /**
      * 
      * @property processingInterval
      * @type {DateTime}
      */
    self.processingInterval = initialize_field(schema.fields[2], options.processingInterval);

    /**
      * 
      * @property aggregateType
      * @type {NodeId[]}
      */
    self.aggregateType = initialize_field_array(schema.fields[3], options.aggregateType);

    /**
      * 
      * @property aggregateConfiguration
      * @type {AggregateConfiguration}
      */
    self.aggregateConfiguration =  new AggregateConfiguration( options.aggregateConfiguration);

   // Object.preventExtensions(self);
}
util.inherits(ReadProcessedDetails,HistoryReadDetails);
ReadProcessedDetails.prototype.encodingDefaultBinary = makeExpandedNodeId(652,0);
ReadProcessedDetails.prototype.encodingDefaultXml = makeExpandedNodeId(651,0);
ReadProcessedDetails.prototype._schema = schema;

const encode_DateTime = _defaultTypeMap.DateTime.encode;
const decode_DateTime = _defaultTypeMap.DateTime.decode;
const encode_NodeId = _defaultTypeMap.NodeId.encode;
const decode_NodeId = _defaultTypeMap.NodeId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ReadProcessedDetails.prototype.encode = function(stream,options) {
    // call base class implementation first
    HistoryReadDetails.prototype.encode.call(this,stream,options);
    encode_DateTime(this.startTime,stream);
    encode_DateTime(this.endTime,stream);
    encode_DateTime(this.processingInterval,stream);
    encodeArray(this.aggregateType, stream, encode_NodeId);
   this.aggregateConfiguration.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
ReadProcessedDetails.prototype.decode = function(stream) {
    // call base class implementation first
    HistoryReadDetails.prototype.decode.call(this,stream);
    this.startTime = decode_DateTime(stream);
    this.endTime = decode_DateTime(stream);
    this.processingInterval = decode_DateTime(stream);
    this.aggregateType = decodeArray(stream, decode_NodeId);
    this.aggregateConfiguration.decode(stream);
};
ReadProcessedDetails.possibleFields = [
  "startTime",
         "endTime",
         "processingInterval",
         "aggregateType",
         "aggregateConfiguration"
];


exports.ReadProcessedDetails = ReadProcessedDetails;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ReadProcessedDetails",ReadProcessedDetails);
