// node-opcua-data-value/_generated_/_auto_generated_DataValue.js
// --------- This code has been automatically generated !!! 2018-03-24T16:23:36.609Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/DataValue_schema").DataValue_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var Variant = getFactory("Variant");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class DataValue
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.value = null] {Variant} 
 * @param  [options.statusCode = 0] {StatusCode} 
 * @param  [options.sourceTimestamp = null] {DateTime} 
 * @param  [options.sourcePicoseconds = 0] {UInt16} 
 * @param  [options.serverTimestamp = null] {DateTime} 
 * @param  [options.serverPicoseconds = 0] {UInt16} 
 */
function DataValue(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.value =  null; /* new Variant(null); */
        return ;
    }

    /**
      * 
      * @property value
      * @type {Variant}
      * @default  null
      */
    self.value = (options.value) ? new Variant( options.value) : null;

    /**
      * 
      * @property statusCode
      * @type {StatusCode}
      * @default  0
      */
    self.statusCode = initialize_field(schema.fields[1], options.statusCode);

    /**
      * 
      * @property sourceTimestamp
      * @type {DateTime}
      * @default  null
      */
    self.sourceTimestamp = initialize_field(schema.fields[2], options.sourceTimestamp);

    /**
      * 
      * @property sourcePicoseconds
      * @type {UInt16}
      * @default  0
      */
    self.sourcePicoseconds = initialize_field(schema.fields[3], options.sourcePicoseconds);

    /**
      * 
      * @property serverTimestamp
      * @type {DateTime}
      * @default  null
      */
    self.serverTimestamp = initialize_field(schema.fields[4], options.serverTimestamp);

    /**
      * 
      * @property serverPicoseconds
      * @type {UInt16}
      * @default  0
      */
    self.serverPicoseconds = initialize_field(schema.fields[5], options.serverPicoseconds);

   // Object.preventExtensions(self);
}
util.inherits(DataValue,BaseUAObject);
schema.id = generate_new_id();
DataValue.prototype.encodingDefaultBinary = makeExpandedNodeId(schema.id);
DataValue.prototype._schema = schema;

var encode_StatusCode = _defaultTypeMap.StatusCode.encode;
var decode_StatusCode = _defaultTypeMap.StatusCode.decode;
var encode_DateTime = _defaultTypeMap.DateTime.encode;
var decode_DateTime = _defaultTypeMap.DateTime.decode;
var encode_UInt16 = _defaultTypeMap.UInt16.encode;
var decode_UInt16 = _defaultTypeMap.UInt16.decode;
DataValue.prototype.encode = function(stream,options) {
   schema.encode(this,stream,options); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
DataValue.prototype.decode = function(stream,options) {
   schema.decode(this,stream,options); 
};
DataValue.prototype.decode_debug = function(stream,options) {
   schema.decode_debug(this,stream,options); 
};
/**
 *
 * verify that all object attributes values are valid according to schema
 * @method isValid
 * @return {Boolean}
 */
DataValue.prototype.isValid = function() { return schema.isValid(this); };
DataValue.possibleFields = [
  "value",
         "statusCode",
         "sourceTimestamp",
         "sourcePicoseconds",
         "serverTimestamp",
         "serverPicoseconds"
];


exports.DataValue = DataValue;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("DataValue",DataValue);
