// node-opcua-common/_generated_/_auto_generated_ServerStatus.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:13.416Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/ServerStatus_schema").ServerStatus_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BuildInfo = require("./_auto_generated_BuildInfo").BuildInfo;
const LocalizedText = getFactory("LocalizedText");
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ServerStatus
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function ServerStatus(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.buildInfo =  null; /* new BuildInfo(null); */
        self.shutdownReason =  null; /* new LocalizedText(null); */
        return ;
    }

    /**
      * 
      * @property startTime
      * @type {UtcTime}
      */
    self.startTime = initialize_field(schema.fields[0], options.startTime);

    /**
      * 
      * @property currentTime
      * @type {UtcTime}
      */
    self.currentTime = initialize_field(schema.fields[1], options.currentTime);

    /**
      * 
      * @property state
      * @type {ServerState}
      */
    self.setState(initialize_field(schema.fields[2], options.state));

    /**
      * 
      * @property buildInfo
      * @type {BuildInfo}
      */
    self.buildInfo =  new BuildInfo( options.buildInfo);

    /**
      * 
      * @property secondsTillShutdown
      * @type {UInt32}
      */
    self.secondsTillShutdown = initialize_field(schema.fields[4], options.secondsTillShutdown);

    /**
      * 
      * @property shutdownReason
      * @type {LocalizedText}
      */
    self.shutdownReason =  new LocalizedText( options.shutdownReason);

   // Object.preventExtensions(self);
}
util.inherits(ServerStatus,BaseUAObject);

//## Define Enumeration setters
ServerStatus.prototype.setState = function(value) {
   const coercedValue = _enumerations.ServerState.typedEnum.get(value);
   /* istanbul ignore next */
   if (coercedValue === undefined || coercedValue === null) {
      throw new Error("value cannot be coerced to ServerState: " + value);
   }
   this.state = coercedValue;
};
ServerStatus.prototype.encodingDefaultBinary = makeExpandedNodeId(864,0);
ServerStatus.prototype._schema = schema;

const encode_UtcTime = _defaultTypeMap.UtcTime.encode;
const decode_UtcTime = _defaultTypeMap.UtcTime.decode;
const encode_ServerState = _enumerations.ServerState.encode;
const decode_ServerState = _enumerations.ServerState.decode;
const encode_UInt32 = _defaultTypeMap.UInt32.encode;
const decode_UInt32 = _defaultTypeMap.UInt32.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ServerStatus.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_UtcTime(this.startTime,stream);
    encode_UtcTime(this.currentTime,stream);
    encode_ServerState(this.state,stream);
   this.buildInfo.encode(stream,options);
    encode_UInt32(this.secondsTillShutdown,stream);
   this.shutdownReason.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
ServerStatus.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.startTime = decode_UtcTime(stream);
    this.currentTime = decode_UtcTime(stream);
    this.state = decode_ServerState(stream);
    this.buildInfo.decode(stream);
    this.secondsTillShutdown = decode_UInt32(stream);
    this.shutdownReason.decode(stream);
};
ServerStatus.possibleFields = [
  "startTime",
         "currentTime",
         "state",
         "buildInfo",
         "secondsTillShutdown",
         "shutdownReason"
];


exports.ServerStatus = ServerStatus;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ServerStatus",ServerStatus);
