// node-opcua-common/_generated_/_auto_generated_ApplicationInstanceCertificate.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:13.341Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/ApplicationInstanceCertificate_schema").ApplicationInstanceCertificate_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ApplicationInstanceCertificate
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function ApplicationInstanceCertificate(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property version
      * @type {String}
      */
    self.version = initialize_field(schema.fields[0], options.version);

    /**
      * 
      * @property serialNumber
      * @type {ByteString}
      */
    self.serialNumber = initialize_field(schema.fields[1], options.serialNumber);

    /**
      * 
      * @property signatureAlgorithm
      * @type {String}
      */
    self.signatureAlgorithm = initialize_field(schema.fields[2], options.signatureAlgorithm);

    /**
      * 
      * @property signature
      * @type {ByteString}
      */
    self.signature = initialize_field(schema.fields[3], options.signature);

    /**
      * 
      * @property issuer
      * @type {String}
      */
    self.issuer = initialize_field(schema.fields[4], options.issuer);

    /**
      * 
      * @property validFrom
      * @type {UtcTime}
      */
    self.validFrom = initialize_field(schema.fields[5], options.validFrom);

    /**
      * 
      * @property validTo
      * @type {UtcTime}
      */
    self.validTo = initialize_field(schema.fields[6], options.validTo);

    /**
      * 
      * @property subject
      * @type {String}
      */
    self.subject = initialize_field(schema.fields[7], options.subject);

    /**
      * 
      * @property applicationUri
      * @type {String}
      */
    self.applicationUri = initialize_field(schema.fields[8], options.applicationUri);

    /**
      * 
      * @property hostnames
      * @type {String[]}
      */
    self.hostnames = initialize_field_array(schema.fields[9], options.hostnames);

    /**
      * 
      * @property publicKey
      * @type {ByteString}
      */
    self.publicKey = initialize_field(schema.fields[10], options.publicKey);

    /**
      * 
      * @property keyUsage
      * @type {String}
      */
    self.keyUsage = initialize_field(schema.fields[11], options.keyUsage);

   // Object.preventExtensions(self);
}
util.inherits(ApplicationInstanceCertificate,BaseUAObject);
schema.id = generate_new_id();
ApplicationInstanceCertificate.prototype.encodingDefaultBinary = makeExpandedNodeId(schema.id);
ApplicationInstanceCertificate.prototype._schema = schema;

const encode_String = _defaultTypeMap.String.encode;
const decode_String = _defaultTypeMap.String.decode;
const encode_ByteString = _defaultTypeMap.ByteString.encode;
const decode_ByteString = _defaultTypeMap.ByteString.decode;
const encode_UtcTime = _defaultTypeMap.UtcTime.encode;
const decode_UtcTime = _defaultTypeMap.UtcTime.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ApplicationInstanceCertificate.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_String(this.version,stream);
    encode_ByteString(this.serialNumber,stream);
    encode_String(this.signatureAlgorithm,stream);
    encode_ByteString(this.signature,stream);
    encode_String(this.issuer,stream);
    encode_UtcTime(this.validFrom,stream);
    encode_UtcTime(this.validTo,stream);
    encode_String(this.subject,stream);
    encode_String(this.applicationUri,stream);
    encodeArray(this.hostnames, stream, encode_String);
    encode_ByteString(this.publicKey,stream);
    encode_String(this.keyUsage,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
ApplicationInstanceCertificate.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.version = decode_String(stream);
    this.serialNumber = decode_ByteString(stream);
    this.signatureAlgorithm = decode_String(stream);
    this.signature = decode_ByteString(stream);
    this.issuer = decode_String(stream);
    this.validFrom = decode_UtcTime(stream);
    this.validTo = decode_UtcTime(stream);
    this.subject = decode_String(stream);
    this.applicationUri = decode_String(stream);
    this.hostnames = decodeArray(stream, decode_String);
    this.publicKey = decode_ByteString(stream);
    this.keyUsage = decode_String(stream);
};
ApplicationInstanceCertificate.possibleFields = [
  "version",
         "serialNumber",
         "signatureAlgorithm",
         "signature",
         "issuer",
         "validFrom",
         "validTo",
         "subject",
         "applicationUri",
         "hostnames",
         "publicKey",
         "keyUsage"
];


exports.ApplicationInstanceCertificate = ApplicationInstanceCertificate;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ApplicationInstanceCertificate",ApplicationInstanceCertificate);
