// node-opcua-common/_generated_/_auto_generated_RedundantServer.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:13.421Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/RedundantServer_schema").RedundantServer_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class RedundantServer
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function RedundantServer(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property serverId
      * @type {String}
      */
    self.serverId = initialize_field(schema.fields[0], options.serverId);

    /**
      * 
      * @property serviceLevel
      * @type {Byte}
      */
    self.serviceLevel = initialize_field(schema.fields[1], options.serviceLevel);

    /**
      * 
      * @property serverState
      * @type {ServerState}
      */
    self.setServerState(initialize_field(schema.fields[2], options.serverState));

   // Object.preventExtensions(self);
}
util.inherits(RedundantServer,BaseUAObject);

//## Define Enumeration setters
RedundantServer.prototype.setServerState = function(value) {
   const coercedValue = _enumerations.ServerState.typedEnum.get(value);
   /* istanbul ignore next */
   if (coercedValue === undefined || coercedValue === null) {
      throw new Error("value cannot be coerced to ServerState: " + value);
   }
   this.serverState = coercedValue;
};
RedundantServer.prototype.encodingDefaultBinary = makeExpandedNodeId(855,0);
RedundantServer.prototype._schema = schema;

const encode_String = _defaultTypeMap.String.encode;
const decode_String = _defaultTypeMap.String.decode;
const encode_Byte = _defaultTypeMap.Byte.encode;
const decode_Byte = _defaultTypeMap.Byte.decode;
const encode_ServerState = _enumerations.ServerState.encode;
const decode_ServerState = _enumerations.ServerState.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
RedundantServer.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_String(this.serverId,stream);
    encode_Byte(this.serviceLevel,stream);
    encode_ServerState(this.serverState,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
RedundantServer.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.serverId = decode_String(stream);
    this.serviceLevel = decode_Byte(stream);
    this.serverState = decode_ServerState(stream);
};
RedundantServer.possibleFields = [
  "serverId",
         "serviceLevel",
         "serverState"
];


exports.RedundantServer = RedundantServer;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("RedundantServer",RedundantServer);
