// node-opcua-common/_generated_/_auto_generated_SamplingIntervalDiagnostics.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:13.427Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/SamplingIntervalDiagnostics_schema").SamplingIntervalDiagnostics_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class SamplingIntervalDiagnostics
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function SamplingIntervalDiagnostics(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property samplingInterval
      * @type {Duration}
      */
    self.samplingInterval = initialize_field(schema.fields[0], options.samplingInterval);

    /**
      * 
      * @property monitoredItemCount
      * @type {UInt32}
      */
    self.monitoredItemCount = initialize_field(schema.fields[1], options.monitoredItemCount);

    /**
      * 
      * @property maxMonitoredItemCount
      * @type {UInt32}
      */
    self.maxMonitoredItemCount = initialize_field(schema.fields[2], options.maxMonitoredItemCount);

    /**
      * 
      * @property disabledMonitoredItemCount
      * @type {UInt32}
      */
    self.disabledMonitoredItemCount = initialize_field(schema.fields[3], options.disabledMonitoredItemCount);

   // Object.preventExtensions(self);
}
util.inherits(SamplingIntervalDiagnostics,BaseUAObject);
SamplingIntervalDiagnostics.prototype.encodingDefaultBinary = makeExpandedNodeId(858,0);
SamplingIntervalDiagnostics.prototype._schema = schema;

const encode_Duration = _defaultTypeMap.Duration.encode;
const decode_Duration = _defaultTypeMap.Duration.decode;
const encode_UInt32 = _defaultTypeMap.UInt32.encode;
const decode_UInt32 = _defaultTypeMap.UInt32.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
SamplingIntervalDiagnostics.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_Duration(this.samplingInterval,stream);
    encode_UInt32(this.monitoredItemCount,stream);
    encode_UInt32(this.maxMonitoredItemCount,stream);
    encode_UInt32(this.disabledMonitoredItemCount,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
SamplingIntervalDiagnostics.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.samplingInterval = decode_Duration(stream);
    this.monitoredItemCount = decode_UInt32(stream);
    this.maxMonitoredItemCount = decode_UInt32(stream);
    this.disabledMonitoredItemCount = decode_UInt32(stream);
};
SamplingIntervalDiagnostics.possibleFields = [
  "samplingInterval",
         "monitoredItemCount",
         "maxMonitoredItemCount",
         "disabledMonitoredItemCount"
];


exports.SamplingIntervalDiagnostics = SamplingIntervalDiagnostics;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("SamplingIntervalDiagnostics",SamplingIntervalDiagnostics);
