// node-opcua-service-register-node/_generated_/_auto_generated_UnregisterNodesRequest.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:11.761Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/UnregisterNodesRequest_schema").UnregisterNodesRequest_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const RequestHeader = getFactory("RequestHeader");
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * Registers Nodes
 * 
 * @class UnregisterNodesRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function UnregisterNodesRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestHeader =  null; /* new RequestHeader(null); */
        return ;
    }

    /**
      * A standard header included in all requests sent to a server.
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * A list of NodeIds that have been obtained via the RegisterNodes service.
      * @property nodesToUnregister
      * @type {NodeId[]}
      */
    self.nodesToUnregister = initialize_field_array(schema.fields[1], options.nodesToUnregister);

   // Object.preventExtensions(self);
}
util.inherits(UnregisterNodesRequest,BaseUAObject);
UnregisterNodesRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(566,0);
UnregisterNodesRequest.prototype.encodingDefaultXml = makeExpandedNodeId(565,0);
UnregisterNodesRequest.prototype._schema = schema;

const encode_NodeId = _defaultTypeMap.NodeId.encode;
const decode_NodeId = _defaultTypeMap.NodeId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
UnregisterNodesRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encodeArray(this.nodesToUnregister, stream, encode_NodeId);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
UnregisterNodesRequest.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.requestHeader.decode(stream);
    this.nodesToUnregister = decodeArray(stream, decode_NodeId);
};
UnregisterNodesRequest.possibleFields = [
  "requestHeader",
         "nodesToUnregister"
];


exports.UnregisterNodesRequest = UnregisterNodesRequest;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("UnregisterNodesRequest",UnregisterNodesRequest);
