// node-opcua-service-secure-channel/_generated_/_auto_generated_ChannelSecurityToken.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:07.951Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/ChannelSecurityToken_schema").ChannelSecurityToken_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ChannelSecurityToken
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function ChannelSecurityToken(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property secureChannelId
      * @type {UInt32}
      */
    self.secureChannelId = initialize_field(schema.fields[0], options.secureChannelId);

    /**
      * 
      * @property tokenId
      * @type {UInt32}
      */
    self.tokenId = initialize_field(schema.fields[1], options.tokenId);

    /**
      * 
      * @property createdAt
      * @type {UtcTime}
      * @default  function () {
            return new Date();
        }
      */
    self.createdAt = initialize_field(schema.fields[2], options.createdAt);

    /**
      * 
      * @property revisedLifeTime
      * @type {UInt32}
      * @default  30000
      */
    self.revisedLifeTime = initialize_field(schema.fields[3], options.revisedLifeTime);

   // Object.preventExtensions(self);
}
util.inherits(ChannelSecurityToken,BaseUAObject);
ChannelSecurityToken.prototype.encodingDefaultBinary = makeExpandedNodeId(443,0);
ChannelSecurityToken.prototype.encodingDefaultXml = makeExpandedNodeId(442,0);
ChannelSecurityToken.prototype._schema = schema;

const encode_UInt32 = _defaultTypeMap.UInt32.encode;
const decode_UInt32 = _defaultTypeMap.UInt32.decode;
const encode_UtcTime = _defaultTypeMap.UtcTime.encode;
const decode_UtcTime = _defaultTypeMap.UtcTime.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ChannelSecurityToken.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_UInt32(this.secureChannelId,stream);
    encode_UInt32(this.tokenId,stream);
    encode_UtcTime(this.createdAt,stream);
    encode_UInt32(this.revisedLifeTime,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
ChannelSecurityToken.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.secureChannelId = decode_UInt32(stream);
    this.tokenId = decode_UInt32(stream);
    this.createdAt = decode_UtcTime(stream);
    this.revisedLifeTime = decode_UInt32(stream);
};
ChannelSecurityToken.possibleFields = [
  "secureChannelId",
         "tokenId",
         "createdAt",
         "revisedLifeTime"
];


exports.ChannelSecurityToken = ChannelSecurityToken;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ChannelSecurityToken",ChannelSecurityToken);
