// node-opcua-service-secure-channel/_generated_/_auto_generated_RequestHeader.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:07.894Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/RequestHeader_schema").RequestHeader_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class RequestHeader
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function RequestHeader(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property authenticationToken
      * @type {NodeId}
      */
    self.authenticationToken = initialize_field(schema.fields[0], options.authenticationToken);

    /**
      * 
      * @property timeStamp
      * @type {UtcTime}
      * @default  function () {
            return new Date();
        }
      */
    self.timeStamp = initialize_field(schema.fields[1], options.timeStamp);

    /**
      * 
      * @property requestHandle
      * @type {IntegerId}
      * @default  function () {
            return 0xDEADBEEF;
        }
      */
    self.requestHandle = initialize_field(schema.fields[2], options.requestHandle);

    /**
      * 
      * @property returnDiagnostics
      * @type {UInt32}
      */
    self.returnDiagnostics = initialize_field(schema.fields[3], options.returnDiagnostics);

    /**
      * 
      * @property auditEntryId
      * @type {UAString}
      */
    self.auditEntryId = initialize_field(schema.fields[4], options.auditEntryId);

    /**
      * 
      * @property timeoutHint
      * @type {UInt32}
      */
    self.timeoutHint = initialize_field(schema.fields[5], options.timeoutHint);

    /**
      * 
      * @property additionalHeader
      * @type {ExtensionObject}
      */
    self.additionalHeader = initialize_field(schema.fields[6], options.additionalHeader);

   // Object.preventExtensions(self);
}
util.inherits(RequestHeader,BaseUAObject);
RequestHeader.prototype.encodingDefaultBinary = makeExpandedNodeId(391,0);
RequestHeader.prototype.encodingDefaultXml = makeExpandedNodeId(390,0);
RequestHeader.prototype._schema = schema;

const encode_NodeId = _defaultTypeMap.NodeId.encode;
const decode_NodeId = _defaultTypeMap.NodeId.decode;
const encode_UtcTime = _defaultTypeMap.UtcTime.encode;
const decode_UtcTime = _defaultTypeMap.UtcTime.decode;
const encode_IntegerId = _defaultTypeMap.IntegerId.encode;
const decode_IntegerId = _defaultTypeMap.IntegerId.decode;
const encode_UInt32 = _defaultTypeMap.UInt32.encode;
const decode_UInt32 = _defaultTypeMap.UInt32.decode;
const encode_UAString = _defaultTypeMap.UAString.encode;
const decode_UAString = _defaultTypeMap.UAString.decode;
const encode_ExtensionObject = _defaultTypeMap.ExtensionObject.encode;
const decode_ExtensionObject = _defaultTypeMap.ExtensionObject.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
RequestHeader.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_NodeId(this.authenticationToken,stream);
    encode_UtcTime(this.timeStamp,stream);
    encode_IntegerId(this.requestHandle,stream);
    encode_UInt32(this.returnDiagnostics,stream);
    encode_UAString(this.auditEntryId,stream);
    encode_UInt32(this.timeoutHint,stream);
    encode_ExtensionObject(this.additionalHeader,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
RequestHeader.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.authenticationToken = decode_NodeId(stream);
    this.timeStamp = decode_UtcTime(stream);
    this.requestHandle = decode_IntegerId(stream);
    this.returnDiagnostics = decode_UInt32(stream);
    this.auditEntryId = decode_UAString(stream);
    this.timeoutHint = decode_UInt32(stream);
    this.additionalHeader = decode_ExtensionObject(stream);
};
RequestHeader.possibleFields = [
  "authenticationToken",
         "timeStamp",
         "requestHandle",
         "returnDiagnostics",
         "auditEntryId",
         "timeoutHint",
         "additionalHeader"
];


exports.RequestHeader = RequestHeader;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("RequestHeader",RequestHeader);
