// node-opcua-service-secure-channel/_generated_/_auto_generated_SignatureData.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:07.955Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/SignatureData_schema").SignatureData_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class SignatureData
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function SignatureData(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * The cryptography algorithm used to create the signature.
      * @property algorithm
      * @type {String}
      * @default  null
      */
    self.algorithm = initialize_field(schema.fields[0], options.algorithm);

    /**
      * The digital signature.
      * @property signature
      * @type {ByteString}
      * @default  null
      */
    self.signature = initialize_field(schema.fields[1], options.signature);

   // Object.preventExtensions(self);
}
util.inherits(SignatureData,BaseUAObject);
SignatureData.prototype.encodingDefaultBinary = makeExpandedNodeId(458,0);
SignatureData.prototype.encodingDefaultXml = makeExpandedNodeId(457,0);
SignatureData.prototype._schema = schema;

const encode_String = _defaultTypeMap.String.encode;
const decode_String = _defaultTypeMap.String.decode;
const encode_ByteString = _defaultTypeMap.ByteString.encode;
const decode_ByteString = _defaultTypeMap.ByteString.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
SignatureData.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_String(this.algorithm,stream);
    encode_ByteString(this.signature,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
SignatureData.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.algorithm = decode_String(stream);
    this.signature = decode_ByteString(stream);
};
SignatureData.possibleFields = [
  "algorithm",
         "signature"
];


exports.SignatureData = SignatureData;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("SignatureData",SignatureData);
