// node-opcua-service-secure-channel/_generated_/_auto_generated_ErrorMessage.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:07.890Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/ErrorMessage_schema").ErrorMessage_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ErrorMessage
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function ErrorMessage(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * The numeric code for the error. This shall be one of the values listed in Table 40.
      * @property Error
      * @type {UInt32}
      */
    self.Error = initialize_field(schema.fields[0], options.Error);

    /**
      * A more verbose description of the error.This string shall not be more than 4096 characters.
      * @property Reason
      * @type {String}
      */
    self.Reason = initialize_field(schema.fields[1], options.Reason);

   // Object.preventExtensions(self);
}
util.inherits(ErrorMessage,BaseUAObject);
schema.id = generate_new_id();
ErrorMessage.prototype.encodingDefaultBinary = makeExpandedNodeId(schema.id);
ErrorMessage.prototype._schema = schema;

const encode_UInt32 = _defaultTypeMap.UInt32.encode;
const decode_UInt32 = _defaultTypeMap.UInt32.decode;
const encode_String = _defaultTypeMap.String.encode;
const decode_String = _defaultTypeMap.String.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ErrorMessage.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_UInt32(this.Error,stream);
    encode_String(this.Reason,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
ErrorMessage.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.Error = decode_UInt32(stream);
    this.Reason = decode_String(stream);
};
ErrorMessage.possibleFields = [
  "Error",
         "Reason"
];


exports.ErrorMessage = ErrorMessage;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ErrorMessage",ErrorMessage);
