// node-opcua-data-model/_generated_/_auto_generated_EnumValueType.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:06.431Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/EnumValueType_schema").EnumValueType_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const LocalizedText = require("./_auto_generated_LocalizedText").LocalizedText;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class EnumValueType
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function EnumValueType(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.displayName =  null; /* new LocalizedText(null); */
        self.description =  null; /* new LocalizedText(null); */
        return ;
    }

    /**
      * The Integer representation of an Enumeration.
      * @property value
      * @type {Int64}
      */
    self.value = initialize_field(schema.fields[0], options.value);

    /**
      * A human-readable representation of the Value of the Enumeration.
      * @property displayName
      * @type {LocalizedText}
      */
    self.displayName =  new LocalizedText( options.displayName);

    /**
      * A localized description of the enumeration value. This field can contain an empty string if no description is available
      * @property description
      * @type {LocalizedText}
      */
    self.description =  new LocalizedText( options.description);

   // Object.preventExtensions(self);
}
util.inherits(EnumValueType,BaseUAObject);
EnumValueType.prototype.encodingDefaultBinary = makeExpandedNodeId(8251,0);
EnumValueType.prototype.encodingDefaultXml = makeExpandedNodeId(7616,0);
EnumValueType.prototype._schema = schema;

const encode_Int64 = _defaultTypeMap.Int64.encode;
const decode_Int64 = _defaultTypeMap.Int64.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
EnumValueType.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_Int64(this.value,stream);
   this.displayName.encode(stream,options);
   this.description.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
EnumValueType.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.value = decode_Int64(stream);
    this.displayName.decode(stream);
    this.description.decode(stream);
};
EnumValueType.possibleFields = [
  "value",
         "displayName",
         "description"
];


exports.EnumValueType = EnumValueType;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("EnumValueType",EnumValueType);
