// node-opcua-data-model/_generated_/_auto_generated_QualifiedName.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:06.411Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/QualifiedName_schema").QualifiedName_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class QualifiedName
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function QualifiedName(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * The namespace index
      * @property namespaceIndex
      * @type {UInt16}
      */
    self.namespaceIndex = initialize_field(schema.fields[0], options.namespaceIndex);

    /**
      * The name
      * @property name
      * @type {String}
      * @default  function () {
            return null;
        }
      */
    self.name = initialize_field(schema.fields[1], options.name);

   // Object.preventExtensions(self);
}
util.inherits(QualifiedName,BaseUAObject);
schema.id = generate_new_id();
QualifiedName.prototype.encodingDefaultBinary = makeExpandedNodeId(schema.id);
QualifiedName.prototype._schema = schema;

const encode_UInt16 = _defaultTypeMap.UInt16.encode;
const decode_UInt16 = _defaultTypeMap.UInt16.decode;
const encode_String = _defaultTypeMap.String.encode;
const decode_String = _defaultTypeMap.String.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
QualifiedName.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_UInt16(this.namespaceIndex,stream);
    encode_String(this.name,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
QualifiedName.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.namespaceIndex = decode_UInt16(stream);
    this.name = decode_String(stream);
};
QualifiedName.possibleFields = [
  "namespaceIndex",
         "name"
];


exports.QualifiedName = QualifiedName;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("QualifiedName",QualifiedName);
