// node-opcua-data-model/_generated_/_auto_generated_EnumValueType.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:12.214Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/EnumValueType_schema").EnumValueType_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var LocalizedText = require("./_auto_generated_LocalizedText").LocalizedText;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class EnumValueType
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.value] {Int64} The Integer representation of an Enumeration.
 * @param  [options.displayName] {LocalizedText} A human-readable representation of the Value of the Enumeration.
 * @param  [options.description] {LocalizedText} A localized description of the enumeration value. This field can contain an empty string if no description is available
 */
function EnumValueType(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.displayName =  null; /* new LocalizedText(null); */
        self.description =  null; /* new LocalizedText(null); */
        return ;
    }

    /**
      * The Integer representation of an Enumeration.
      * @property value
      * @type {Int64}
      */
    self.value = initialize_field(schema.fields[0], options.value);

    /**
      * A human-readable representation of the Value of the Enumeration.
      * @property displayName
      * @type {LocalizedText}
      */
    self.displayName =  new LocalizedText( options.displayName);

    /**
      * A localized description of the enumeration value. This field can contain an empty string if no description is available
      * @property description
      * @type {LocalizedText}
      */
    self.description =  new LocalizedText( options.description);

   // Object.preventExtensions(self);
}
util.inherits(EnumValueType,BaseUAObject);
EnumValueType.prototype.encodingDefaultBinary = makeExpandedNodeId(8251,0);
EnumValueType.prototype.encodingDefaultXml = makeExpandedNodeId(7616,0);
EnumValueType.prototype._schema = schema;

var encode_Int64 = _defaultTypeMap.Int64.encode;
var decode_Int64 = _defaultTypeMap.Int64.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
EnumValueType.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_Int64(this.value,stream);
   this.displayName.encode(stream,options);
   this.description.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
EnumValueType.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.value = decode_Int64(stream,options);
    this.displayName.decode(stream,options);
    this.description.decode(stream,options);
};
EnumValueType.possibleFields = [
  "value",
         "displayName",
         "description"
];


exports.EnumValueType = EnumValueType;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("EnumValueType",EnumValueType);
