// node-opcua-service-filter/_generated_/_auto_generated_ContentFilterElement.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:25.457Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ContentFilterElement_schema").ContentFilterElement_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ContentFilterElement
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.filterOperator] {FilterOperator} 
 * @param  [options.filterOperands] {ExtensionObject[]} 
 */
function ContentFilterElement(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property filterOperator
      * @type {FilterOperator}
      */
    self.setFilterOperator(initialize_field(schema.fields[0], options.filterOperator));

    /**
      * 
      * @property filterOperands
      * @type {ExtensionObject[]}
      */
    self.filterOperands = initialize_field_array(schema.fields[1], options.filterOperands);

   // Object.preventExtensions(self);
}
util.inherits(ContentFilterElement,BaseUAObject);

//## Define Enumeration setters
ContentFilterElement.prototype.setFilterOperator = function(value) {
   var coercedValue = _enumerations.FilterOperator.typedEnum.get(value);
   /* istanbul ignore next */
   if (coercedValue === undefined || coercedValue === null) {
      throw new Error("value cannot be coerced to FilterOperator: " + value);
   }
   this.filterOperator = coercedValue;
};
ContentFilterElement.prototype.encodingDefaultBinary = makeExpandedNodeId(585,0);
ContentFilterElement.prototype.encodingDefaultXml = makeExpandedNodeId(584,0);
ContentFilterElement.prototype._schema = schema;

var encode_FilterOperator = _enumerations.FilterOperator.encode;
var decode_FilterOperator = _enumerations.FilterOperator.decode;
var encode_ExtensionObject = _defaultTypeMap.ExtensionObject.encode;
var decode_ExtensionObject = _defaultTypeMap.ExtensionObject.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ContentFilterElement.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_FilterOperator(this.filterOperator,stream);
    encodeArray(this.filterOperands, stream, encode_ExtensionObject);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
ContentFilterElement.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.filterOperator = decode_FilterOperator(stream,options);
    this.filterOperands = decodeArray(stream, decode_ExtensionObject);
};
ContentFilterElement.possibleFields = [
  "filterOperator",
         "filterOperands"
];


exports.ContentFilterElement = ContentFilterElement;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ContentFilterElement",ContentFilterElement);
