// node-opcua-service-filter/_generated_/_auto_generated_LiteralOperand.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:25.356Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/LiteralOperand_schema").LiteralOperand_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var Variant = getFactory("Variant");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;
var FilterOperand = require("./_auto_generated_FilterOperand").FilterOperand;

/**
 * 
 * @class LiteralOperand
 * @constructor
 * @extends FilterOperand
 * @param  options {Object}
 * @param  [options.value] {Variant} A literal value.
 */
function LiteralOperand(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    FilterOperand.call(this,options);
    if (options === null) { 
        FilterOperand.call(this,options);
        self.value =  null; /* new Variant(null); */
        return ;
    }

    /**
      * A literal value.
      * @property value
      * @type {Variant}
      */
    self.value =  new Variant( options.value);

   // Object.preventExtensions(self);
}
util.inherits(LiteralOperand,FilterOperand);
LiteralOperand.prototype.encodingDefaultBinary = makeExpandedNodeId(597,0);
LiteralOperand.prototype.encodingDefaultXml = makeExpandedNodeId(596,0);
LiteralOperand.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
LiteralOperand.prototype.encode = function(stream,options) {
    // call base class implementation first
    FilterOperand.prototype.encode.call(this,stream,options);
   this.value.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
LiteralOperand.prototype.decode = function(stream,options) {
    // call base class implementation first
    FilterOperand.prototype.decode.call(this,stream,options);
    this.value.decode(stream,options);
};
LiteralOperand.possibleFields = [
  "value"
];


exports.LiteralOperand = LiteralOperand;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("LiteralOperand",LiteralOperand);
