// node-opcua-service-filter/_generated_/_auto_generated_EventFilter.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:25.472Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/EventFilter_schema").EventFilter_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var SimpleAttributeOperand = require("./_auto_generated_SimpleAttributeOperand").SimpleAttributeOperand;
var ContentFilter = require("./_auto_generated_ContentFilter").ContentFilter;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;
var MonitoringFilter = require("./_auto_generated_MonitoringFilter").MonitoringFilter;

/**
 * 
 * @class EventFilter
 * @constructor
 * @extends MonitoringFilter
 * @param  options {Object}
 * @param  [options.selectClauses] {SimpleAttributeOperand[]} 
 * @param  [options.whereClause] {ContentFilter} 
 */
function EventFilter(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    MonitoringFilter.call(this,options);
    if (options === null) { 
        MonitoringFilter.call(this,options);
        self.selectClauses =  null; /* null array */
        self.whereClause =  null; /* new ContentFilter(null); */
        return ;
    }

    /**
      * 
      * @property selectClauses
      * @type {SimpleAttributeOperand[]}
      */
    self.selectClauses = [];
    if (options.selectClauses) {
        assert(_.isArray(options.selectClauses));
        self.selectClauses = options.selectClauses.map(function(e){ return new SimpleAttributeOperand(e); } );
    }

    /**
      * 
      * @property whereClause
      * @type {ContentFilter}
      */
    self.whereClause =  new ContentFilter( options.whereClause);

   // Object.preventExtensions(self);
}
util.inherits(EventFilter,MonitoringFilter);
EventFilter.prototype.encodingDefaultBinary = makeExpandedNodeId(727,0);
EventFilter.prototype.encodingDefaultXml = makeExpandedNodeId(726,0);
EventFilter.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
EventFilter.prototype.encode = function(stream,options) {
    // call base class implementation first
    MonitoringFilter.prototype.encode.call(this,stream,options);
    encodeArray(this.selectClauses,stream,function(obj,stream){ obj.encode(stream,options); }); 
   this.whereClause.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
EventFilter.prototype.decode = function(stream,options) {
    // call base class implementation first
    MonitoringFilter.prototype.decode.call(this,stream,options);
    this.selectClauses = decodeArray(stream, function(stream) { 
       var obj = new SimpleAttributeOperand(null);
       obj.decode(stream,options);
       return obj; 
    });
    this.whereClause.decode(stream,options);
};
EventFilter.possibleFields = [
  "selectClauses",
         "whereClause"
];


exports.EventFilter = EventFilter;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("EventFilter",EventFilter);
