// node-opcua-service-filter/_generated_/_auto_generated_SimpleAttributeOperand.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:25.308Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/SimpleAttributeOperand_schema").SimpleAttributeOperand_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var QualifiedName = getFactory("QualifiedName");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;
var FilterOperand = require("./_auto_generated_FilterOperand").FilterOperand;

/**
 * 
 * @class SimpleAttributeOperand
 * @constructor
 * @extends FilterOperand
 * @param  options {Object}
 * @param  [options.typeId] {NodeId} NodeId of a TypeDefinitionNode.
 * @param  [options.browsePath] {QualifiedName[]} 
 * @param  [options.attributeId] {IntegerId} 
 * @param  [options.indexRange] {NumericRange} 
 */
function SimpleAttributeOperand(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    FilterOperand.call(this,options);
    if (options === null) { 
        FilterOperand.call(this,options);
        self.browsePath =  null; /* null array */
        return ;
    }

    /**
      * NodeId of a TypeDefinitionNode.
      * @property typeId
      * @type {NodeId}
      */
    self.typeId = initialize_field(schema.fields[0], options.typeId);

    /**
      * 
      * @property browsePath
      * @type {QualifiedName[]}
      */
    self.browsePath = [];
    if (options.browsePath) {
        assert(_.isArray(options.browsePath));
        self.browsePath = options.browsePath.map(function(e){ return new QualifiedName(e); } );
    }

    /**
      * 
      * @property attributeId
      * @type {IntegerId}
      */
    self.attributeId = initialize_field(schema.fields[2], options.attributeId);

    /**
      * 
      * @property indexRange
      * @type {NumericRange}
      */
    self.indexRange = initialize_field(schema.fields[3], options.indexRange);

   // Object.preventExtensions(self);
}
util.inherits(SimpleAttributeOperand,FilterOperand);
SimpleAttributeOperand.prototype.encodingDefaultBinary = makeExpandedNodeId(603,0);
SimpleAttributeOperand.prototype.encodingDefaultXml = makeExpandedNodeId(602,0);
SimpleAttributeOperand.prototype._schema = schema;

var encode_NodeId = _defaultTypeMap.NodeId.encode;
var decode_NodeId = _defaultTypeMap.NodeId.decode;
var encode_IntegerId = _defaultTypeMap.IntegerId.encode;
var decode_IntegerId = _defaultTypeMap.IntegerId.decode;
var encode_NumericRange = _defaultTypeMap.NumericRange.encode;
var decode_NumericRange = _defaultTypeMap.NumericRange.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
SimpleAttributeOperand.prototype.encode = function(stream,options) {
    // call base class implementation first
    FilterOperand.prototype.encode.call(this,stream,options);
    encode_NodeId(this.typeId,stream);
    encodeArray(this.browsePath,stream,function(obj,stream){ obj.encode(stream,options); }); 
    encode_IntegerId(this.attributeId,stream);
    encode_NumericRange(this.indexRange,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
SimpleAttributeOperand.prototype.decode = function(stream,options) {
    // call base class implementation first
    FilterOperand.prototype.decode.call(this,stream,options);
    this.typeId = decode_NodeId(stream,options);
    this.browsePath = decodeArray(stream, function(stream) { 
       var obj = new QualifiedName(null);
       obj.decode(stream,options);
       return obj; 
    });
    this.attributeId = decode_IntegerId(stream,options);
    this.indexRange = decode_NumericRange(stream,options);
};
SimpleAttributeOperand.possibleFields = [
  "typeId",
         "browsePath",
         "attributeId",
         "indexRange"
];


exports.SimpleAttributeOperand = SimpleAttributeOperand;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("SimpleAttributeOperand",SimpleAttributeOperand);
