// node-opcua-service-translate-browse-path/_generated_/_auto_generated_RelativePathElement.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:11.737Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/RelativePathElement_schema").RelativePathElement_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const QualifiedName = getFactory("QualifiedName");
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * An element in a relative path.
 * 
 * @class RelativePathElement
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function RelativePathElement(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.targetName =  null; /* new QualifiedName(null); */
        return ;
    }

    /**
      * The type of reference to follow.
      * @property referenceTypeId
      * @type {NodeId}
      */
    self.referenceTypeId = initialize_field(schema.fields[0], options.referenceTypeId);

    /**
      * If TRUE the reverse reference is followed.
      * @property isInverse
      * @type {Boolean}
      */
    self.isInverse = initialize_field(schema.fields[1], options.isInverse);

    /**
      * If TRUE then subtypes of the reference type are followed.
      * @property includeSubtypes
      * @type {Boolean}
      */
    self.includeSubtypes = initialize_field(schema.fields[2], options.includeSubtypes);

    /**
      * The browse name of the target.
      * @property targetName
      * @type {QualifiedName}
      */
    self.targetName =  new QualifiedName( options.targetName);

   // Object.preventExtensions(self);
}
util.inherits(RelativePathElement,BaseUAObject);
RelativePathElement.prototype.encodingDefaultBinary = makeExpandedNodeId(539,0);
RelativePathElement.prototype.encodingDefaultXml = makeExpandedNodeId(538,0);
RelativePathElement.prototype._schema = schema;

const encode_NodeId = _defaultTypeMap.NodeId.encode;
const decode_NodeId = _defaultTypeMap.NodeId.decode;
const encode_Boolean = _defaultTypeMap.Boolean.encode;
const decode_Boolean = _defaultTypeMap.Boolean.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
RelativePathElement.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_NodeId(this.referenceTypeId,stream);
    encode_Boolean(this.isInverse,stream);
    encode_Boolean(this.includeSubtypes,stream);
   this.targetName.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
RelativePathElement.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.referenceTypeId = decode_NodeId(stream);
    this.isInverse = decode_Boolean(stream);
    this.includeSubtypes = decode_Boolean(stream);
    this.targetName.decode(stream);
};
RelativePathElement.possibleFields = [
  "referenceTypeId",
         "isInverse",
         "includeSubtypes",
         "targetName"
];


exports.RelativePathElement = RelativePathElement;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("RelativePathElement",RelativePathElement);
