// node-opcua-service-translate-browse-path/_generated_/_auto_generated_TranslateBrowsePathsToNodeIdsRequest.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:11.820Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/TranslateBrowsePathsToNodeIdsRequest_schema").TranslateBrowsePathsToNodeIdsRequest_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const RequestHeader = getFactory("RequestHeader");
const BrowsePath = require("./_auto_generated_BrowsePath").BrowsePath;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class TranslateBrowsePathsToNodeIdsRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function TranslateBrowsePathsToNodeIdsRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestHeader =  null; /* new RequestHeader(null); */
        self.browsePath =  null; /* null array */
        return ;
    }

    /**
      * 
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * The list of paths to translate.
      * @property browsePath
      * @type {BrowsePath[]}
      */
    self.browsePath = [];
    if (options.browsePath) {
        assert(_.isArray(options.browsePath));
        self.browsePath = options.browsePath.map(function(e){ return new BrowsePath(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(TranslateBrowsePathsToNodeIdsRequest,BaseUAObject);
TranslateBrowsePathsToNodeIdsRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(554,0);
TranslateBrowsePathsToNodeIdsRequest.prototype.encodingDefaultXml = makeExpandedNodeId(553,0);
TranslateBrowsePathsToNodeIdsRequest.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
TranslateBrowsePathsToNodeIdsRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encodeArray(this.browsePath,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
TranslateBrowsePathsToNodeIdsRequest.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.requestHeader.decode(stream);
    this.browsePath = decodeArray(stream, function(stream) { 
       const obj = new BrowsePath(null);
       obj.decode(stream);
       return obj; 
    });
};
TranslateBrowsePathsToNodeIdsRequest.possibleFields = [
  "requestHeader",
         "browsePath"
];


exports.TranslateBrowsePathsToNodeIdsRequest = TranslateBrowsePathsToNodeIdsRequest;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("TranslateBrowsePathsToNodeIdsRequest",TranslateBrowsePathsToNodeIdsRequest);
