// node-opcua-service-translate-browse-path/_generated_/_auto_generated_BrowsePath.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:11.763Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/BrowsePath_schema").BrowsePath_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const RelativePath = require("./_auto_generated_RelativePath").RelativePath;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class BrowsePath
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function BrowsePath(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.relativePath =  null; /* new RelativePath(null); */
        return ;
    }

    /**
      * 
      * @property startingNode
      * @type {NodeId}
      */
    self.startingNode = initialize_field(schema.fields[0], options.startingNode);

    /**
      * 
      * @property relativePath
      * @type {RelativePath}
      */
    self.relativePath =  new RelativePath( options.relativePath);

   // Object.preventExtensions(self);
}
util.inherits(BrowsePath,BaseUAObject);
BrowsePath.prototype.encodingDefaultBinary = makeExpandedNodeId(545,0);
BrowsePath.prototype.encodingDefaultXml = makeExpandedNodeId(544,0);
BrowsePath.prototype._schema = schema;

const encode_NodeId = _defaultTypeMap.NodeId.encode;
const decode_NodeId = _defaultTypeMap.NodeId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
BrowsePath.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_NodeId(this.startingNode,stream);
   this.relativePath.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
BrowsePath.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.startingNode = decode_NodeId(stream);
    this.relativePath.decode(stream);
};
BrowsePath.possibleFields = [
  "startingNode",
         "relativePath"
];


exports.BrowsePath = BrowsePath;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("BrowsePath",BrowsePath);
