// node-opcua-service-translate-browse-path/_generated_/_auto_generated_BrowsePathResult.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:11.771Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/BrowsePathResult_schema").BrowsePathResult_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BrowsePathTarget = require("./_auto_generated_BrowsePathTarget").BrowsePathTarget;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * The result of a translate operation.
 * 
 * @class BrowsePathResult
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function BrowsePathResult(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.targets =  null; /* null array */
        return ;
    }

    /**
      * A code indicating any error during the operation.
      * @property statusCode
      * @type {StatusCode}
      */
    self.statusCode = initialize_field(schema.fields[0], options.statusCode);

    /**
      * A list of nodes found. The first element matches the type definition.
      * @property targets
      * @type {BrowsePathTarget[]}
      */
    self.targets = [];
    if (options.targets) {
        assert(_.isArray(options.targets));
        self.targets = options.targets.map(function(e){ return new BrowsePathTarget(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(BrowsePathResult,BaseUAObject);
BrowsePathResult.prototype.encodingDefaultBinary = makeExpandedNodeId(551,0);
BrowsePathResult.prototype.encodingDefaultXml = makeExpandedNodeId(550,0);
BrowsePathResult.prototype._schema = schema;

const encode_StatusCode = _defaultTypeMap.StatusCode.encode;
const decode_StatusCode = _defaultTypeMap.StatusCode.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
BrowsePathResult.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_StatusCode(this.statusCode,stream);
    encodeArray(this.targets,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
BrowsePathResult.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.statusCode = decode_StatusCode(stream);
    this.targets = decodeArray(stream, function(stream) { 
       const obj = new BrowsePathTarget(null);
       obj.decode(stream);
       return obj; 
    });
};
BrowsePathResult.possibleFields = [
  "statusCode",
         "targets"
];


exports.BrowsePathResult = BrowsePathResult;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("BrowsePathResult",BrowsePathResult);
