// node-opcua-service-session/_generated_/_auto_generated_ActivateSessionRequest.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:15.031Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/ActivateSessionRequest_schema").ActivateSessionRequest_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const RequestHeader = getFactory("RequestHeader");
const SignatureData = getFactory("SignatureData");
const SignedSoftwareCertificate = require("./_auto_generated_SignedSoftwareCertificate").SignedSoftwareCertificate;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * Activates a session with the server.
 * 
 * @class ActivateSessionRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function ActivateSessionRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestHeader =  null; /* new RequestHeader(null); */
        self.clientSignature =  null; /* new SignatureData(null); */
        self.clientSoftwareCertificates =  null; /* null array */
        self.userTokenSignature =  null; /* new SignatureData(null); */
        return ;
    }

    /**
      * A standard header included in all requests sent to a server.
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * A signature created with the client certificate from the last server nonce returned by the server.
      * @property clientSignature
      * @type {SignatureData}
      */
    self.clientSignature =  new SignatureData( options.clientSignature);

    /**
      * The software certificates owned by the client.
      * @property clientSoftwareCertificates
      * @type {SignedSoftwareCertificate[]}
      */
    self.clientSoftwareCertificates = [];
    if (options.clientSoftwareCertificates) {
        assert(_.isArray(options.clientSoftwareCertificates));
        self.clientSoftwareCertificates = options.clientSoftwareCertificates.map(function(e){ return new SignedSoftwareCertificate(e); } );
    }

    /**
      * The locales to use with the session.
      * @property localeIds
      * @type {LocaleId[]}
      */
    self.localeIds = initialize_field_array(schema.fields[3], options.localeIds);

    /**
      * The user identity to use with the session.
      * @property userIdentityToken
      * @type {ExtensionObject}
      */
    self.userIdentityToken = initialize_field(schema.fields[4], options.userIdentityToken);

    /**
      * A digital signature created with the user identity token.
      * @property userTokenSignature
      * @type {SignatureData}
      */
    self.userTokenSignature =  new SignatureData( options.userTokenSignature);

   // Object.preventExtensions(self);
}
util.inherits(ActivateSessionRequest,BaseUAObject);
ActivateSessionRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(467,0);
ActivateSessionRequest.prototype.encodingDefaultXml = makeExpandedNodeId(466,0);
ActivateSessionRequest.prototype._schema = schema;

const encode_LocaleId = _defaultTypeMap.LocaleId.encode;
const decode_LocaleId = _defaultTypeMap.LocaleId.decode;
const encode_ExtensionObject = _defaultTypeMap.ExtensionObject.encode;
const decode_ExtensionObject = _defaultTypeMap.ExtensionObject.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ActivateSessionRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
   this.clientSignature.encode(stream,options);
    encodeArray(this.clientSoftwareCertificates,stream,function(obj,stream){ obj.encode(stream,options); }); 
    encodeArray(this.localeIds, stream, encode_LocaleId);
    encode_ExtensionObject(this.userIdentityToken,stream);
   this.userTokenSignature.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
ActivateSessionRequest.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.requestHeader.decode(stream);
    this.clientSignature.decode(stream);
    this.clientSoftwareCertificates = decodeArray(stream, function(stream) { 
       const obj = new SignedSoftwareCertificate(null);
       obj.decode(stream);
       return obj; 
    });
    this.localeIds = decodeArray(stream, decode_LocaleId);
    this.userIdentityToken = decode_ExtensionObject(stream);
    this.userTokenSignature.decode(stream);
};
ActivateSessionRequest.possibleFields = [
  "requestHeader",
         "clientSignature",
         "clientSoftwareCertificates",
         "localeIds",
         "userIdentityToken",
         "userTokenSignature"
];


exports.ActivateSessionRequest = ActivateSessionRequest;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ActivateSessionRequest",ActivateSessionRequest);
