// node-opcua-service-secure-channel/_generated_/_auto_generated_AsymmetricAlgorithmSecurityHeader.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:15.010Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/AsymmetricAlgorithmSecurityHeader_schema").AsymmetricAlgorithmSecurityHeader_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class AsymmetricAlgorithmSecurityHeader
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.securityPolicyUri] {String} 
 * @param  [options.senderCertificate = null] {ByteString} 
 * @param  [options.receiverCertificateThumbprint = null] {ByteString} 
 */
function AsymmetricAlgorithmSecurityHeader(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property securityPolicyUri
      * @type {String}
      */
    self.securityPolicyUri = initialize_field(schema.fields[0], options.securityPolicyUri);

    /**
      * 
      * @property senderCertificate
      * @type {ByteString}
      * @default  null
      */
    self.senderCertificate = initialize_field(schema.fields[1], options.senderCertificate);

    /**
      * 
      * @property receiverCertificateThumbprint
      * @type {ByteString}
      * @default  null
      */
    self.receiverCertificateThumbprint = initialize_field(schema.fields[2], options.receiverCertificateThumbprint);

   // Object.preventExtensions(self);
}
util.inherits(AsymmetricAlgorithmSecurityHeader,BaseUAObject);
schema.id = generate_new_id();
AsymmetricAlgorithmSecurityHeader.prototype.encodingDefaultBinary = makeExpandedNodeId(schema.id);
AsymmetricAlgorithmSecurityHeader.prototype._schema = schema;

var encode_String = _defaultTypeMap.String.encode;
var decode_String = _defaultTypeMap.String.decode;
var encode_ByteString = _defaultTypeMap.ByteString.encode;
var decode_ByteString = _defaultTypeMap.ByteString.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
AsymmetricAlgorithmSecurityHeader.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_String(this.securityPolicyUri,stream);
    encode_ByteString(this.senderCertificate,stream);
    encode_ByteString(this.receiverCertificateThumbprint,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
AsymmetricAlgorithmSecurityHeader.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.securityPolicyUri = decode_String(stream,options);
    this.senderCertificate = decode_ByteString(stream,options);
    this.receiverCertificateThumbprint = decode_ByteString(stream,options);
};
AsymmetricAlgorithmSecurityHeader.possibleFields = [
  "securityPolicyUri",
         "senderCertificate",
         "receiverCertificateThumbprint"
];


exports.AsymmetricAlgorithmSecurityHeader = AsymmetricAlgorithmSecurityHeader;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("AsymmetricAlgorithmSecurityHeader",AsymmetricAlgorithmSecurityHeader);
