// node-opcua-service-secure-channel/_generated_/_auto_generated_ErrorMessage.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:15.044Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ErrorMessage_schema").ErrorMessage_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ErrorMessage
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.Error] {UInt32} The numeric code for the error. This shall be one of the values listed in Table 40.
 * @param  [options.Reason] {String} A more verbose description of the error.This string shall not be more than 4096 characters.
 */
function ErrorMessage(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * The numeric code for the error. This shall be one of the values listed in Table 40.
      * @property Error
      * @type {UInt32}
      */
    self.Error = initialize_field(schema.fields[0], options.Error);

    /**
      * A more verbose description of the error.This string shall not be more than 4096 characters.
      * @property Reason
      * @type {String}
      */
    self.Reason = initialize_field(schema.fields[1], options.Reason);

   // Object.preventExtensions(self);
}
util.inherits(ErrorMessage,BaseUAObject);
schema.id = generate_new_id();
ErrorMessage.prototype.encodingDefaultBinary = makeExpandedNodeId(schema.id);
ErrorMessage.prototype._schema = schema;

var encode_UInt32 = _defaultTypeMap.UInt32.encode;
var decode_UInt32 = _defaultTypeMap.UInt32.decode;
var encode_String = _defaultTypeMap.String.encode;
var decode_String = _defaultTypeMap.String.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ErrorMessage.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_UInt32(this.Error,stream);
    encode_String(this.Reason,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
ErrorMessage.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.Error = decode_UInt32(stream,options);
    this.Reason = decode_String(stream,options);
};
ErrorMessage.possibleFields = [
  "Error",
         "Reason"
];


exports.ErrorMessage = ErrorMessage;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ErrorMessage",ErrorMessage);
