// node-opcua-service-secure-channel/_generated_/_auto_generated_OpenSecureChannelResponse.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:15.149Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/OpenSecureChannelResponse_schema").OpenSecureChannelResponse_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var ResponseHeader = require("./_auto_generated_ResponseHeader").ResponseHeader;
var ChannelSecurityToken = require("./_auto_generated_ChannelSecurityToken").ChannelSecurityToken;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class OpenSecureChannelResponse
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.responseHeader] {ResponseHeader} 
 * @param  [options.serverProtocolVersion] {UInt32} 
 * @param  [options.securityToken] {ChannelSecurityToken} 
 * @param  [options.serverNonce] {ByteString} 
 */
function OpenSecureChannelResponse(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.responseHeader =  null; /* new ResponseHeader(null); */
        self.securityToken =  null; /* new ChannelSecurityToken(null); */
        return ;
    }

    /**
      * 
      * @property responseHeader
      * @type {ResponseHeader}
      */
    self.responseHeader =  new ResponseHeader( options.responseHeader);

    /**
      * 
      * @property serverProtocolVersion
      * @type {UInt32}
      */
    self.serverProtocolVersion = initialize_field(schema.fields[1], options.serverProtocolVersion);

    /**
      * 
      * @property securityToken
      * @type {ChannelSecurityToken}
      */
    self.securityToken =  new ChannelSecurityToken( options.securityToken);

    /**
      * 
      * @property serverNonce
      * @type {ByteString}
      */
    self.serverNonce = initialize_field(schema.fields[3], options.serverNonce);

   // Object.preventExtensions(self);
}
util.inherits(OpenSecureChannelResponse,BaseUAObject);
OpenSecureChannelResponse.prototype.encodingDefaultBinary = makeExpandedNodeId(449,0);
OpenSecureChannelResponse.prototype.encodingDefaultXml = makeExpandedNodeId(448,0);
OpenSecureChannelResponse.prototype._schema = schema;

var encode_UInt32 = _defaultTypeMap.UInt32.encode;
var decode_UInt32 = _defaultTypeMap.UInt32.decode;
var encode_ByteString = _defaultTypeMap.ByteString.encode;
var decode_ByteString = _defaultTypeMap.ByteString.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
OpenSecureChannelResponse.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.responseHeader.encode(stream,options);
    encode_UInt32(this.serverProtocolVersion,stream);
   this.securityToken.encode(stream,options);
    encode_ByteString(this.serverNonce,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
OpenSecureChannelResponse.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.responseHeader.decode(stream,options);
    this.serverProtocolVersion = decode_UInt32(stream,options);
    this.securityToken.decode(stream,options);
    this.serverNonce = decode_ByteString(stream,options);
};
OpenSecureChannelResponse.possibleFields = [
  "responseHeader",
         "serverProtocolVersion",
         "securityToken",
         "serverNonce"
];


exports.OpenSecureChannelResponse = OpenSecureChannelResponse;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("OpenSecureChannelResponse",OpenSecureChannelResponse);
