// node-opcua-service-secure-channel/_generated_/_auto_generated_ChannelSecurityToken.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:15.120Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ChannelSecurityToken_schema").ChannelSecurityToken_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ChannelSecurityToken
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.secureChannelId] {UInt32} 
 * @param  [options.tokenId] {UInt32} 
 * @param  [options.createdAt = Thu Feb 08 2018 08:25:15 GMT+0100 (Paris, Madrid)] {UtcTime} 
 * @param  [options.revisedLifeTime = 30000] {UInt32} 
 */
function ChannelSecurityToken(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property secureChannelId
      * @type {UInt32}
      */
    self.secureChannelId = initialize_field(schema.fields[0], options.secureChannelId);

    /**
      * 
      * @property tokenId
      * @type {UInt32}
      */
    self.tokenId = initialize_field(schema.fields[1], options.tokenId);

    /**
      * 
      * @property createdAt
      * @type {UtcTime}
      * @default  function () {
            return new Date();
        }
      */
    self.createdAt = initialize_field(schema.fields[2], options.createdAt);

    /**
      * 
      * @property revisedLifeTime
      * @type {UInt32}
      * @default  30000
      */
    self.revisedLifeTime = initialize_field(schema.fields[3], options.revisedLifeTime);

   // Object.preventExtensions(self);
}
util.inherits(ChannelSecurityToken,BaseUAObject);
ChannelSecurityToken.prototype.encodingDefaultBinary = makeExpandedNodeId(443,0);
ChannelSecurityToken.prototype.encodingDefaultXml = makeExpandedNodeId(442,0);
ChannelSecurityToken.prototype._schema = schema;

var encode_UInt32 = _defaultTypeMap.UInt32.encode;
var decode_UInt32 = _defaultTypeMap.UInt32.decode;
var encode_UtcTime = _defaultTypeMap.UtcTime.encode;
var decode_UtcTime = _defaultTypeMap.UtcTime.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ChannelSecurityToken.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_UInt32(this.secureChannelId,stream);
    encode_UInt32(this.tokenId,stream);
    encode_UtcTime(this.createdAt,stream);
    encode_UInt32(this.revisedLifeTime,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
ChannelSecurityToken.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.secureChannelId = decode_UInt32(stream,options);
    this.tokenId = decode_UInt32(stream,options);
    this.createdAt = decode_UtcTime(stream,options);
    this.revisedLifeTime = decode_UInt32(stream,options);
};
ChannelSecurityToken.possibleFields = [
  "secureChannelId",
         "tokenId",
         "createdAt",
         "revisedLifeTime"
];


exports.ChannelSecurityToken = ChannelSecurityToken;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ChannelSecurityToken",ChannelSecurityToken);
