// node-opcua-service-secure-channel/_generated_/_auto_generated_SignatureData.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:15.126Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/SignatureData_schema").SignatureData_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class SignatureData
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.algorithm = null] {String} The cryptography algorithm used to create the signature.
 * @param  [options.signature = null] {ByteString} The digital signature.
 */
function SignatureData(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * The cryptography algorithm used to create the signature.
      * @property algorithm
      * @type {String}
      * @default  null
      */
    self.algorithm = initialize_field(schema.fields[0], options.algorithm);

    /**
      * The digital signature.
      * @property signature
      * @type {ByteString}
      * @default  null
      */
    self.signature = initialize_field(schema.fields[1], options.signature);

   // Object.preventExtensions(self);
}
util.inherits(SignatureData,BaseUAObject);
SignatureData.prototype.encodingDefaultBinary = makeExpandedNodeId(458,0);
SignatureData.prototype.encodingDefaultXml = makeExpandedNodeId(457,0);
SignatureData.prototype._schema = schema;

var encode_String = _defaultTypeMap.String.encode;
var decode_String = _defaultTypeMap.String.decode;
var encode_ByteString = _defaultTypeMap.ByteString.encode;
var decode_ByteString = _defaultTypeMap.ByteString.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
SignatureData.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_String(this.algorithm,stream);
    encode_ByteString(this.signature,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
SignatureData.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.algorithm = decode_String(stream,options);
    this.signature = decode_ByteString(stream,options);
};
SignatureData.possibleFields = [
  "algorithm",
         "signature"
];


exports.SignatureData = SignatureData;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("SignatureData",SignatureData);
