// node-opcua-service-secure-channel/_generated_/_auto_generated_SymmetricAlgorithmSecurityHeader.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:15.034Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/SymmetricAlgorithmSecurityHeader_schema").SymmetricAlgorithmSecurityHeader_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class SymmetricAlgorithmSecurityHeader
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.tokenId = 3735928559] {UInt32} 
 */
function SymmetricAlgorithmSecurityHeader(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property tokenId
      * @type {UInt32}
      * @default  3735928559
      */
    self.tokenId = initialize_field(schema.fields[0], options.tokenId);

   // Object.preventExtensions(self);
}
util.inherits(SymmetricAlgorithmSecurityHeader,BaseUAObject);
schema.id = generate_new_id();
SymmetricAlgorithmSecurityHeader.prototype.encodingDefaultBinary = makeExpandedNodeId(schema.id);
SymmetricAlgorithmSecurityHeader.prototype._schema = schema;

var encode_UInt32 = _defaultTypeMap.UInt32.encode;
var decode_UInt32 = _defaultTypeMap.UInt32.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
SymmetricAlgorithmSecurityHeader.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_UInt32(this.tokenId,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
SymmetricAlgorithmSecurityHeader.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.tokenId = decode_UInt32(stream,options);
};
SymmetricAlgorithmSecurityHeader.possibleFields = [
  "tokenId"
];


exports.SymmetricAlgorithmSecurityHeader = SymmetricAlgorithmSecurityHeader;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("SymmetricAlgorithmSecurityHeader",SymmetricAlgorithmSecurityHeader);
