// node-opcua-service-subscription/_generated_/_auto_generated_AggregateFilter.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:18.690Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/AggregateFilter_schema").AggregateFilter_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const AggregateConfiguration = getFactory("AggregateConfiguration");
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class AggregateFilter
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function AggregateFilter(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.AggregateType =  null; /* new AggregateConfiguration(null); */
        return ;
    }

    /**
      * 
      * @property StartTime
      * @type {DateTime}
      */
    self.StartTime = initialize_field(schema.fields[0], options.StartTime);

    /**
      * 
      * @property AggregateType
      * @type {NodeId}
      */
    self.AggregateType = initialize_field(schema.fields[1], options.AggregateType);

    /**
      * 
      * @property ProcessingInterval
      * @type {Double}
      */
    self.ProcessingInterval = initialize_field(schema.fields[2], options.ProcessingInterval);

    /**
      * 
      * @property AggregateType
      * @type {AggregateConfiguration}
      */
    self.AggregateType =  new AggregateConfiguration( options.AggregateType);

   // Object.preventExtensions(self);
}
util.inherits(AggregateFilter,BaseUAObject);
AggregateFilter.prototype.encodingDefaultBinary = makeExpandedNodeId(730,0);
AggregateFilter.prototype.encodingDefaultXml = makeExpandedNodeId(729,0);
AggregateFilter.prototype._schema = schema;

const encode_DateTime = _defaultTypeMap.DateTime.encode;
const decode_DateTime = _defaultTypeMap.DateTime.decode;
const encode_NodeId = _defaultTypeMap.NodeId.encode;
const decode_NodeId = _defaultTypeMap.NodeId.decode;
const encode_Double = _defaultTypeMap.Double.encode;
const decode_Double = _defaultTypeMap.Double.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
AggregateFilter.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_DateTime(this.StartTime,stream);
    encode_NodeId(this.AggregateType,stream);
    encode_Double(this.ProcessingInterval,stream);
   this.AggregateType.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
AggregateFilter.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.StartTime = decode_DateTime(stream);
    this.AggregateType = decode_NodeId(stream);
    this.ProcessingInterval = decode_Double(stream);
    this.AggregateType.decode(stream);
};
AggregateFilter.possibleFields = [
  "StartTime",
         "AggregateType",
         "ProcessingInterval",
         "AggregateType"
];


exports.AggregateFilter = AggregateFilter;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("AggregateFilter",AggregateFilter);
