// node-opcua-service-subscription/_generated_/_auto_generated_ModifySubscriptionResponse.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:18.510Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/ModifySubscriptionResponse_schema").ModifySubscriptionResponse_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const ResponseHeader = getFactory("ResponseHeader");
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ModifySubscriptionResponse
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function ModifySubscriptionResponse(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.responseHeader =  null; /* new ResponseHeader(null); */
        return ;
    }

    /**
      * 
      * @property responseHeader
      * @type {ResponseHeader}
      */
    self.responseHeader =  new ResponseHeader( options.responseHeader);

    /**
      * 
      * @property revisedPublishingInterval
      * @type {Duration}
      */
    self.revisedPublishingInterval = initialize_field(schema.fields[1], options.revisedPublishingInterval);

    /**
      * 
      * @property revisedLifetimeCount
      * @type {Counter}
      */
    self.revisedLifetimeCount = initialize_field(schema.fields[2], options.revisedLifetimeCount);

    /**
      * 
      * @property revisedMaxKeepAliveCount
      * @type {Counter}
      */
    self.revisedMaxKeepAliveCount = initialize_field(schema.fields[3], options.revisedMaxKeepAliveCount);

   // Object.preventExtensions(self);
}
util.inherits(ModifySubscriptionResponse,BaseUAObject);
ModifySubscriptionResponse.prototype.encodingDefaultBinary = makeExpandedNodeId(796,0);
ModifySubscriptionResponse.prototype.encodingDefaultXml = makeExpandedNodeId(795,0);
ModifySubscriptionResponse.prototype._schema = schema;

const encode_Duration = _defaultTypeMap.Duration.encode;
const decode_Duration = _defaultTypeMap.Duration.decode;
const encode_Counter = _defaultTypeMap.Counter.encode;
const decode_Counter = _defaultTypeMap.Counter.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ModifySubscriptionResponse.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.responseHeader.encode(stream,options);
    encode_Duration(this.revisedPublishingInterval,stream);
    encode_Counter(this.revisedLifetimeCount,stream);
    encode_Counter(this.revisedMaxKeepAliveCount,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
ModifySubscriptionResponse.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.responseHeader.decode(stream);
    this.revisedPublishingInterval = decode_Duration(stream);
    this.revisedLifetimeCount = decode_Counter(stream);
    this.revisedMaxKeepAliveCount = decode_Counter(stream);
};
ModifySubscriptionResponse.possibleFields = [
  "responseHeader",
         "revisedPublishingInterval",
         "revisedLifetimeCount",
         "revisedMaxKeepAliveCount"
];


exports.ModifySubscriptionResponse = ModifySubscriptionResponse;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ModifySubscriptionResponse",ModifySubscriptionResponse);
