// node-opcua-service-subscription/_generated_/_auto_generated_StatusChangeNotification.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:18.650Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/StatusChangeNotification_schema").StatusChangeNotification_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const DiagnosticInfo = getFactory("DiagnosticInfo");
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class StatusChangeNotification
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function StatusChangeNotification(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.diagnosticInfo =  null; /* new DiagnosticInfo(null); */
        return ;
    }

    /**
      * The StatusCode that indicates the status change.
      * @property statusCode
      * @type {StatusCode}
      */
    self.statusCode = initialize_field(schema.fields[0], options.statusCode);

    /**
      * 
      * @property diagnosticInfo
      * @type {DiagnosticInfo}
      */
    self.diagnosticInfo =  new DiagnosticInfo( options.diagnosticInfo);

   // Object.preventExtensions(self);
}
util.inherits(StatusChangeNotification,BaseUAObject);
StatusChangeNotification.prototype.encodingDefaultBinary = makeExpandedNodeId(820,0);
StatusChangeNotification.prototype.encodingDefaultXml = makeExpandedNodeId(819,0);
StatusChangeNotification.prototype._schema = schema;

const encode_StatusCode = _defaultTypeMap.StatusCode.encode;
const decode_StatusCode = _defaultTypeMap.StatusCode.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
StatusChangeNotification.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_StatusCode(this.statusCode,stream);
   this.diagnosticInfo.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
StatusChangeNotification.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.statusCode = decode_StatusCode(stream);
    this.diagnosticInfo.decode(stream);
};
StatusChangeNotification.possibleFields = [
  "statusCode",
         "diagnosticInfo"
];


exports.StatusChangeNotification = StatusChangeNotification;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("StatusChangeNotification",StatusChangeNotification);
