// node-opcua-service-subscription/_generated_/_auto_generated_NotificationMessage.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:18.547Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/NotificationMessage_schema").NotificationMessage_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class NotificationMessage
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function NotificationMessage(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property sequenceNumber
      * @type {Counter}
      */
    self.sequenceNumber = initialize_field(schema.fields[0], options.sequenceNumber);

    /**
      * 
      * @property publishTime
      * @type {UtcTime}
      */
    self.publishTime = initialize_field(schema.fields[1], options.publishTime);

    /**
      * 
      * @property notificationData
      * @type {ExtensionObject[]}
      */
    self.notificationData = initialize_field_array(schema.fields[2], options.notificationData);

   // Object.preventExtensions(self);
}
util.inherits(NotificationMessage,BaseUAObject);
NotificationMessage.prototype.encodingDefaultBinary = makeExpandedNodeId(805,0);
NotificationMessage.prototype.encodingDefaultXml = makeExpandedNodeId(804,0);
NotificationMessage.prototype._schema = schema;

const encode_Counter = _defaultTypeMap.Counter.encode;
const decode_Counter = _defaultTypeMap.Counter.decode;
const encode_UtcTime = _defaultTypeMap.UtcTime.encode;
const decode_UtcTime = _defaultTypeMap.UtcTime.decode;
const encode_ExtensionObject = _defaultTypeMap.ExtensionObject.encode;
const decode_ExtensionObject = _defaultTypeMap.ExtensionObject.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
NotificationMessage.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_Counter(this.sequenceNumber,stream);
    encode_UtcTime(this.publishTime,stream);
    encodeArray(this.notificationData, stream, encode_ExtensionObject);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
NotificationMessage.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.sequenceNumber = decode_Counter(stream);
    this.publishTime = decode_UtcTime(stream);
    this.notificationData = decodeArray(stream, decode_ExtensionObject);
};
NotificationMessage.possibleFields = [
  "sequenceNumber",
         "publishTime",
         "notificationData"
];


exports.NotificationMessage = NotificationMessage;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("NotificationMessage",NotificationMessage);
