// node-opcua-service-subscription/_generated_/_auto_generated_ModifyMonitoredItemsRequest.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:18.624Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/ModifyMonitoredItemsRequest_schema").ModifyMonitoredItemsRequest_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const RequestHeader = getFactory("RequestHeader");
const MonitoredItemModifyRequest = require("./_auto_generated_MonitoredItemModifyRequest").MonitoredItemModifyRequest;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ModifyMonitoredItemsRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function ModifyMonitoredItemsRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestHeader =  null; /* new RequestHeader(null); */
        self.itemsToModify =  null; /* null array */
        return ;
    }

    /**
      * 
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * 
      * @property subscriptionId
      * @type {IntegerId}
      */
    self.subscriptionId = initialize_field(schema.fields[1], options.subscriptionId);

    /**
      * 
      * @property timestampsToReturn
      * @type {TimestampsToReturn}
      */
    self.setTimestampsToReturn(initialize_field(schema.fields[2], options.timestampsToReturn));

    /**
      * 
      * @property itemsToModify
      * @type {MonitoredItemModifyRequest[]}
      */
    self.itemsToModify = [];
    if (options.itemsToModify) {
        assert(_.isArray(options.itemsToModify));
        self.itemsToModify = options.itemsToModify.map(function(e){ return new MonitoredItemModifyRequest(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(ModifyMonitoredItemsRequest,BaseUAObject);

//## Define Enumeration setters
ModifyMonitoredItemsRequest.prototype.setTimestampsToReturn = function(value) {
   const coercedValue = _enumerations.TimestampsToReturn.typedEnum.get(value);
   /* istanbul ignore next */
   if (coercedValue === undefined || coercedValue === null) {
      throw new Error("value cannot be coerced to TimestampsToReturn: " + value);
   }
   this.timestampsToReturn = coercedValue;
};
ModifyMonitoredItemsRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(763,0);
ModifyMonitoredItemsRequest.prototype.encodingDefaultXml = makeExpandedNodeId(762,0);
ModifyMonitoredItemsRequest.prototype._schema = schema;

const encode_IntegerId = _defaultTypeMap.IntegerId.encode;
const decode_IntegerId = _defaultTypeMap.IntegerId.decode;
const encode_TimestampsToReturn = _enumerations.TimestampsToReturn.encode;
const decode_TimestampsToReturn = _enumerations.TimestampsToReturn.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ModifyMonitoredItemsRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encode_IntegerId(this.subscriptionId,stream);
    encode_TimestampsToReturn(this.timestampsToReturn,stream);
    encodeArray(this.itemsToModify,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
ModifyMonitoredItemsRequest.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.requestHeader.decode(stream);
    this.subscriptionId = decode_IntegerId(stream);
    this.timestampsToReturn = decode_TimestampsToReturn(stream);
    this.itemsToModify = decodeArray(stream, function(stream) { 
       const obj = new MonitoredItemModifyRequest(null);
       obj.decode(stream);
       return obj; 
    });
};
ModifyMonitoredItemsRequest.possibleFields = [
  "requestHeader",
         "subscriptionId",
         "timestampsToReturn",
         "itemsToModify"
];


exports.ModifyMonitoredItemsRequest = ModifyMonitoredItemsRequest;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ModifyMonitoredItemsRequest",ModifyMonitoredItemsRequest);
