// node-opcua-service-subscription/_generated_/_auto_generated_MonitoredItemNotification.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:18.600Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/MonitoredItemNotification_schema").MonitoredItemNotification_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const DataValue = getFactory("DataValue");
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class MonitoredItemNotification
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function MonitoredItemNotification(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.value =  null; /* new DataValue(null); */
        return ;
    }

    /**
      * Client-supplied handle for the MonitoredItem. 
      * @property clientHandle
      * @type {IntegerId}
      */
    self.clientHandle = initialize_field(schema.fields[0], options.clientHandle);

    /**
      * The StatusCode, value and timestamp(s) of the monitored Attribute depending on the sampling and queuing configuration.
      * @property value
      * @type {DataValue}
      */
    self.value =  new DataValue( options.value);

   // Object.preventExtensions(self);
}
util.inherits(MonitoredItemNotification,BaseUAObject);
MonitoredItemNotification.prototype.encodingDefaultBinary = makeExpandedNodeId(808,0);
MonitoredItemNotification.prototype.encodingDefaultXml = makeExpandedNodeId(807,0);
MonitoredItemNotification.prototype._schema = schema;

const encode_IntegerId = _defaultTypeMap.IntegerId.encode;
const decode_IntegerId = _defaultTypeMap.IntegerId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
MonitoredItemNotification.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_IntegerId(this.clientHandle,stream);
   this.value.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
MonitoredItemNotification.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.clientHandle = decode_IntegerId(stream);
    this.value.decode(stream);
};
MonitoredItemNotification.possibleFields = [
  "clientHandle",
         "value"
];


exports.MonitoredItemNotification = MonitoredItemNotification;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("MonitoredItemNotification",MonitoredItemNotification);
