// node-opcua-service-subscription/_generated_/_auto_generated_MonitoringParameters.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:18.515Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/MonitoringParameters_schema").MonitoringParameters_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class MonitoringParameters
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function MonitoringParameters(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property clientHandle
      * @type {IntegerId}
      */
    self.clientHandle = initialize_field(schema.fields[0], options.clientHandle);

    /**
      * 
      * @property samplingInterval
      * @type {Duration}
      */
    self.samplingInterval = initialize_field(schema.fields[1], options.samplingInterval);

    /**
      * 
      * @property filter
      * @type {ExtensionObject}
      */
    self.filter = initialize_field(schema.fields[2], options.filter);

    /**
      * 
      * @property queueSize
      * @type {Counter}
      */
    self.queueSize = initialize_field(schema.fields[3], options.queueSize);

    /**
      * 
      * @property discardOldest
      * @type {Boolean}
      */
    self.discardOldest = initialize_field(schema.fields[4], options.discardOldest);

   // Object.preventExtensions(self);
}
util.inherits(MonitoringParameters,BaseUAObject);
MonitoringParameters.prototype.encodingDefaultBinary = makeExpandedNodeId(742,0);
MonitoringParameters.prototype.encodingDefaultXml = makeExpandedNodeId(741,0);
MonitoringParameters.prototype._schema = schema;

const encode_IntegerId = _defaultTypeMap.IntegerId.encode;
const decode_IntegerId = _defaultTypeMap.IntegerId.decode;
const encode_Duration = _defaultTypeMap.Duration.encode;
const decode_Duration = _defaultTypeMap.Duration.decode;
const encode_ExtensionObject = _defaultTypeMap.ExtensionObject.encode;
const decode_ExtensionObject = _defaultTypeMap.ExtensionObject.decode;
const encode_Counter = _defaultTypeMap.Counter.encode;
const decode_Counter = _defaultTypeMap.Counter.decode;
const encode_Boolean = _defaultTypeMap.Boolean.encode;
const decode_Boolean = _defaultTypeMap.Boolean.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
MonitoringParameters.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_IntegerId(this.clientHandle,stream);
    encode_Duration(this.samplingInterval,stream);
    encode_ExtensionObject(this.filter,stream);
    encode_Counter(this.queueSize,stream);
    encode_Boolean(this.discardOldest,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
MonitoringParameters.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.clientHandle = decode_IntegerId(stream);
    this.samplingInterval = decode_Duration(stream);
    this.filter = decode_ExtensionObject(stream);
    this.queueSize = decode_Counter(stream);
    this.discardOldest = decode_Boolean(stream);
};
MonitoringParameters.possibleFields = [
  "clientHandle",
         "samplingInterval",
         "filter",
         "queueSize",
         "discardOldest"
];


exports.MonitoringParameters = MonitoringParameters;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("MonitoringParameters",MonitoringParameters);
