// node-opcua-service-subscription/_generated_/_auto_generated_ModifySubscriptionRequest.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:18.507Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/ModifySubscriptionRequest_schema").ModifySubscriptionRequest_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const RequestHeader = getFactory("RequestHeader");
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ModifySubscriptionRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function ModifySubscriptionRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestHeader =  null; /* new RequestHeader(null); */
        return ;
    }

    /**
      * 
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * 
      * @property subscriptionId
      * @type {IntegerId}
      */
    self.subscriptionId = initialize_field(schema.fields[1], options.subscriptionId);

    /**
      * 
      * @property requestedPublishingInterval
      * @type {Duration}
      */
    self.requestedPublishingInterval = initialize_field(schema.fields[2], options.requestedPublishingInterval);

    /**
      * 
      * @property requestedLifetimeCount
      * @type {Counter}
      */
    self.requestedLifetimeCount = initialize_field(schema.fields[3], options.requestedLifetimeCount);

    /**
      * 
      * @property requestedMaxKeepAliveCount
      * @type {Counter}
      */
    self.requestedMaxKeepAliveCount = initialize_field(schema.fields[4], options.requestedMaxKeepAliveCount);

    /**
      * 
      * @property maxNotificationsPerPublish
      * @type {Counter}
      */
    self.maxNotificationsPerPublish = initialize_field(schema.fields[5], options.maxNotificationsPerPublish);

    /**
      * 
      * @property priority
      * @type {Byte}
      */
    self.priority = initialize_field(schema.fields[6], options.priority);

   // Object.preventExtensions(self);
}
util.inherits(ModifySubscriptionRequest,BaseUAObject);
ModifySubscriptionRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(793,0);
ModifySubscriptionRequest.prototype.encodingDefaultXml = makeExpandedNodeId(792,0);
ModifySubscriptionRequest.prototype._schema = schema;

const encode_IntegerId = _defaultTypeMap.IntegerId.encode;
const decode_IntegerId = _defaultTypeMap.IntegerId.decode;
const encode_Duration = _defaultTypeMap.Duration.encode;
const decode_Duration = _defaultTypeMap.Duration.decode;
const encode_Counter = _defaultTypeMap.Counter.encode;
const decode_Counter = _defaultTypeMap.Counter.decode;
const encode_Byte = _defaultTypeMap.Byte.encode;
const decode_Byte = _defaultTypeMap.Byte.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ModifySubscriptionRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encode_IntegerId(this.subscriptionId,stream);
    encode_Duration(this.requestedPublishingInterval,stream);
    encode_Counter(this.requestedLifetimeCount,stream);
    encode_Counter(this.requestedMaxKeepAliveCount,stream);
    encode_Counter(this.maxNotificationsPerPublish,stream);
    encode_Byte(this.priority,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
ModifySubscriptionRequest.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.requestHeader.decode(stream);
    this.subscriptionId = decode_IntegerId(stream);
    this.requestedPublishingInterval = decode_Duration(stream);
    this.requestedLifetimeCount = decode_Counter(stream);
    this.requestedMaxKeepAliveCount = decode_Counter(stream);
    this.maxNotificationsPerPublish = decode_Counter(stream);
    this.priority = decode_Byte(stream);
};
ModifySubscriptionRequest.possibleFields = [
  "requestHeader",
         "subscriptionId",
         "requestedPublishingInterval",
         "requestedLifetimeCount",
         "requestedMaxKeepAliveCount",
         "maxNotificationsPerPublish",
         "priority"
];


exports.ModifySubscriptionRequest = ModifySubscriptionRequest;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ModifySubscriptionRequest",ModifySubscriptionRequest);
