// node-opcua-service-subscription/_generated_/_auto_generated_EventFieldList.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:18.641Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/EventFieldList_schema").EventFieldList_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const Variant = getFactory("Variant");
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class EventFieldList
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function EventFieldList(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.eventFields =  null; /* null array */
        return ;
    }

    /**
      * Client-supplied handle for the MonitoredItem
      * @property clientHandle
      * @type {IntegerId}
      */
    self.clientHandle = initialize_field(schema.fields[0], options.clientHandle);

    /**
      * List of selected Event fields. This shall be a one to one match with the fields selected in the EventFilter.
      * @property eventFields
      * @type {Variant[]}
      */
    self.eventFields = [];
    if (options.eventFields) {
        assert(_.isArray(options.eventFields));
        self.eventFields = options.eventFields.map(function(e){ return new Variant(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(EventFieldList,BaseUAObject);
schema.id = generate_new_id();
EventFieldList.prototype.encodingDefaultBinary = makeExpandedNodeId(schema.id);
EventFieldList.prototype._schema = schema;

const encode_IntegerId = _defaultTypeMap.IntegerId.encode;
const decode_IntegerId = _defaultTypeMap.IntegerId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
EventFieldList.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_IntegerId(this.clientHandle,stream);
    encodeArray(this.eventFields,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
EventFieldList.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.clientHandle = decode_IntegerId(stream);
    this.eventFields = decodeArray(stream, function(stream) { 
       const obj = new Variant(null);
       obj.decode(stream);
       return obj; 
    });
};
EventFieldList.possibleFields = [
  "clientHandle",
         "eventFields"
];


exports.EventFieldList = EventFieldList;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("EventFieldList",EventFieldList);
