// node-opcua-service-subscription/_generated_/_auto_generated_MonitoredItemCreateRequest.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:18.521Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/MonitoredItemCreateRequest_schema").MonitoredItemCreateRequest_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const ReadValueId = getFactory("ReadValueId");
const MonitoringParameters = require("./_auto_generated_MonitoringParameters").MonitoringParameters;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class MonitoredItemCreateRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function MonitoredItemCreateRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.itemToMonitor =  null; /* new ReadValueId(null); */
        self.requestedParameters =  null; /* new MonitoringParameters(null); */
        return ;
    }

    /**
      * 
      * @property itemToMonitor
      * @type {ReadValueId}
      */
    self.itemToMonitor =  new ReadValueId( options.itemToMonitor);

    /**
      * 
      * @property monitoringMode
      * @type {MonitoringMode}
      */
    self.setMonitoringMode(initialize_field(schema.fields[1], options.monitoringMode));

    /**
      * 
      * @property requestedParameters
      * @type {MonitoringParameters}
      */
    self.requestedParameters =  new MonitoringParameters( options.requestedParameters);

   // Object.preventExtensions(self);
}
util.inherits(MonitoredItemCreateRequest,BaseUAObject);

//## Define Enumeration setters
MonitoredItemCreateRequest.prototype.setMonitoringMode = function(value) {
   const coercedValue = _enumerations.MonitoringMode.typedEnum.get(value);
   /* istanbul ignore next */
   if (coercedValue === undefined || coercedValue === null) {
      throw new Error("value cannot be coerced to MonitoringMode: " + value);
   }
   this.monitoringMode = coercedValue;
};
MonitoredItemCreateRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(745,0);
MonitoredItemCreateRequest.prototype.encodingDefaultXml = makeExpandedNodeId(744,0);
MonitoredItemCreateRequest.prototype._schema = schema;

const encode_MonitoringMode = _enumerations.MonitoringMode.encode;
const decode_MonitoringMode = _enumerations.MonitoringMode.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
MonitoredItemCreateRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.itemToMonitor.encode(stream,options);
    encode_MonitoringMode(this.monitoringMode,stream);
   this.requestedParameters.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
MonitoredItemCreateRequest.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.itemToMonitor.decode(stream);
    this.monitoringMode = decode_MonitoringMode(stream);
    this.requestedParameters.decode(stream);
};
MonitoredItemCreateRequest.possibleFields = [
  "itemToMonitor",
         "monitoringMode",
         "requestedParameters"
];


exports.MonitoredItemCreateRequest = MonitoredItemCreateRequest;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("MonitoredItemCreateRequest",MonitoredItemCreateRequest);
