// node-opcua-service-subscription/_generated_/_auto_generated_MonitoredItemModifyRequest.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:18.616Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/MonitoredItemModifyRequest_schema").MonitoredItemModifyRequest_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const MonitoringParameters = require("./_auto_generated_MonitoringParameters").MonitoringParameters;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class MonitoredItemModifyRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function MonitoredItemModifyRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestedParameters =  null; /* new MonitoringParameters(null); */
        return ;
    }

    /**
      * 
      * @property monitoredItemId
      * @type {IntegerId}
      */
    self.monitoredItemId = initialize_field(schema.fields[0], options.monitoredItemId);

    /**
      * 
      * @property requestedParameters
      * @type {MonitoringParameters}
      */
    self.requestedParameters =  new MonitoringParameters( options.requestedParameters);

   // Object.preventExtensions(self);
}
util.inherits(MonitoredItemModifyRequest,BaseUAObject);
MonitoredItemModifyRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(757,0);
MonitoredItemModifyRequest.prototype.encodingDefaultXml = makeExpandedNodeId(756,0);
MonitoredItemModifyRequest.prototype._schema = schema;

const encode_IntegerId = _defaultTypeMap.IntegerId.encode;
const decode_IntegerId = _defaultTypeMap.IntegerId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
MonitoredItemModifyRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_IntegerId(this.monitoredItemId,stream);
   this.requestedParameters.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
MonitoredItemModifyRequest.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.monitoredItemId = decode_IntegerId(stream);
    this.requestedParameters.decode(stream);
};
MonitoredItemModifyRequest.possibleFields = [
  "monitoredItemId",
         "requestedParameters"
];


exports.MonitoredItemModifyRequest = MonitoredItemModifyRequest;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("MonitoredItemModifyRequest",MonitoredItemModifyRequest);
