// node-opcua-service-subscription/_generated_/_auto_generated_EventFilterResult.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:18.685Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/EventFilterResult_schema").EventFilterResult_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const DiagnosticInfo = getFactory("DiagnosticInfo");
const ContentFilterResult = require("./_auto_generated_ContentFilterResult").ContentFilterResult;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;
const MonitoringFilter = getFactory("MonitoringFilter");

/**
 * 
 * @class EventFilterResult
 * @constructor
 * @extends MonitoringFilter
 * @param  options {Object}
 */
function EventFilterResult(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    MonitoringFilter.call(this,options);
    if (options === null) { 
        MonitoringFilter.call(this,options);
        self.selectClauseDiagnosticInfos =  null; /* null array */
        self.whereClauseResult =  null; /* new ContentFilterResult(null); */
        return ;
    }

    /**
      * 
      * @property selectClauseResults
      * @type {StatusCode[]}
      */
    self.selectClauseResults = initialize_field_array(schema.fields[0], options.selectClauseResults);

    /**
      * 
      * @property selectClauseDiagnosticInfos
      * @type {DiagnosticInfo[]}
      */
    self.selectClauseDiagnosticInfos = [];
    if (options.selectClauseDiagnosticInfos) {
        assert(_.isArray(options.selectClauseDiagnosticInfos));
        self.selectClauseDiagnosticInfos = options.selectClauseDiagnosticInfos.map(function(e){ return new DiagnosticInfo(e); } );
    }

    /**
      * 
      * @property whereClauseResult
      * @type {ContentFilterResult}
      */
    self.whereClauseResult =  new ContentFilterResult( options.whereClauseResult);

   // Object.preventExtensions(self);
}
util.inherits(EventFilterResult,MonitoringFilter);
EventFilterResult.prototype.encodingDefaultBinary = makeExpandedNodeId(736,0);
EventFilterResult.prototype.encodingDefaultXml = makeExpandedNodeId(735,0);
EventFilterResult.prototype._schema = schema;

const encode_StatusCode = _defaultTypeMap.StatusCode.encode;
const decode_StatusCode = _defaultTypeMap.StatusCode.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
EventFilterResult.prototype.encode = function(stream,options) {
    // call base class implementation first
    MonitoringFilter.prototype.encode.call(this,stream,options);
    encodeArray(this.selectClauseResults, stream, encode_StatusCode);
    encodeArray(this.selectClauseDiagnosticInfos,stream,function(obj,stream){ obj.encode(stream,options); }); 
   this.whereClauseResult.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
EventFilterResult.prototype.decode = function(stream) {
    // call base class implementation first
    MonitoringFilter.prototype.decode.call(this,stream);
    this.selectClauseResults = decodeArray(stream, decode_StatusCode);
    this.selectClauseDiagnosticInfos = decodeArray(stream, function(stream) { 
       const obj = new DiagnosticInfo(null);
       obj.decode(stream);
       return obj; 
    });
    this.whereClauseResult.decode(stream);
};
EventFilterResult.possibleFields = [
  "selectClauseResults",
         "selectClauseDiagnosticInfos",
         "whereClauseResult"
];


exports.EventFilterResult = EventFilterResult;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("EventFilterResult",EventFilterResult);
