// node-opcua-service-subscription/_generated_/_auto_generated_SubscriptionAcknowledgement.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:18.538Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/SubscriptionAcknowledgement_schema").SubscriptionAcknowledgement_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class SubscriptionAcknowledgement
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function SubscriptionAcknowledgement(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property subscriptionId
      * @type {IntegerId}
      */
    self.subscriptionId = initialize_field(schema.fields[0], options.subscriptionId);

    /**
      * 
      * @property sequenceNumber
      * @type {Counter}
      */
    self.sequenceNumber = initialize_field(schema.fields[1], options.sequenceNumber);

   // Object.preventExtensions(self);
}
util.inherits(SubscriptionAcknowledgement,BaseUAObject);
SubscriptionAcknowledgement.prototype.encodingDefaultBinary = makeExpandedNodeId(823,0);
SubscriptionAcknowledgement.prototype.encodingDefaultXml = makeExpandedNodeId(822,0);
SubscriptionAcknowledgement.prototype._schema = schema;

const encode_IntegerId = _defaultTypeMap.IntegerId.encode;
const decode_IntegerId = _defaultTypeMap.IntegerId.decode;
const encode_Counter = _defaultTypeMap.Counter.encode;
const decode_Counter = _defaultTypeMap.Counter.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
SubscriptionAcknowledgement.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_IntegerId(this.subscriptionId,stream);
    encode_Counter(this.sequenceNumber,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
SubscriptionAcknowledgement.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.subscriptionId = decode_IntegerId(stream);
    this.sequenceNumber = decode_Counter(stream);
};
SubscriptionAcknowledgement.possibleFields = [
  "subscriptionId",
         "sequenceNumber"
];


exports.SubscriptionAcknowledgement = SubscriptionAcknowledgement;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("SubscriptionAcknowledgement",SubscriptionAcknowledgement);
