// node-opcua-service-subscription/_generated_/_auto_generated_SetPublishingModeRequest.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:18.581Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/SetPublishingModeRequest_schema").SetPublishingModeRequest_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const RequestHeader = getFactory("RequestHeader");
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class SetPublishingModeRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function SetPublishingModeRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestHeader =  null; /* new RequestHeader(null); */
        return ;
    }

    /**
      * 
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * 
      * @property publishingEnabled
      * @type {Boolean}
      */
    self.publishingEnabled = initialize_field(schema.fields[1], options.publishingEnabled);

    /**
      * 
      * @property subscriptionIds
      * @type {IntegerId[]}
      */
    self.subscriptionIds = initialize_field_array(schema.fields[2], options.subscriptionIds);

   // Object.preventExtensions(self);
}
util.inherits(SetPublishingModeRequest,BaseUAObject);
SetPublishingModeRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(799,0);
SetPublishingModeRequest.prototype.encodingDefaultXml = makeExpandedNodeId(798,0);
SetPublishingModeRequest.prototype._schema = schema;

const encode_Boolean = _defaultTypeMap.Boolean.encode;
const decode_Boolean = _defaultTypeMap.Boolean.decode;
const encode_IntegerId = _defaultTypeMap.IntegerId.encode;
const decode_IntegerId = _defaultTypeMap.IntegerId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
SetPublishingModeRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encode_Boolean(this.publishingEnabled,stream);
    encodeArray(this.subscriptionIds, stream, encode_IntegerId);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
SetPublishingModeRequest.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.requestHeader.decode(stream);
    this.publishingEnabled = decode_Boolean(stream);
    this.subscriptionIds = decodeArray(stream, decode_IntegerId);
};
SetPublishingModeRequest.possibleFields = [
  "requestHeader",
         "publishingEnabled",
         "subscriptionIds"
];


exports.SetPublishingModeRequest = SetPublishingModeRequest;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("SetPublishingModeRequest",SetPublishingModeRequest);
