// node-opcua-service-call/_generated_/_auto_generated_CallMethodResult.js
// --------- This code has been automatically generated !!! 2018-03-24T16:23:36.694Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/CallMethodResult_schema").CallMethodResult_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var DiagnosticInfo = getFactory("DiagnosticInfo");
var Variant = getFactory("Variant");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * The result of a Method call.
 * 
 * @class CallMethodResult
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.statusCode] {StatusCode} The StatusCode of the Method executed in the server.
 * @param  [options.inputArgumentResults] {StatusCode[]} The list of StatusCodes corresponding to the inputArguments.
 * @param  [options.inputArgumentDiagnosticInfos] {DiagnosticInfo[]} The list of diagnostic information corresponding to the inputArguments.
 * @param  [options.outputArguments] {Variant[]} The list of output argument values. 
 */
function CallMethodResult(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.inputArgumentDiagnosticInfos =  null; /* null array */
        self.outputArguments =  null; /* null array */
        return ;
    }

    /**
      * The StatusCode of the Method executed in the server.
      * @property statusCode
      * @type {StatusCode}
      */
    self.statusCode = initialize_field(schema.fields[0], options.statusCode);

    /**
      * The list of StatusCodes corresponding to the inputArguments.
      * @property inputArgumentResults
      * @type {StatusCode[]}
      */
    self.inputArgumentResults = initialize_field_array(schema.fields[1], options.inputArgumentResults);

    /**
      * The list of diagnostic information corresponding to the inputArguments.
      * @property inputArgumentDiagnosticInfos
      * @type {DiagnosticInfo[]}
      */
    self.inputArgumentDiagnosticInfos = [];
    if (options.inputArgumentDiagnosticInfos) {
        assert(_.isArray(options.inputArgumentDiagnosticInfos));
        self.inputArgumentDiagnosticInfos = options.inputArgumentDiagnosticInfos.map(function(e){ return new DiagnosticInfo(e); } );
    }

    /**
      * The list of output argument values. 
      * @property outputArguments
      * @type {Variant[]}
      */
    self.outputArguments = [];
    if (options.outputArguments) {
        assert(_.isArray(options.outputArguments));
        self.outputArguments = options.outputArguments.map(function(e){ return new Variant(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(CallMethodResult,BaseUAObject);
CallMethodResult.prototype.encodingDefaultBinary = makeExpandedNodeId(709,0);
CallMethodResult.prototype.encodingDefaultXml = makeExpandedNodeId(708,0);
CallMethodResult.prototype._schema = schema;

var encode_StatusCode = _defaultTypeMap.StatusCode.encode;
var decode_StatusCode = _defaultTypeMap.StatusCode.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
CallMethodResult.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_StatusCode(this.statusCode,stream);
    encodeArray(this.inputArgumentResults, stream, encode_StatusCode);
    encodeArray(this.inputArgumentDiagnosticInfos,stream,function(obj,stream){ obj.encode(stream,options); }); 
    encodeArray(this.outputArguments,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
CallMethodResult.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.statusCode = decode_StatusCode(stream,options);
    this.inputArgumentResults = decodeArray(stream, decode_StatusCode);
    this.inputArgumentDiagnosticInfos = decodeArray(stream, function(stream) { 
       var obj = new DiagnosticInfo(null);
       obj.decode(stream,options);
       return obj; 
    });
    this.outputArguments = decodeArray(stream, function(stream) { 
       var obj = new Variant(null);
       obj.decode(stream,options);
       return obj; 
    });
};
CallMethodResult.possibleFields = [
  "statusCode",
         "inputArgumentResults",
         "inputArgumentDiagnosticInfos",
         "outputArguments"
];


exports.CallMethodResult = CallMethodResult;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("CallMethodResult",CallMethodResult);
