// node-opcua-common/_generated_/_auto_generated_SessionSecurityDiagnostics.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:25.390Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/SessionSecurityDiagnostics_schema").SessionSecurityDiagnostics_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class SessionSecurityDiagnostics
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.sessionId] {NodeId} 
 * @param  [options.clientUserIdOfSession] {String} 
 * @param  [options.clientUserIdHistory] {String} 
 * @param  [options.authenticationMechanism] {String} 
 * @param  [options.encoding] {String} 
 * @param  [options.transportProtocol] {String} 
 * @param  [options.securityMode] {MessageSecurityMode} 
 * @param  [options.securityPolicyUri] {String} 
 * @param  [options.clientCertificate] {ByteString} 
 */
function SessionSecurityDiagnostics(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property sessionId
      * @type {NodeId}
      */
    self.sessionId = initialize_field(schema.fields[0], options.sessionId);

    /**
      * 
      * @property clientUserIdOfSession
      * @type {String}
      */
    self.clientUserIdOfSession = initialize_field(schema.fields[1], options.clientUserIdOfSession);

    /**
      * 
      * @property clientUserIdHistory
      * @type {String}
      */
    self.clientUserIdHistory = initialize_field(schema.fields[2], options.clientUserIdHistory);

    /**
      * 
      * @property authenticationMechanism
      * @type {String}
      */
    self.authenticationMechanism = initialize_field(schema.fields[3], options.authenticationMechanism);

    /**
      * 
      * @property encoding
      * @type {String}
      */
    self.encoding = initialize_field(schema.fields[4], options.encoding);

    /**
      * 
      * @property transportProtocol
      * @type {String}
      */
    self.transportProtocol = initialize_field(schema.fields[5], options.transportProtocol);

    /**
      * 
      * @property securityMode
      * @type {MessageSecurityMode}
      */
    self.setSecurityMode(initialize_field(schema.fields[6], options.securityMode));

    /**
      * 
      * @property securityPolicyUri
      * @type {String}
      */
    self.securityPolicyUri = initialize_field(schema.fields[7], options.securityPolicyUri);

    /**
      * 
      * @property clientCertificate
      * @type {ByteString}
      */
    self.clientCertificate = initialize_field(schema.fields[8], options.clientCertificate);

   // Object.preventExtensions(self);
}
util.inherits(SessionSecurityDiagnostics,BaseUAObject);

//## Define Enumeration setters
SessionSecurityDiagnostics.prototype.setSecurityMode = function(value) {
   var coercedValue = _enumerations.MessageSecurityMode.typedEnum.get(value);
   /* istanbul ignore next */
   if (coercedValue === undefined || coercedValue === null) {
      throw new Error("value cannot be coerced to MessageSecurityMode: " + value);
   }
   this.securityMode = coercedValue;
};
SessionSecurityDiagnostics.prototype.encodingDefaultBinary = makeExpandedNodeId(870,0);
SessionSecurityDiagnostics.prototype._schema = schema;

var encode_NodeId = _defaultTypeMap.NodeId.encode;
var decode_NodeId = _defaultTypeMap.NodeId.decode;
var encode_String = _defaultTypeMap.String.encode;
var decode_String = _defaultTypeMap.String.decode;
var encode_MessageSecurityMode = _enumerations.MessageSecurityMode.encode;
var decode_MessageSecurityMode = _enumerations.MessageSecurityMode.decode;
var encode_ByteString = _defaultTypeMap.ByteString.encode;
var decode_ByteString = _defaultTypeMap.ByteString.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
SessionSecurityDiagnostics.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_NodeId(this.sessionId,stream);
    encode_String(this.clientUserIdOfSession,stream);
    encode_String(this.clientUserIdHistory,stream);
    encode_String(this.authenticationMechanism,stream);
    encode_String(this.encoding,stream);
    encode_String(this.transportProtocol,stream);
    encode_MessageSecurityMode(this.securityMode,stream);
    encode_String(this.securityPolicyUri,stream);
    encode_ByteString(this.clientCertificate,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
SessionSecurityDiagnostics.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.sessionId = decode_NodeId(stream,options);
    this.clientUserIdOfSession = decode_String(stream,options);
    this.clientUserIdHistory = decode_String(stream,options);
    this.authenticationMechanism = decode_String(stream,options);
    this.encoding = decode_String(stream,options);
    this.transportProtocol = decode_String(stream,options);
    this.securityMode = decode_MessageSecurityMode(stream,options);
    this.securityPolicyUri = decode_String(stream,options);
    this.clientCertificate = decode_ByteString(stream,options);
};
SessionSecurityDiagnostics.possibleFields = [
  "sessionId",
         "clientUserIdOfSession",
         "clientUserIdHistory",
         "authenticationMechanism",
         "encoding",
         "transportProtocol",
         "securityMode",
         "securityPolicyUri",
         "clientCertificate"
];


exports.SessionSecurityDiagnostics = SessionSecurityDiagnostics;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("SessionSecurityDiagnostics",SessionSecurityDiagnostics);
