// node-opcua-common/_generated_/_auto_generated_ServiceCounter.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:25.258Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ServiceCounter_schema").ServiceCounter_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ServiceCounter
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.totalCount] {UInt32} 
 * @param  [options.errorCount] {UInt32} 
 */
function ServiceCounter(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property totalCount
      * @type {UInt32}
      */
    self.totalCount = initialize_field(schema.fields[0], options.totalCount);

    /**
      * 
      * @property errorCount
      * @type {UInt32}
      */
    self.errorCount = initialize_field(schema.fields[1], options.errorCount);

   // Object.preventExtensions(self);
}
util.inherits(ServiceCounter,BaseUAObject);
ServiceCounter.prototype.encodingDefaultBinary = makeExpandedNodeId(873,0);
ServiceCounter.prototype._schema = schema;

var encode_UInt32 = _defaultTypeMap.UInt32.encode;
var decode_UInt32 = _defaultTypeMap.UInt32.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ServiceCounter.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_UInt32(this.totalCount,stream);
    encode_UInt32(this.errorCount,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
ServiceCounter.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.totalCount = decode_UInt32(stream,options);
    this.errorCount = decode_UInt32(stream,options);
};
ServiceCounter.possibleFields = [
  "totalCount",
         "errorCount"
];


exports.ServiceCounter = ServiceCounter;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ServiceCounter",ServiceCounter);
