// node-opcua-common/_generated_/_auto_generated_ModelChangeStructure.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:25.107Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ModelChangeStructure_schema").ModelChangeStructure_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ModelChangeStructure
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.affected] {NodeId} 
 * @param  [options.affectedType] {NodeId} 
 * @param  [options.verb] {Byte} 
 */
function ModelChangeStructure(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property affected
      * @type {NodeId}
      */
    self.affected = initialize_field(schema.fields[0], options.affected);

    /**
      * 
      * @property affectedType
      * @type {NodeId}
      */
    self.affectedType = initialize_field(schema.fields[1], options.affectedType);

    /**
      * 
      * @property verb
      * @type {Byte}
      */
    self.verb = initialize_field(schema.fields[2], options.verb);

   // Object.preventExtensions(self);
}
util.inherits(ModelChangeStructure,BaseUAObject);
ModelChangeStructure.prototype.encodingDefaultBinary = makeExpandedNodeId(879,0);
ModelChangeStructure.prototype._schema = schema;

var encode_NodeId = _defaultTypeMap.NodeId.encode;
var decode_NodeId = _defaultTypeMap.NodeId.decode;
var encode_Byte = _defaultTypeMap.Byte.encode;
var decode_Byte = _defaultTypeMap.Byte.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ModelChangeStructure.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_NodeId(this.affected,stream);
    encode_NodeId(this.affectedType,stream);
    encode_Byte(this.verb,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
ModelChangeStructure.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.affected = decode_NodeId(stream,options);
    this.affectedType = decode_NodeId(stream,options);
    this.verb = decode_Byte(stream,options);
};
ModelChangeStructure.possibleFields = [
  "affected",
         "affectedType",
         "verb"
];


exports.ModelChangeStructure = ModelChangeStructure;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ModelChangeStructure",ModelChangeStructure);
