// node-opcua-common/_generated_/_auto_generated_ServerDiagnosticsSummary.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:25.250Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ServerDiagnosticsSummary_schema").ServerDiagnosticsSummary_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ServerDiagnosticsSummary
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.serverViewCount] {UInt32} 
 * @param  [options.currentSessionCount] {UInt32} 
 * @param  [options.cumulatedSessionCount] {UInt32} 
 * @param  [options.securityRejectedSessionCount] {UInt32} 
 * @param  [options.rejectedSessionCount] {UInt32} 
 * @param  [options.sessionTimeoutCount] {UInt32} 
 * @param  [options.sessionAbortCount] {UInt32} 
 * @param  [options.currentSubscriptionCount] {UInt32} 
 * @param  [options.cumulatedSubscriptionCount] {UInt32} 
 * @param  [options.publishingIntervalCount] {UInt32} 
 * @param  [options.securityRejectedRequestsCount] {UInt32} 
 * @param  [options.rejectedRequestsCount] {UInt32} 
 */
function ServerDiagnosticsSummary(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property serverViewCount
      * @type {UInt32}
      */
    self.serverViewCount = initialize_field(schema.fields[0], options.serverViewCount);

    /**
      * 
      * @property currentSessionCount
      * @type {UInt32}
      */
    self.currentSessionCount = initialize_field(schema.fields[1], options.currentSessionCount);

    /**
      * 
      * @property cumulatedSessionCount
      * @type {UInt32}
      */
    self.cumulatedSessionCount = initialize_field(schema.fields[2], options.cumulatedSessionCount);

    /**
      * 
      * @property securityRejectedSessionCount
      * @type {UInt32}
      */
    self.securityRejectedSessionCount = initialize_field(schema.fields[3], options.securityRejectedSessionCount);

    /**
      * 
      * @property rejectedSessionCount
      * @type {UInt32}
      */
    self.rejectedSessionCount = initialize_field(schema.fields[4], options.rejectedSessionCount);

    /**
      * 
      * @property sessionTimeoutCount
      * @type {UInt32}
      */
    self.sessionTimeoutCount = initialize_field(schema.fields[5], options.sessionTimeoutCount);

    /**
      * 
      * @property sessionAbortCount
      * @type {UInt32}
      */
    self.sessionAbortCount = initialize_field(schema.fields[6], options.sessionAbortCount);

    /**
      * 
      * @property currentSubscriptionCount
      * @type {UInt32}
      */
    self.currentSubscriptionCount = initialize_field(schema.fields[7], options.currentSubscriptionCount);

    /**
      * 
      * @property cumulatedSubscriptionCount
      * @type {UInt32}
      */
    self.cumulatedSubscriptionCount = initialize_field(schema.fields[8], options.cumulatedSubscriptionCount);

    /**
      * 
      * @property publishingIntervalCount
      * @type {UInt32}
      */
    self.publishingIntervalCount = initialize_field(schema.fields[9], options.publishingIntervalCount);

    /**
      * 
      * @property securityRejectedRequestsCount
      * @type {UInt32}
      */
    self.securityRejectedRequestsCount = initialize_field(schema.fields[10], options.securityRejectedRequestsCount);

    /**
      * 
      * @property rejectedRequestsCount
      * @type {UInt32}
      */
    self.rejectedRequestsCount = initialize_field(schema.fields[11], options.rejectedRequestsCount);

   // Object.preventExtensions(self);
}
util.inherits(ServerDiagnosticsSummary,BaseUAObject);
ServerDiagnosticsSummary.prototype.encodingDefaultBinary = makeExpandedNodeId(861,0);
ServerDiagnosticsSummary.prototype._schema = schema;

var encode_UInt32 = _defaultTypeMap.UInt32.encode;
var decode_UInt32 = _defaultTypeMap.UInt32.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ServerDiagnosticsSummary.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_UInt32(this.serverViewCount,stream);
    encode_UInt32(this.currentSessionCount,stream);
    encode_UInt32(this.cumulatedSessionCount,stream);
    encode_UInt32(this.securityRejectedSessionCount,stream);
    encode_UInt32(this.rejectedSessionCount,stream);
    encode_UInt32(this.sessionTimeoutCount,stream);
    encode_UInt32(this.sessionAbortCount,stream);
    encode_UInt32(this.currentSubscriptionCount,stream);
    encode_UInt32(this.cumulatedSubscriptionCount,stream);
    encode_UInt32(this.publishingIntervalCount,stream);
    encode_UInt32(this.securityRejectedRequestsCount,stream);
    encode_UInt32(this.rejectedRequestsCount,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
ServerDiagnosticsSummary.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.serverViewCount = decode_UInt32(stream,options);
    this.currentSessionCount = decode_UInt32(stream,options);
    this.cumulatedSessionCount = decode_UInt32(stream,options);
    this.securityRejectedSessionCount = decode_UInt32(stream,options);
    this.rejectedSessionCount = decode_UInt32(stream,options);
    this.sessionTimeoutCount = decode_UInt32(stream,options);
    this.sessionAbortCount = decode_UInt32(stream,options);
    this.currentSubscriptionCount = decode_UInt32(stream,options);
    this.cumulatedSubscriptionCount = decode_UInt32(stream,options);
    this.publishingIntervalCount = decode_UInt32(stream,options);
    this.securityRejectedRequestsCount = decode_UInt32(stream,options);
    this.rejectedRequestsCount = decode_UInt32(stream,options);
};
ServerDiagnosticsSummary.possibleFields = [
  "serverViewCount",
         "currentSessionCount",
         "cumulatedSessionCount",
         "securityRejectedSessionCount",
         "rejectedSessionCount",
         "sessionTimeoutCount",
         "sessionAbortCount",
         "currentSubscriptionCount",
         "cumulatedSubscriptionCount",
         "publishingIntervalCount",
         "securityRejectedRequestsCount",
         "rejectedRequestsCount"
];


exports.ServerDiagnosticsSummary = ServerDiagnosticsSummary;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ServerDiagnosticsSummary",ServerDiagnosticsSummary);
