// node-opcua-service-node-management/_generated_/_auto_generated_DeleteReferencesItem.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:11.720Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/DeleteReferencesItem_schema").DeleteReferencesItem_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class DeleteReferencesItem
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function DeleteReferencesItem(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * 
      * @property sourceNodeId
      * @type {NodeId}
      */
    self.sourceNodeId = initialize_field(schema.fields[0], options.sourceNodeId);

    /**
      * 
      * @property referenceNodeId
      * @type {NodeId}
      */
    self.referenceNodeId = initialize_field(schema.fields[1], options.referenceNodeId);

    /**
      * 
      * @property isForward
      * @type {Boolean}
      */
    self.isForward = initialize_field(schema.fields[2], options.isForward);

    /**
      * 
      * @property targetNodeId
      * @type {ExpandedNodeId}
      */
    self.targetNodeId = initialize_field(schema.fields[3], options.targetNodeId);

    /**
      * 
      * @property deleteBidirectional
      * @type {Boolean}
      */
    self.deleteBidirectional = initialize_field(schema.fields[4], options.deleteBidirectional);

   // Object.preventExtensions(self);
}
util.inherits(DeleteReferencesItem,BaseUAObject);
DeleteReferencesItem.prototype.encodingDefaultBinary = makeExpandedNodeId(387,0);
DeleteReferencesItem.prototype.encodingDefaultXml = makeExpandedNodeId(386,0);
DeleteReferencesItem.prototype._schema = schema;

const encode_NodeId = _defaultTypeMap.NodeId.encode;
const decode_NodeId = _defaultTypeMap.NodeId.decode;
const encode_Boolean = _defaultTypeMap.Boolean.encode;
const decode_Boolean = _defaultTypeMap.Boolean.decode;
const encode_ExpandedNodeId = _defaultTypeMap.ExpandedNodeId.encode;
const decode_ExpandedNodeId = _defaultTypeMap.ExpandedNodeId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
DeleteReferencesItem.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_NodeId(this.sourceNodeId,stream);
    encode_NodeId(this.referenceNodeId,stream);
    encode_Boolean(this.isForward,stream);
    encode_ExpandedNodeId(this.targetNodeId,stream);
    encode_Boolean(this.deleteBidirectional,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
DeleteReferencesItem.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.sourceNodeId = decode_NodeId(stream);
    this.referenceNodeId = decode_NodeId(stream);
    this.isForward = decode_Boolean(stream);
    this.targetNodeId = decode_ExpandedNodeId(stream);
    this.deleteBidirectional = decode_Boolean(stream);
};
DeleteReferencesItem.possibleFields = [
  "sourceNodeId",
         "referenceNodeId",
         "isForward",
         "targetNodeId",
         "deleteBidirectional"
];


exports.DeleteReferencesItem = DeleteReferencesItem;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("DeleteReferencesItem",DeleteReferencesItem);
