// node-opcua-service-discovery/_generated_/_auto_generated_RegisteredServer.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:13.529Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/RegisteredServer_schema").RegisteredServer_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const LocalizedText = getFactory("LocalizedText");
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * The information required to register a server with a discovery server.
 * 
 * @class RegisteredServer
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function RegisteredServer(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.serverNames =  null; /* null array */
        return ;
    }

    /**
      * The globally unique identifier for the server.
      * @property serverUri
      * @type {String}
      */
    self.serverUri = initialize_field(schema.fields[0], options.serverUri);

    /**
      * The globally unique identifier for the product.
      * @property productUri
      * @type {String}
      */
    self.productUri = initialize_field(schema.fields[1], options.productUri);

    /**
      * The name of server in multiple lcoales.
      * @property serverNames
      * @type {LocalizedText[]}
      */
    self.serverNames = [];
    if (options.serverNames) {
        assert(_.isArray(options.serverNames));
        self.serverNames = options.serverNames.map(function(e){ return new LocalizedText(e); } );
    }

    /**
      * The type of server.
      * @property serverType
      * @type {ApplicationType}
      */
    self.setServerType(initialize_field(schema.fields[3], options.serverType));

    /**
      * The globally unique identifier for the server that is acting as a gateway for the server.
      * @property gatewayServerUri
      * @type {String}
      */
    self.gatewayServerUri = initialize_field(schema.fields[4], options.gatewayServerUri);

    /**
      * The URLs for the server's discovery endpoints.
      * @property discoveryUrls
      * @type {String[]}
      */
    self.discoveryUrls = initialize_field_array(schema.fields[5], options.discoveryUrls);

    /**
      * A path to a file that is deleted when the server is no longer accepting connections.
      * @property semaphoreFilePath
      * @type {String}
      */
    self.semaphoreFilePath = initialize_field(schema.fields[6], options.semaphoreFilePath);

    /**
      * If FALSE the server will save the registration information to a persistent datastore.
      * @property isOnline
      * @type {Boolean}
      */
    self.isOnline = initialize_field(schema.fields[7], options.isOnline);

   // Object.preventExtensions(self);
}
util.inherits(RegisteredServer,BaseUAObject);

//## Define Enumeration setters
RegisteredServer.prototype.setServerType = function(value) {
   const coercedValue = _enumerations.ApplicationType.typedEnum.get(value);
   /* istanbul ignore next */
   if (coercedValue === undefined || coercedValue === null) {
      throw new Error("value cannot be coerced to ApplicationType: " + value);
   }
   this.serverType = coercedValue;
};
schema.id = generate_new_id();
RegisteredServer.prototype.encodingDefaultBinary = makeExpandedNodeId(schema.id);
RegisteredServer.prototype._schema = schema;

const encode_String = _defaultTypeMap.String.encode;
const decode_String = _defaultTypeMap.String.decode;
const encode_ApplicationType = _enumerations.ApplicationType.encode;
const decode_ApplicationType = _enumerations.ApplicationType.decode;
const encode_Boolean = _defaultTypeMap.Boolean.encode;
const decode_Boolean = _defaultTypeMap.Boolean.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
RegisteredServer.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_String(this.serverUri,stream);
    encode_String(this.productUri,stream);
    encodeArray(this.serverNames,stream,function(obj,stream){ obj.encode(stream,options); }); 
    encode_ApplicationType(this.serverType,stream);
    encode_String(this.gatewayServerUri,stream);
    encodeArray(this.discoveryUrls, stream, encode_String);
    encode_String(this.semaphoreFilePath,stream);
    encode_Boolean(this.isOnline,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
RegisteredServer.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.serverUri = decode_String(stream);
    this.productUri = decode_String(stream);
    this.serverNames = decodeArray(stream, function(stream) { 
       const obj = new LocalizedText(null);
       obj.decode(stream);
       return obj; 
    });
    this.serverType = decode_ApplicationType(stream);
    this.gatewayServerUri = decode_String(stream);
    this.discoveryUrls = decodeArray(stream, decode_String);
    this.semaphoreFilePath = decode_String(stream);
    this.isOnline = decode_Boolean(stream);
};
RegisteredServer.possibleFields = [
  "serverUri",
         "productUri",
         "serverNames",
         "serverType",
         "gatewayServerUri",
         "discoveryUrls",
         "semaphoreFilePath",
         "isOnline"
];


exports.RegisteredServer = RegisteredServer;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("RegisteredServer",RegisteredServer);
