// node-opcua-service-discovery/_generated_/_auto_generated_FindServersRequest.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:13.586Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/FindServersRequest_schema").FindServersRequest_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const RequestHeader = getFactory("RequestHeader");
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * Finds the servers known to the discovery server.
 * 
 * @class FindServersRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function FindServersRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestHeader =  null; /* new RequestHeader(null); */
        return ;
    }

    /**
      * A standard header included in all requests sent to a server.
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * The URL used by the client to send the request.
      * @property endpointUrl
      * @type {String}
      */
    self.endpointUrl = initialize_field(schema.fields[1], options.endpointUrl);

    /**
      * The locales to use when constructing a response.
      * @property localeIds
      * @type {LocaleId[]}
      */
    self.localeIds = initialize_field_array(schema.fields[2], options.localeIds);

    /**
      * The URIs of the servers to return (all servers returned if none specified).
      * @property serverUris
      * @type {String[]}
      */
    self.serverUris = initialize_field_array(schema.fields[3], options.serverUris);

   // Object.preventExtensions(self);
}
util.inherits(FindServersRequest,BaseUAObject);
FindServersRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(422,0);
FindServersRequest.prototype.encodingDefaultXml = makeExpandedNodeId(421,0);
FindServersRequest.prototype._schema = schema;

const encode_String = _defaultTypeMap.String.encode;
const decode_String = _defaultTypeMap.String.decode;
const encode_LocaleId = _defaultTypeMap.LocaleId.encode;
const decode_LocaleId = _defaultTypeMap.LocaleId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
FindServersRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encode_String(this.endpointUrl,stream);
    encodeArray(this.localeIds, stream, encode_LocaleId);
    encodeArray(this.serverUris, stream, encode_String);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
FindServersRequest.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.requestHeader.decode(stream);
    this.endpointUrl = decode_String(stream);
    this.localeIds = decodeArray(stream, decode_LocaleId);
    this.serverUris = decodeArray(stream, decode_String);
};
FindServersRequest.possibleFields = [
  "requestHeader",
         "endpointUrl",
         "localeIds",
         "serverUris"
];


exports.FindServersRequest = FindServersRequest;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("FindServersRequest",FindServersRequest);
