// node-opcua-service-discovery/_generated_/_auto_generated_FindServersOnNetworkResponse.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:13.609Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/FindServersOnNetworkResponse_schema").FindServersOnNetworkResponse_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const ResponseHeader = getFactory("ResponseHeader");
const ServerOnNetwork = require("./_auto_generated_ServerOnNetwork").ServerOnNetwork;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 *  A standard header included in all responses returned by servers.
 * 
 * @class FindServersOnNetworkResponse
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function FindServersOnNetworkResponse(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.responseHeader =  null; /* new ResponseHeader(null); */
        self.servers =  null; /* null array */
        return ;
    }

    /**
      * A standard header included in all responses returned by servers.
      * @property responseHeader
      * @type {ResponseHeader}
      */
    self.responseHeader =  new ResponseHeader( options.responseHeader);

    /**
      * The last time the counters were reset.
      * @property lastCounterResetTime
      * @type {UtcTime}
      */
    self.lastCounterResetTime = initialize_field(schema.fields[1], options.lastCounterResetTime);

    /**
      * The servers that met the criteria specified in the request.
      * @property servers
      * @type {ServerOnNetwork[]}
      */
    self.servers = [];
    if (options.servers) {
        assert(_.isArray(options.servers));
        self.servers = options.servers.map(function(e){ return new ServerOnNetwork(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(FindServersOnNetworkResponse,BaseUAObject);
FindServersOnNetworkResponse.prototype.encodingDefaultBinary = makeExpandedNodeId(12209,0);
FindServersOnNetworkResponse.prototype.encodingDefaultXml = makeExpandedNodeId(12197,0);
FindServersOnNetworkResponse.prototype._schema = schema;

const encode_UtcTime = _defaultTypeMap.UtcTime.encode;
const decode_UtcTime = _defaultTypeMap.UtcTime.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
FindServersOnNetworkResponse.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.responseHeader.encode(stream,options);
    encode_UtcTime(this.lastCounterResetTime,stream);
    encodeArray(this.servers,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
FindServersOnNetworkResponse.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.responseHeader.decode(stream);
    this.lastCounterResetTime = decode_UtcTime(stream);
    this.servers = decodeArray(stream, function(stream) { 
       const obj = new ServerOnNetwork(null);
       obj.decode(stream);
       return obj; 
    });
};
FindServersOnNetworkResponse.possibleFields = [
  "responseHeader",
         "lastCounterResetTime",
         "servers"
];


exports.FindServersOnNetworkResponse = FindServersOnNetworkResponse;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("FindServersOnNetworkResponse",FindServersOnNetworkResponse);
