// node-opcua-service-discovery/_generated_/_auto_generated_ServerOnNetwork.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:13.598Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/ServerOnNetwork_schema").ServerOnNetwork_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ServerOnNetwork
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function ServerOnNetwork(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * A unique identifier for the record.
This can be used to fetch the next batch of Servers in a subsequent
call to FindServersOnNetwork
      * @property recordId
      * @type {UInt32}
      */
    self.recordId = initialize_field(schema.fields[0], options.recordId);

    /**
      * The name of the Server specified in the mDNS announcement (see Part 12).This may be the same as the ApplicationName for the Server.
      * @property serverName
      * @type {String}
      */
    self.serverName = initialize_field(schema.fields[1], options.serverName);

    /**
      * The URL of the discovery Endpoint.
      * @property discoveryUrl
      * @type {String}
      */
    self.discoveryUrl = initialize_field(schema.fields[2], options.discoveryUrl);

    /**
      * The set of Server capabilities supported by the Server.The set of allowed Server capabilities are defined in Part 12
      * @property serverCapabilities
      * @type {String[]}
      */
    self.serverCapabilities = initialize_field_array(schema.fields[3], options.serverCapabilities);

   // Object.preventExtensions(self);
}
util.inherits(ServerOnNetwork,BaseUAObject);
ServerOnNetwork.prototype.encodingDefaultBinary = makeExpandedNodeId(12207,0);
ServerOnNetwork.prototype.encodingDefaultXml = makeExpandedNodeId(12195,0);
ServerOnNetwork.prototype._schema = schema;

const encode_UInt32 = _defaultTypeMap.UInt32.encode;
const decode_UInt32 = _defaultTypeMap.UInt32.decode;
const encode_String = _defaultTypeMap.String.encode;
const decode_String = _defaultTypeMap.String.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ServerOnNetwork.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_UInt32(this.recordId,stream);
    encode_String(this.serverName,stream);
    encode_String(this.discoveryUrl,stream);
    encodeArray(this.serverCapabilities, stream, encode_String);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
ServerOnNetwork.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.recordId = decode_UInt32(stream);
    this.serverName = decode_String(stream);
    this.discoveryUrl = decode_String(stream);
    this.serverCapabilities = decodeArray(stream, decode_String);
};
ServerOnNetwork.possibleFields = [
  "recordId",
         "serverName",
         "discoveryUrl",
         "serverCapabilities"
];


exports.ServerOnNetwork = ServerOnNetwork;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ServerOnNetwork",ServerOnNetwork);
