// node-opcua-service-browse/_generated_/_auto_generated_BrowseRequest.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:18.692Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/BrowseRequest_schema").BrowseRequest_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var RequestHeader = getFactory("RequestHeader");
var ViewDescription = require("./_auto_generated_ViewDescription").ViewDescription;
var BrowseDescription = require("./_auto_generated_BrowseDescription").BrowseDescription;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * Browse the references for one or more nodes from the server address space.
 * 
 * @class BrowseRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} A standard header included in all requests sent to a server.
 * @param  [options.view] {ViewDescription} The view to browse.
 * @param  [options.requestedMaxReferencesPerNode] {Counter} The maximum number of references to return in the response.
 * @param  [options.nodesToBrowse] {BrowseDescription[]} The list of nodes to browse.
 */
function BrowseRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestHeader =  null; /* new RequestHeader(null); */
        self.view =  null; /* new ViewDescription(null); */
        self.nodesToBrowse =  null; /* null array */
        return ;
    }

    /**
      * A standard header included in all requests sent to a server.
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * The view to browse.
      * @property view
      * @type {ViewDescription}
      */
    self.view =  new ViewDescription( options.view);

    /**
      * The maximum number of references to return in the response.
      * @property requestedMaxReferencesPerNode
      * @type {Counter}
      */
    self.requestedMaxReferencesPerNode = initialize_field(schema.fields[2], options.requestedMaxReferencesPerNode);

    /**
      * The list of nodes to browse.
      * @property nodesToBrowse
      * @type {BrowseDescription[]}
      */
    self.nodesToBrowse = [];
    if (options.nodesToBrowse) {
        assert(_.isArray(options.nodesToBrowse));
        self.nodesToBrowse = options.nodesToBrowse.map(function(e){ return new BrowseDescription(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(BrowseRequest,BaseUAObject);
BrowseRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(527,0);
BrowseRequest.prototype.encodingDefaultXml = makeExpandedNodeId(526,0);
BrowseRequest.prototype._schema = schema;

var encode_Counter = _defaultTypeMap.Counter.encode;
var decode_Counter = _defaultTypeMap.Counter.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
BrowseRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
   this.view.encode(stream,options);
    encode_Counter(this.requestedMaxReferencesPerNode,stream);
    encodeArray(this.nodesToBrowse,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
BrowseRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.view.decode(stream,options);
    this.requestedMaxReferencesPerNode = decode_Counter(stream,options);
    this.nodesToBrowse = decodeArray(stream, function(stream) { 
       var obj = new BrowseDescription(null);
       obj.decode(stream,options);
       return obj; 
    });
};
BrowseRequest.possibleFields = [
  "requestHeader",
         "view",
         "requestedMaxReferencesPerNode",
         "nodesToBrowse"
];


exports.BrowseRequest = BrowseRequest;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("BrowseRequest",BrowseRequest);
