// node-opcua-service-browse/_generated_/_auto_generated_ViewDescription.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:18.585Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ViewDescription_schema").ViewDescription_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * the view to browse.
 * 
 * @class ViewDescription
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.viewId] {NodeId} The node id of the view.
 * @param  [options.timestamp] {UtcTime} Browses the view at or before this time.
 * @param  [options.viewVersion] {UInt32} Browses a specific version of the view .
 */
function ViewDescription(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * The node id of the view.
      * @property viewId
      * @type {NodeId}
      */
    self.viewId = initialize_field(schema.fields[0], options.viewId);

    /**
      * Browses the view at or before this time.
      * @property timestamp
      * @type {UtcTime}
      */
    self.timestamp = initialize_field(schema.fields[1], options.timestamp);

    /**
      * Browses a specific version of the view .
      * @property viewVersion
      * @type {UInt32}
      */
    self.viewVersion = initialize_field(schema.fields[2], options.viewVersion);

   // Object.preventExtensions(self);
}
util.inherits(ViewDescription,BaseUAObject);
ViewDescription.prototype.encodingDefaultBinary = makeExpandedNodeId(513,0);
ViewDescription.prototype.encodingDefaultXml = makeExpandedNodeId(512,0);
ViewDescription.prototype._schema = schema;

var encode_NodeId = _defaultTypeMap.NodeId.encode;
var decode_NodeId = _defaultTypeMap.NodeId.decode;
var encode_UtcTime = _defaultTypeMap.UtcTime.encode;
var decode_UtcTime = _defaultTypeMap.UtcTime.decode;
var encode_UInt32 = _defaultTypeMap.UInt32.encode;
var decode_UInt32 = _defaultTypeMap.UInt32.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ViewDescription.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_NodeId(this.viewId,stream);
    encode_UtcTime(this.timestamp,stream);
    encode_UInt32(this.viewVersion,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
ViewDescription.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.viewId = decode_NodeId(stream,options);
    this.timestamp = decode_UtcTime(stream,options);
    this.viewVersion = decode_UInt32(stream,options);
};
ViewDescription.possibleFields = [
  "viewId",
         "timestamp",
         "viewVersion"
];


exports.ViewDescription = ViewDescription;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ViewDescription",ViewDescription);
