// node-opcua-service-endpoints/_generated_/_auto_generated_GetEndpointsResponse.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:09.869Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/GetEndpointsResponse_schema").GetEndpointsResponse_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const ResponseHeader = getFactory("ResponseHeader");
const EndpointDescription = require("./_auto_generated_EndpointDescription").EndpointDescription;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class GetEndpointsResponse
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function GetEndpointsResponse(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.responseHeader =  null; /* new ResponseHeader(null); */
        self.endpoints =  null; /* null array */
        return ;
    }

    /**
      * 
      * @property responseHeader
      * @type {ResponseHeader}
      */
    self.responseHeader =  new ResponseHeader( options.responseHeader);

    /**
      * 
      * @property endpoints
      * @type {EndpointDescription[]}
      */
    self.endpoints = [];
    if (options.endpoints) {
        assert(_.isArray(options.endpoints));
        self.endpoints = options.endpoints.map(function(e){ return new EndpointDescription(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(GetEndpointsResponse,BaseUAObject);
GetEndpointsResponse.prototype.encodingDefaultBinary = makeExpandedNodeId(431,0);
GetEndpointsResponse.prototype.encodingDefaultXml = makeExpandedNodeId(430,0);
GetEndpointsResponse.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
GetEndpointsResponse.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.responseHeader.encode(stream,options);
    encodeArray(this.endpoints,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
GetEndpointsResponse.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.responseHeader.decode(stream);
    this.endpoints = decodeArray(stream, function(stream) { 
       const obj = new EndpointDescription(null);
       obj.decode(stream);
       return obj; 
    });
};
GetEndpointsResponse.possibleFields = [
  "responseHeader",
         "endpoints"
];


exports.GetEndpointsResponse = GetEndpointsResponse;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("GetEndpointsResponse",GetEndpointsResponse);
