// @google-cloud/common-grpc/build/src/index.js
"use strict";
/*!
 * Copyright 2017 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var service_1 = require("./service");
var service_object_1 = require("./service-object");
var operation_1 = require("./operation");
var grpc = __importStar(require("grpc"));
exports.grpc = grpc;
var Service = service_1.GrpcService;
exports.Service = Service;
var ServiceObject = service_object_1.GrpcServiceObject;
exports.ServiceObject = ServiceObject;
var Operation = operation_1.GrpcOperation;
exports.Operation = Operation;
/*!
 * @module commonGrpc
 */
'use strict';
/**
 * @type {module:common/util}
 * @private
 */
var common_1 = require("@google-cloud/common");
exports.util = common_1.util;
/**
 * @type {module:common/paginator}
 * @private
 */
var common_2 = require("@google-cloud/common");
exports.paginator = common_2.paginator;
exports.Paginator = common_2.Paginator;
