// node-opcua-service-query/_generated_/_auto_generated_QueryNextResponse.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:20.464Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/QueryNextResponse_schema").QueryNextResponse_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const ResponseHeader = getFactory("ResponseHeader");
const QueryDataSet = require("./_auto_generated_QueryDataSet").QueryDataSet;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class QueryNextResponse
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function QueryNextResponse(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.responseHeader =  null; /* new ResponseHeader(null); */
        self.queryDataSet =  null; /* null array */
        return ;
    }

    /**
      * 
      * @property responseHeader
      * @type {ResponseHeader}
      */
    self.responseHeader =  new ResponseHeader( options.responseHeader);

    /**
      * 
      * @property queryDataSet
      * @type {QueryDataSet[]}
      */
    self.queryDataSet = [];
    if (options.queryDataSet) {
        assert(_.isArray(options.queryDataSet));
        self.queryDataSet = options.queryDataSet.map(function(e){ return new QueryDataSet(e); } );
    }

    /**
      * 
      * @property revisedContinuationPoint
      * @type {ContinuationPoint}
      */
    self.revisedContinuationPoint = initialize_field(schema.fields[2], options.revisedContinuationPoint);

   // Object.preventExtensions(self);
}
util.inherits(QueryNextResponse,BaseUAObject);
QueryNextResponse.prototype.encodingDefaultBinary = makeExpandedNodeId(624,0);
QueryNextResponse.prototype.encodingDefaultXml = makeExpandedNodeId(623,0);
QueryNextResponse.prototype._schema = schema;

const encode_ContinuationPoint = _defaultTypeMap.ContinuationPoint.encode;
const decode_ContinuationPoint = _defaultTypeMap.ContinuationPoint.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
QueryNextResponse.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.responseHeader.encode(stream,options);
    encodeArray(this.queryDataSet,stream,function(obj,stream){ obj.encode(stream,options); }); 
    encode_ContinuationPoint(this.revisedContinuationPoint,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
QueryNextResponse.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.responseHeader.decode(stream);
    this.queryDataSet = decodeArray(stream, function(stream) { 
       const obj = new QueryDataSet(null);
       obj.decode(stream);
       return obj; 
    });
    this.revisedContinuationPoint = decode_ContinuationPoint(stream);
};
QueryNextResponse.possibleFields = [
  "responseHeader",
         "queryDataSet",
         "revisedContinuationPoint"
];


exports.QueryNextResponse = QueryNextResponse;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("QueryNextResponse",QueryNextResponse);
