// node-opcua-service-query/_generated_/_auto_generated_QueryDataSet.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:20.291Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/QueryDataSet_schema").QueryDataSet_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const Variant = getFactory("Variant");
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * Data related to a Node returned in a Query response.
 * 
 * @class QueryDataSet
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function QueryDataSet(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.values =  null; /* null array */
        return ;
    }

    /**
      * 
      * @property nodeId
      * @type {ExpandedNodeId}
      */
    self.nodeId = initialize_field(schema.fields[0], options.nodeId);

    /**
      * 
      * @property typeDefinitionNode
      * @type {ExpandedNodeId}
      */
    self.typeDefinitionNode = initialize_field(schema.fields[1], options.typeDefinitionNode);

    /**
      * 
      * @property values
      * @type {Variant[]}
      */
    self.values = [];
    if (options.values) {
        assert(_.isArray(options.values));
        self.values = options.values.map(function(e){ return new Variant(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(QueryDataSet,BaseUAObject);
QueryDataSet.prototype.encodingDefaultBinary = makeExpandedNodeId(579,0);
QueryDataSet.prototype.encodingDefaultXml = makeExpandedNodeId(578,0);
QueryDataSet.prototype._schema = schema;

const encode_ExpandedNodeId = _defaultTypeMap.ExpandedNodeId.encode;
const decode_ExpandedNodeId = _defaultTypeMap.ExpandedNodeId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
QueryDataSet.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_ExpandedNodeId(this.nodeId,stream);
    encode_ExpandedNodeId(this.typeDefinitionNode,stream);
    encodeArray(this.values,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
QueryDataSet.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.nodeId = decode_ExpandedNodeId(stream);
    this.typeDefinitionNode = decode_ExpandedNodeId(stream);
    this.values = decodeArray(stream, function(stream) { 
       const obj = new Variant(null);
       obj.decode(stream);
       return obj; 
    });
};
QueryDataSet.possibleFields = [
  "nodeId",
         "typeDefinitionNode",
         "values"
];


exports.QueryDataSet = QueryDataSet;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("QueryDataSet",QueryDataSet);
