// node-opcua-service-query/_generated_/_auto_generated_QueryNextRequest.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:20.461Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/QueryNextRequest_schema").QueryNextRequest_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const RequestHeader = getFactory("RequestHeader");
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class QueryNextRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function QueryNextRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestHeader =  null; /* new RequestHeader(null); */
        return ;
    }

    /**
      * A standard header included in all requests sent to a server.
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * 
      * @property releaseContinuationPoint
      * @type {Boolean}
      */
    self.releaseContinuationPoint = initialize_field(schema.fields[1], options.releaseContinuationPoint);

    /**
      * 
      * @property continuationPoint
      * @type {ContinuationPoint}
      */
    self.continuationPoint = initialize_field(schema.fields[2], options.continuationPoint);

   // Object.preventExtensions(self);
}
util.inherits(QueryNextRequest,BaseUAObject);
QueryNextRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(621,0);
QueryNextRequest.prototype.encodingDefaultXml = makeExpandedNodeId(620,0);
QueryNextRequest.prototype._schema = schema;

const encode_Boolean = _defaultTypeMap.Boolean.encode;
const decode_Boolean = _defaultTypeMap.Boolean.decode;
const encode_ContinuationPoint = _defaultTypeMap.ContinuationPoint.encode;
const decode_ContinuationPoint = _defaultTypeMap.ContinuationPoint.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
QueryNextRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encode_Boolean(this.releaseContinuationPoint,stream);
    encode_ContinuationPoint(this.continuationPoint,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
QueryNextRequest.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.requestHeader.decode(stream);
    this.releaseContinuationPoint = decode_Boolean(stream);
    this.continuationPoint = decode_ContinuationPoint(stream);
};
QueryNextRequest.possibleFields = [
  "requestHeader",
         "releaseContinuationPoint",
         "continuationPoint"
];


exports.QueryNextRequest = QueryNextRequest;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("QueryNextRequest",QueryNextRequest);
