// node-opcua-service-call/_generated_/_auto_generated_CallRequest.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:09.825Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/CallRequest_schema").CallRequest_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const RequestHeader = getFactory("RequestHeader");
const CallMethodRequest = require("./_auto_generated_CallMethodRequest").CallMethodRequest;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * This Service is used to call (invoke) a list of Methods.
 * 
 * @class CallRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function CallRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestHeader =  null; /* new RequestHeader(null); */
        self.methodsToCall =  null; /* null array */
        return ;
    }

    /**
      * A standard header included in all requests sent to a server. 
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * List of Methods to call.
      * @property methodsToCall
      * @type {CallMethodRequest[]}
      */
    self.methodsToCall = [];
    if (options.methodsToCall) {
        assert(_.isArray(options.methodsToCall));
        self.methodsToCall = options.methodsToCall.map(function(e){ return new CallMethodRequest(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(CallRequest,BaseUAObject);
CallRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(712,0);
CallRequest.prototype.encodingDefaultXml = makeExpandedNodeId(711,0);
CallRequest.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
CallRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encodeArray(this.methodsToCall,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
CallRequest.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.requestHeader.decode(stream);
    this.methodsToCall = decodeArray(stream, function(stream) { 
       const obj = new CallMethodRequest(null);
       obj.decode(stream);
       return obj; 
    });
};
CallRequest.possibleFields = [
  "requestHeader",
         "methodsToCall"
];


exports.CallRequest = CallRequest;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("CallRequest",CallRequest);
