// node-opcua-service-call/_generated_/_auto_generated_CallMethodRequest.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:09.811Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/CallMethodRequest_schema").CallMethodRequest_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const Variant = getFactory("Variant");
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class CallMethodRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function CallMethodRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.inputArguments =  null; /* null array */
        return ;
    }

    /**
      * The NodeId shall be that of the Object or ObjectType that is the source of a HasComponent Reference (or subtype of HasComponent Reference) to the Method specified in methodId.
      * @property objectId
      * @type {NodeId}
      */
    self.objectId = initialize_field(schema.fields[0], options.objectId);

    /**
      * The NodeId of the Method to invoke.
      * @property methodId
      * @type {NodeId}
      */
    self.methodId = initialize_field(schema.fields[1], options.methodId);

    /**
      * The list of input argument values.
      * @property inputArguments
      * @type {Variant[]}
      */
    self.inputArguments = [];
    if (options.inputArguments) {
        assert(_.isArray(options.inputArguments));
        self.inputArguments = options.inputArguments.map(function(e){ return new Variant(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(CallMethodRequest,BaseUAObject);
CallMethodRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(706,0);
CallMethodRequest.prototype.encodingDefaultXml = makeExpandedNodeId(705,0);
CallMethodRequest.prototype._schema = schema;

const encode_NodeId = _defaultTypeMap.NodeId.encode;
const decode_NodeId = _defaultTypeMap.NodeId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
CallMethodRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_NodeId(this.objectId,stream);
    encode_NodeId(this.methodId,stream);
    encodeArray(this.inputArguments,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
CallMethodRequest.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.objectId = decode_NodeId(stream);
    this.methodId = decode_NodeId(stream);
    this.inputArguments = decodeArray(stream, function(stream) { 
       const obj = new Variant(null);
       obj.decode(stream);
       return obj; 
    });
};
CallMethodRequest.possibleFields = [
  "objectId",
         "methodId",
         "inputArguments"
];


exports.CallMethodRequest = CallMethodRequest;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("CallMethodRequest",CallMethodRequest);
