// node-opcua-service-translate-browse-path/_generated_/_auto_generated_TranslateBrowsePathsToNodeIdsRequest.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:21.706Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/TranslateBrowsePathsToNodeIdsRequest_schema").TranslateBrowsePathsToNodeIdsRequest_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var RequestHeader = getFactory("RequestHeader");
var BrowsePath = require("./_auto_generated_BrowsePath").BrowsePath;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class TranslateBrowsePathsToNodeIdsRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} 
 * @param  [options.browsePath] {BrowsePath[]} The list of paths to translate.
 */
function TranslateBrowsePathsToNodeIdsRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestHeader =  null; /* new RequestHeader(null); */
        self.browsePath =  null; /* null array */
        return ;
    }

    /**
      * 
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * The list of paths to translate.
      * @property browsePath
      * @type {BrowsePath[]}
      */
    self.browsePath = [];
    if (options.browsePath) {
        assert(_.isArray(options.browsePath));
        self.browsePath = options.browsePath.map(function(e){ return new BrowsePath(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(TranslateBrowsePathsToNodeIdsRequest,BaseUAObject);
TranslateBrowsePathsToNodeIdsRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(554,0);
TranslateBrowsePathsToNodeIdsRequest.prototype.encodingDefaultXml = makeExpandedNodeId(553,0);
TranslateBrowsePathsToNodeIdsRequest.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
TranslateBrowsePathsToNodeIdsRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encodeArray(this.browsePath,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
TranslateBrowsePathsToNodeIdsRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.browsePath = decodeArray(stream, function(stream) { 
       var obj = new BrowsePath(null);
       obj.decode(stream,options);
       return obj; 
    });
};
TranslateBrowsePathsToNodeIdsRequest.possibleFields = [
  "requestHeader",
         "browsePath"
];


exports.TranslateBrowsePathsToNodeIdsRequest = TranslateBrowsePathsToNodeIdsRequest;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("TranslateBrowsePathsToNodeIdsRequest",TranslateBrowsePathsToNodeIdsRequest);
