// node-opcua-service-translate-browse-path/_generated_/_auto_generated_RelativePathElement.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:21.591Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/RelativePathElement_schema").RelativePathElement_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var QualifiedName = getFactory("QualifiedName");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * An element in a relative path.
 * 
 * @class RelativePathElement
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.referenceTypeId] {NodeId} The type of reference to follow.
 * @param  [options.isInverse] {Boolean} If TRUE the reverse reference is followed.
 * @param  [options.includeSubtypes] {Boolean} If TRUE then subtypes of the reference type are followed.
 * @param  [options.targetName] {QualifiedName} The browse name of the target.
 */
function RelativePathElement(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.targetName =  null; /* new QualifiedName(null); */
        return ;
    }

    /**
      * The type of reference to follow.
      * @property referenceTypeId
      * @type {NodeId}
      */
    self.referenceTypeId = initialize_field(schema.fields[0], options.referenceTypeId);

    /**
      * If TRUE the reverse reference is followed.
      * @property isInverse
      * @type {Boolean}
      */
    self.isInverse = initialize_field(schema.fields[1], options.isInverse);

    /**
      * If TRUE then subtypes of the reference type are followed.
      * @property includeSubtypes
      * @type {Boolean}
      */
    self.includeSubtypes = initialize_field(schema.fields[2], options.includeSubtypes);

    /**
      * The browse name of the target.
      * @property targetName
      * @type {QualifiedName}
      */
    self.targetName =  new QualifiedName( options.targetName);

   // Object.preventExtensions(self);
}
util.inherits(RelativePathElement,BaseUAObject);
RelativePathElement.prototype.encodingDefaultBinary = makeExpandedNodeId(539,0);
RelativePathElement.prototype.encodingDefaultXml = makeExpandedNodeId(538,0);
RelativePathElement.prototype._schema = schema;

var encode_NodeId = _defaultTypeMap.NodeId.encode;
var decode_NodeId = _defaultTypeMap.NodeId.decode;
var encode_Boolean = _defaultTypeMap.Boolean.encode;
var decode_Boolean = _defaultTypeMap.Boolean.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
RelativePathElement.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_NodeId(this.referenceTypeId,stream);
    encode_Boolean(this.isInverse,stream);
    encode_Boolean(this.includeSubtypes,stream);
   this.targetName.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
RelativePathElement.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.referenceTypeId = decode_NodeId(stream,options);
    this.isInverse = decode_Boolean(stream,options);
    this.includeSubtypes = decode_Boolean(stream,options);
    this.targetName.decode(stream,options);
};
RelativePathElement.possibleFields = [
  "referenceTypeId",
         "isInverse",
         "includeSubtypes",
         "targetName"
];


exports.RelativePathElement = RelativePathElement;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("RelativePathElement",RelativePathElement);
