// node-opcua-service-browse/_generated_/_auto_generated_BrowseResult.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:09.728Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/BrowseResult_schema").BrowseResult_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const ReferenceDescription = require("./_auto_generated_ReferenceDescription").ReferenceDescription;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * The result of a browse operation.
 * 
 * @class BrowseResult
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 */
function BrowseResult(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.references =  null; /* null array */
        return ;
    }

    /**
      * A code indicating any error during the operation.
      * @property statusCode
      * @type {StatusCode}
      */
    self.statusCode = initialize_field(schema.fields[0], options.statusCode);

    /**
      * A value that indicates the operation is incomplete and can be continued by calling BrowseNext.
      * @property continuationPoint
      * @type {ContinuationPoint}
      * @default  null
      */
    self.continuationPoint = initialize_field(schema.fields[1], options.continuationPoint);

    /**
      * A list of references that meet the criteria specified in the request.
      * @property references
      * @type {ReferenceDescription[]}
      */
    self.references = [];
    if (options.references) {
        assert(_.isArray(options.references));
        self.references = options.references.map(function(e){ return new ReferenceDescription(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(BrowseResult,BaseUAObject);
schema.id = generate_new_id();
BrowseResult.prototype.encodingDefaultBinary = makeExpandedNodeId(schema.id);
BrowseResult.prototype._schema = schema;

const encode_StatusCode = _defaultTypeMap.StatusCode.encode;
const decode_StatusCode = _defaultTypeMap.StatusCode.decode;
const encode_ContinuationPoint = _defaultTypeMap.ContinuationPoint.encode;
const decode_ContinuationPoint = _defaultTypeMap.ContinuationPoint.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
BrowseResult.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_StatusCode(this.statusCode,stream);
    encode_ContinuationPoint(this.continuationPoint,stream);
    encodeArray(this.references,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
BrowseResult.prototype.decode = function(stream) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream);
    this.statusCode = decode_StatusCode(stream);
    this.continuationPoint = decode_ContinuationPoint(stream);
    this.references = decodeArray(stream, function(stream) { 
       const obj = new ReferenceDescription(null);
       obj.decode(stream);
       return obj; 
    });
};
BrowseResult.possibleFields = [
  "statusCode",
         "continuationPoint",
         "references"
];


exports.BrowseResult = BrowseResult;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("BrowseResult",BrowseResult);
