// node-opcua-service-node-management/_generated_/_auto_generated_AddNodesItem.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:21.693Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/AddNodesItem_schema").AddNodesItem_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var QualifiedName = getFactory("QualifiedName");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class AddNodesItem
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.parentNodeId] {ExpandedNodeId} ExpandedNodeId of the parent Node for the Reference.
 * @param  [options.referenceTypeId] {NodeId} NodeId of the hierarchical ReferenceType to use for the Reference from the parent Node to the new Node.
 * @param  [options.requestedNewNodeId] {ExpandedNodeId} 
 * @param  [options.browseName] {QualifiedName} 
 * @param  [options.nodeClass] {NodeClass} 
 * @param  [options.nodeAttributes] {ExtensionObject} 
 * @param  [options.typeDefinition] {ExpandedNodeId} 
 */
function AddNodesItem(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.browseName =  null; /* new QualifiedName(null); */
        return ;
    }

    /**
      * ExpandedNodeId of the parent Node for the Reference.
      * @property parentNodeId
      * @type {ExpandedNodeId}
      */
    self.parentNodeId = initialize_field(schema.fields[0], options.parentNodeId);

    /**
      * NodeId of the hierarchical ReferenceType to use for the Reference from the parent Node to the new Node.
      * @property referenceTypeId
      * @type {NodeId}
      */
    self.referenceTypeId = initialize_field(schema.fields[1], options.referenceTypeId);

    /**
      * 
      * @property requestedNewNodeId
      * @type {ExpandedNodeId}
      */
    self.requestedNewNodeId = initialize_field(schema.fields[2], options.requestedNewNodeId);

    /**
      * 
      * @property browseName
      * @type {QualifiedName}
      */
    self.browseName =  new QualifiedName( options.browseName);

    /**
      * 
      * @property nodeClass
      * @type {NodeClass}
      */
    self.setNodeClass(initialize_field(schema.fields[4], options.nodeClass));

    /**
      * 
      * @property nodeAttributes
      * @type {ExtensionObject}
      */
    self.nodeAttributes = initialize_field(schema.fields[5], options.nodeAttributes);

    /**
      * 
      * @property typeDefinition
      * @type {ExpandedNodeId}
      */
    self.typeDefinition = initialize_field(schema.fields[6], options.typeDefinition);

   // Object.preventExtensions(self);
}
util.inherits(AddNodesItem,BaseUAObject);

//## Define Enumeration setters
AddNodesItem.prototype.setNodeClass = function(value) {
   var coercedValue = _enumerations.NodeClass.typedEnum.get(value);
   /* istanbul ignore next */
   if (coercedValue === undefined || coercedValue === null) {
      throw new Error("value cannot be coerced to NodeClass: " + value);
   }
   this.nodeClass = coercedValue;
};
AddNodesItem.prototype.encodingDefaultBinary = makeExpandedNodeId(378,0);
AddNodesItem.prototype.encodingDefaultXml = makeExpandedNodeId(377,0);
AddNodesItem.prototype._schema = schema;

var encode_ExpandedNodeId = _defaultTypeMap.ExpandedNodeId.encode;
var decode_ExpandedNodeId = _defaultTypeMap.ExpandedNodeId.decode;
var encode_NodeId = _defaultTypeMap.NodeId.encode;
var decode_NodeId = _defaultTypeMap.NodeId.decode;
var encode_NodeClass = _enumerations.NodeClass.encode;
var decode_NodeClass = _enumerations.NodeClass.decode;
var encode_ExtensionObject = _defaultTypeMap.ExtensionObject.encode;
var decode_ExtensionObject = _defaultTypeMap.ExtensionObject.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
AddNodesItem.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_ExpandedNodeId(this.parentNodeId,stream);
    encode_NodeId(this.referenceTypeId,stream);
    encode_ExpandedNodeId(this.requestedNewNodeId,stream);
   this.browseName.encode(stream,options);
    encode_NodeClass(this.nodeClass,stream);
    encode_ExtensionObject(this.nodeAttributes,stream);
    encode_ExpandedNodeId(this.typeDefinition,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
AddNodesItem.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.parentNodeId = decode_ExpandedNodeId(stream,options);
    this.referenceTypeId = decode_NodeId(stream,options);
    this.requestedNewNodeId = decode_ExpandedNodeId(stream,options);
    this.browseName.decode(stream,options);
    this.nodeClass = decode_NodeClass(stream,options);
    this.nodeAttributes = decode_ExtensionObject(stream,options);
    this.typeDefinition = decode_ExpandedNodeId(stream,options);
};
AddNodesItem.possibleFields = [
  "parentNodeId",
         "referenceTypeId",
         "requestedNewNodeId",
         "browseName",
         "nodeClass",
         "nodeAttributes",
         "typeDefinition"
];


exports.AddNodesItem = AddNodesItem;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("AddNodesItem",AddNodesItem);
