// node-opcua-service-node-management/_generated_/_auto_generated_AddNodesRequest.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:21.787Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/AddNodesRequest_schema").AddNodesRequest_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var RequestHeader = getFactory("RequestHeader");
var AddNodesItem = require("./_auto_generated_AddNodesItem").AddNodesItem;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class AddNodesRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} A standard header included in all requests sent to a server.
 * @param  [options.nodesToAdd] {AddNodesItem[]}  
 */
function AddNodesRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestHeader =  null; /* new RequestHeader(null); */
        self.nodesToAdd =  null; /* null array */
        return ;
    }

    /**
      * A standard header included in all requests sent to a server.
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      *  
      * @property nodesToAdd
      * @type {AddNodesItem[]}
      */
    self.nodesToAdd = [];
    if (options.nodesToAdd) {
        assert(_.isArray(options.nodesToAdd));
        self.nodesToAdd = options.nodesToAdd.map(function(e){ return new AddNodesItem(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(AddNodesRequest,BaseUAObject);
AddNodesRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(488,0);
AddNodesRequest.prototype.encodingDefaultXml = makeExpandedNodeId(487,0);
AddNodesRequest.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
AddNodesRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
    encodeArray(this.nodesToAdd,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
AddNodesRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.nodesToAdd = decodeArray(stream, function(stream) { 
       var obj = new AddNodesItem(null);
       obj.decode(stream,options);
       return obj; 
    });
};
AddNodesRequest.possibleFields = [
  "requestHeader",
         "nodesToAdd"
];


exports.AddNodesRequest = AddNodesRequest;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("AddNodesRequest",AddNodesRequest);
