// azure-iot-common/lib/errors.js
/*! Copyright (c) Microsoft. All rights reserved.
 *! Licensed under the MIT license. See LICENSE file in the project root for full license information.
 */
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
/**
 * Error thrown when an argument is invalid.
 *
 * @augments {Error}
 */
var ArgumentError = (function (_super) {
    __extends(ArgumentError, _super);
    function ArgumentError(message) {
        _super.call(this, message);
        this.name = 'ArgumentError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return ArgumentError;
}(Error));
exports.ArgumentError = ArgumentError;
/**
 * Error thrown when an argument has a value that is out of the admissible range.
 *
 * @augments {Error}
 */
var ArgumentOutOfRangeError = (function (_super) {
    __extends(ArgumentOutOfRangeError, _super);
    function ArgumentOutOfRangeError(message) {
        _super.call(this, message);
        this.name = 'ArgumentOutOfRangeError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return ArgumentOutOfRangeError;
}(Error));
exports.ArgumentOutOfRangeError = ArgumentOutOfRangeError;
/**
 * Error thrown when the message queue for a device is full.
 *
 * @augments {Error}
 */
var DeviceMaximumQueueDepthExceededError = (function (_super) {
    __extends(DeviceMaximumQueueDepthExceededError, _super);
    function DeviceMaximumQueueDepthExceededError(message) {
        _super.call(this, message);
        this.name = 'DeviceMaximumQueueDepthExceededError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return DeviceMaximumQueueDepthExceededError;
}(Error));
exports.DeviceMaximumQueueDepthExceededError = DeviceMaximumQueueDepthExceededError;
/**
 * Error thrown when a device cannot be found in the IoT Hub instance registry.
 *
 * @augments {Error}
 */
var DeviceNotFoundError = (function (_super) {
    __extends(DeviceNotFoundError, _super);
    function DeviceNotFoundError(message) {
        _super.call(this, message);
        this.name = 'DeviceNotFoundError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return DeviceNotFoundError;
}(Error));
exports.DeviceNotFoundError = DeviceNotFoundError;
/**
 * Error thrown when a string that is supposed to have a specific formatting is not formatted properly.
 *
 * @augments {Error}
 */
var FormatError = (function (_super) {
    __extends(FormatError, _super);
    function FormatError(message) {
        _super.call(this, message);
        this.name = 'FormatError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return FormatError;
}(Error));
exports.FormatError = FormatError;
/**
 * Error thrown when the connection parameters are wrong and the server refused the connection.
 *
 * @augments {Error}
 */
var UnauthorizedError = (function (_super) {
    __extends(UnauthorizedError, _super);
    function UnauthorizedError(message) {
        _super.call(this, message);
        this.name = 'UnauthorizedError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return UnauthorizedError;
}(Error));
exports.UnauthorizedError = UnauthorizedError;
/**
 * Error thrown when a feature is not implemented yet but the placeholder is present.
 *
 * @augments {Error}
 */
var NotImplementedError = (function (_super) {
    __extends(NotImplementedError, _super);
    function NotImplementedError(message) {
        _super.call(this, message);
        this.name = 'NotImplementedError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return NotImplementedError;
}(Error));
exports.NotImplementedError = NotImplementedError;
/**
 * Error thrown when the device is disconnected and the operation cannot be completed.
 *
 * @augments {Error}
 */
var NotConnectedError = (function (_super) {
    __extends(NotConnectedError, _super);
    function NotConnectedError(message) {
        _super.call(this, message);
        this.name = 'NotConnectedError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return NotConnectedError;
}(Error));
exports.NotConnectedError = NotConnectedError;
/**
 * Error thrown the the Azure IoT hub quota has been exceeded. Quotas are reset periodically, this operation will have to wait until then.
 * To learn more about quotas, see {@link https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-quotas-throttling|Azure IoT Hub quotas and throttling}
 *
 * @augments {Error}
 */
var IotHubQuotaExceededError = (function (_super) {
    __extends(IotHubQuotaExceededError, _super);
    function IotHubQuotaExceededError(message) {
        _super.call(this, message);
        this.name = 'IotHubQuotaExceededError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return IotHubQuotaExceededError;
}(Error));
exports.IotHubQuotaExceededError = IotHubQuotaExceededError;
/**
 * Error thrown when the message sent is too large: the maximum size is 256Kb.
 *
 * @augments {Error}
 */
var MessageTooLargeError = (function (_super) {
    __extends(MessageTooLargeError, _super);
    function MessageTooLargeError(message) {
        _super.call(this, message);
        this.name = 'MessageTooLargeError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return MessageTooLargeError;
}(Error));
exports.MessageTooLargeError = MessageTooLargeError;
/**
 * Error thrown when an internal server error occured. You may have found a bug?
 *
 * @augments {Error}
 */
var InternalServerError = (function (_super) {
    __extends(InternalServerError, _super);
    function InternalServerError(message) {
        _super.call(this, message);
        this.name = 'InternalServerError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return InternalServerError;
}(Error));
exports.InternalServerError = InternalServerError;
/**
 * Error thrown when the service is unavailable. The operation should be retried.
 *
 * @augments {Error}
 */
var ServiceUnavailableError = (function (_super) {
    __extends(ServiceUnavailableError, _super);
    function ServiceUnavailableError(message) {
        _super.call(this, message);
        this.name = 'ServiceUnavailableError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return ServiceUnavailableError;
}(Error));
exports.ServiceUnavailableError = ServiceUnavailableError;
/**
 * Error thrown when the Azure IoT hub was not found.
 *
 * @augments {Error}
 */
var IotHubNotFoundError = (function (_super) {
    __extends(IotHubNotFoundError, _super);
    function IotHubNotFoundError(message) {
        _super.call(this, message);
        this.name = 'IotHubNotFoundError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return IotHubNotFoundError;
}(Error));
exports.IotHubNotFoundError = IotHubNotFoundError;
/**
 * Error thrown when IoT Hub has been suspended.
 *
 * @augments {Error}
 */
var IoTHubSuspendedError = (function (_super) {
    __extends(IoTHubSuspendedError, _super);
    function IoTHubSuspendedError(message) {
        _super.call(this, message);
        this.name = 'IoTHubSuspendedError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return IoTHubSuspendedError;
}(Error));
exports.IoTHubSuspendedError = IoTHubSuspendedError;
/**
 * Error thrown when the job with the specified identifier was not found.
 *
 * @augments {Error}
 */
var JobNotFoundError = (function (_super) {
    __extends(JobNotFoundError, _super);
    function JobNotFoundError(message) {
        _super.call(this, message);
        this.name = 'JobNotFoundError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return JobNotFoundError;
}(Error));
exports.JobNotFoundError = JobNotFoundError;
/**
 * Error thrown when the maximum number of devices on a specific hub has been reached.
 *
 * @augments {Error}
 */
var TooManyDevicesError = (function (_super) {
    __extends(TooManyDevicesError, _super);
    function TooManyDevicesError(message) {
        _super.call(this, message);
        this.name = 'TooManyDevicesError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return TooManyDevicesError;
}(Error));
exports.TooManyDevicesError = TooManyDevicesError;
/**
 * Error thrown when IoT Hub is throttled due to excessive activity.
 * To learn more about quotas, see {@link https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-quotas-throttling|Azure IoT Hub quotas and throttling}
 *
 * @augments {Error}
 */
var ThrottlingError = (function (_super) {
    __extends(ThrottlingError, _super);
    function ThrottlingError(message) {
        _super.call(this, message);
        this.name = 'ThrottlingError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return ThrottlingError;
}(Error));
exports.ThrottlingError = ThrottlingError;
/**
 * Error thrown when the device id used for device creation already exists in the Device Identity Registry.
 *
 * @augments {Error}
 */
var DeviceAlreadyExistsError = (function (_super) {
    __extends(DeviceAlreadyExistsError, _super);
    function DeviceAlreadyExistsError(message) {
        _super.call(this, message);
        this.name = 'DeviceAlreadyExistsError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return DeviceAlreadyExistsError;
}(Error));
exports.DeviceAlreadyExistsError = DeviceAlreadyExistsError;
/**
 * Error thrown when settling a message fails because the lock token associated with the message is lost.
 *
 * @augments {Error}
 */
var DeviceMessageLockLostError = (function (_super) {
    __extends(DeviceMessageLockLostError, _super);
    function DeviceMessageLockLostError(message) {
        _super.call(this, message);
        this.name = 'DeviceMessageLockLostError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return DeviceMessageLockLostError;
}(Error));
exports.DeviceMessageLockLostError = DeviceMessageLockLostError;
/**
 * Error thrown when the eTag specified is incorrectly formatted or out of date.
 *
 * @augments {Error}
 */
var InvalidEtagError = (function (_super) {
    __extends(InvalidEtagError, _super);
    function InvalidEtagError(message) {
        _super.call(this, message);
        this.name = 'InvalidEtagError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return InvalidEtagError;
}(Error));
exports.InvalidEtagError = InvalidEtagError;
/**
 * Error thrown when an operation is attempted but is not allowed.
 *
 * @augments {Error}
 */
var InvalidOperationError = (function (_super) {
    __extends(InvalidOperationError, _super);
    function InvalidOperationError(message) {
        _super.call(this, message);
        this.name = 'InvalidOperationError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return InvalidOperationError;
}(Error));
exports.InvalidOperationError = InvalidOperationError;
/**
 * Error thrown when a condition that should have been met in order to execute an operation was not.
 *
 * @augments {Error}
 */
var PreconditionFailedError = (function (_super) {
    __extends(PreconditionFailedError, _super);
    function PreconditionFailedError(message) {
        _super.call(this, message);
        this.name = 'PreconditionFailedError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return PreconditionFailedError;
}(Error));
exports.PreconditionFailedError = PreconditionFailedError;
/**
 * Error thrown when a timeout occurs
 *
 * @augments {Error}
 */
var TimeoutError = (function (_super) {
    __extends(TimeoutError, _super);
    function TimeoutError(message) {
        _super.call(this, message);
        this.name = 'TimeoutError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return TimeoutError;
}(Error));
exports.TimeoutError = TimeoutError;
/**
 * Error thrown when a device sends a bad response to a device method call.
 *
 * @augments {Error}
 */
var BadDeviceResponseError = (function (_super) {
    __extends(BadDeviceResponseError, _super);
    function BadDeviceResponseError(message) {
        _super.call(this, message);
        this.name = 'BadDeviceResponseError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return BadDeviceResponseError;
}(Error));
exports.BadDeviceResponseError = BadDeviceResponseError;
/**
 * Error thrown when the IoT Hub instance doesn't process the device method call in time.
 *
 * @augments {Error}
 */
var GatewayTimeoutError = (function (_super) {
    __extends(GatewayTimeoutError, _super);
    function GatewayTimeoutError(message) {
        _super.call(this, message);
        this.name = 'GatewayTimeoutError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return GatewayTimeoutError;
}(Error));
exports.GatewayTimeoutError = GatewayTimeoutError;
/**
 * Error thrown when the device doesn't process the method call in time.
 *
 * @augments {Error}
 */
var DeviceTimeoutError = (function (_super) {
    __extends(DeviceTimeoutError, _super);
    function DeviceTimeoutError(message) {
        _super.call(this, message);
        this.name = 'DeviceTimeoutError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return DeviceTimeoutError;
}(Error));
exports.DeviceTimeoutError = DeviceTimeoutError;
/**
 * Error thrown when the c2d feature stopped working at the transport level, requiring the client to retry starting it.
 *
 * @augments {Error}
 */
var CloudToDeviceDetachedError = (function (_super) {
    __extends(CloudToDeviceDetachedError, _super);
    function CloudToDeviceDetachedError(message) {
        _super.call(this, message);
        this.name = 'CloudToDeviceDetachedError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return CloudToDeviceDetachedError;
}(Error));
exports.CloudToDeviceDetachedError = CloudToDeviceDetachedError;
/**
 * Error thrown when the device methods feature stopped working at the transport level, requiring the client to retry starting it.
 *
 * @augments {Error}
 */
var DeviceMethodsDetachedError = (function (_super) {
    __extends(DeviceMethodsDetachedError, _super);
    function DeviceMethodsDetachedError(message) {
        _super.call(this, message);
        this.name = 'DeviceMethodsDetachedError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return DeviceMethodsDetachedError;
}(Error));
exports.DeviceMethodsDetachedError = DeviceMethodsDetachedError;
/**
 * Error thrown when the twin feature stopped working at the transport level, requiring the client to retry starting it.
 *
 * @augments {Error}
 */
var TwinDetachedError = (function (_super) {
    __extends(TwinDetachedError, _super);
    function TwinDetachedError(message) {
        _super.call(this, message);
        this.name = 'TwinDetachedError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return TwinDetachedError;
}(Error));
exports.TwinDetachedError = TwinDetachedError;
/**
 * Generic error thrown when a twin request fails with an unknown error code.
 *
 * @augments {Error}
 */
var TwinRequestError = (function (_super) {
    __extends(TwinRequestError, _super);
    function TwinRequestError(message) {
        _super.call(this, message);
        this.name = 'TwinRequestError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return TwinRequestError;
}(Error));
exports.TwinRequestError = TwinRequestError;
/**
 * Error thrown when any operation (local or remote) is cancelled
 *
 * @augments {Error}
 */
var OperationCancelledError = (function (_super) {
    __extends(OperationCancelledError, _super);
    function OperationCancelledError(message) {
        _super.call(this, message);
        this.name = 'OperationCancelledError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return OperationCancelledError;
}(Error));
exports.OperationCancelledError = OperationCancelledError;
/**
 * Error thrown when a DPS registration operation fails
 *
 * @augments {Error}
 */
var DeviceRegistrationFailedError = (function (_super) {
    __extends(DeviceRegistrationFailedError, _super);
    function DeviceRegistrationFailedError(message) {
        _super.call(this, message);
        this.name = 'ProvisioningRegistrationFailedError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return DeviceRegistrationFailedError;
}(Error));
exports.DeviceRegistrationFailedError = DeviceRegistrationFailedError;
/**
 * Error thrown when a low level security device/driver fails.
 *
 * @augments {Error}
 */
var SecurityDeviceError = (function (_super) {
    __extends(SecurityDeviceError, _super);
    function SecurityDeviceError(message) {
        _super.call(this, message);
        this.name = 'SecurityDeviceError';
        this.message = message;
        Error.captureStackTrace(this, this.constructor);
    }
    return SecurityDeviceError;
}(Error));
exports.SecurityDeviceError = SecurityDeviceError;
