// node-opcua-service-session/_generated_/_auto_generated_CreateSessionRequest.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:28.431Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/CreateSessionRequest_schema").CreateSessionRequest_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var RequestHeader = getFactory("RequestHeader");
var ApplicationDescription = getFactory("ApplicationDescription");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * Creates a new session with the server.
 * 
 * @class CreateSessionRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} 
 * @param  [options.clientDescription] {ApplicationDescription} Describes the client application.
 * @param  [options.serverUri] {String} The URI of the server that the client wants to create a session with.
 * @param  [options.endpointUrl] {String} The URL that the client used to connect to the server.
 * @param  [options.sessionName] {String} A name for the session provided by the client.
 * @param  [options.clientNonce] {ByteString} A random number generated by the client.
 * @param  [options.clientCertificate] {ByteString} The application certificate for the client.
 * @param  [options.requestedSessionTimeout] {Duration} The requested session timeout in milliseconds.
 * @param  [options.maxResponseMessageSize] {UInt32} The maximum message size accepted by the client.
 */
function CreateSessionRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestHeader =  null; /* new RequestHeader(null); */
        self.clientDescription =  null; /* new ApplicationDescription(null); */
        return ;
    }

    /**
      * 
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * Describes the client application.
      * @property clientDescription
      * @type {ApplicationDescription}
      */
    self.clientDescription =  new ApplicationDescription( options.clientDescription);

    /**
      * The URI of the server that the client wants to create a session with.
      * @property serverUri
      * @type {String}
      */
    self.serverUri = initialize_field(schema.fields[2], options.serverUri);

    /**
      * The URL that the client used to connect to the server.
      * @property endpointUrl
      * @type {String}
      */
    self.endpointUrl = initialize_field(schema.fields[3], options.endpointUrl);

    /**
      * A name for the session provided by the client.
      * @property sessionName
      * @type {String}
      */
    self.sessionName = initialize_field(schema.fields[4], options.sessionName);

    /**
      * A random number generated by the client.
      * @property clientNonce
      * @type {ByteString}
      */
    self.clientNonce = initialize_field(schema.fields[5], options.clientNonce);

    /**
      * The application certificate for the client.
      * @property clientCertificate
      * @type {ByteString}
      */
    self.clientCertificate = initialize_field(schema.fields[6], options.clientCertificate);

    /**
      * The requested session timeout in milliseconds.
      * @property requestedSessionTimeout
      * @type {Duration}
      */
    self.requestedSessionTimeout = initialize_field(schema.fields[7], options.requestedSessionTimeout);

    /**
      * The maximum message size accepted by the client.
      * @property maxResponseMessageSize
      * @type {UInt32}
      */
    self.maxResponseMessageSize = initialize_field(schema.fields[8], options.maxResponseMessageSize);

   // Object.preventExtensions(self);
}
util.inherits(CreateSessionRequest,BaseUAObject);
CreateSessionRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(461,0);
CreateSessionRequest.prototype.encodingDefaultXml = makeExpandedNodeId(460,0);
CreateSessionRequest.prototype._schema = schema;

var encode_String = _defaultTypeMap.String.encode;
var decode_String = _defaultTypeMap.String.decode;
var encode_ByteString = _defaultTypeMap.ByteString.encode;
var decode_ByteString = _defaultTypeMap.ByteString.decode;
var encode_Duration = _defaultTypeMap.Duration.encode;
var decode_Duration = _defaultTypeMap.Duration.decode;
var encode_UInt32 = _defaultTypeMap.UInt32.encode;
var decode_UInt32 = _defaultTypeMap.UInt32.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
CreateSessionRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
   this.clientDescription.encode(stream,options);
    encode_String(this.serverUri,stream);
    encode_String(this.endpointUrl,stream);
    encode_String(this.sessionName,stream);
    encode_ByteString(this.clientNonce,stream);
    encode_ByteString(this.clientCertificate,stream);
    encode_Duration(this.requestedSessionTimeout,stream);
    encode_UInt32(this.maxResponseMessageSize,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
CreateSessionRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.clientDescription.decode(stream,options);
    this.serverUri = decode_String(stream,options);
    this.endpointUrl = decode_String(stream,options);
    this.sessionName = decode_String(stream,options);
    this.clientNonce = decode_ByteString(stream,options);
    this.clientCertificate = decode_ByteString(stream,options);
    this.requestedSessionTimeout = decode_Duration(stream,options);
    this.maxResponseMessageSize = decode_UInt32(stream,options);
};
CreateSessionRequest.possibleFields = [
  "requestHeader",
         "clientDescription",
         "serverUri",
         "endpointUrl",
         "sessionName",
         "clientNonce",
         "clientCertificate",
         "requestedSessionTimeout",
         "maxResponseMessageSize"
];


exports.CreateSessionRequest = CreateSessionRequest;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("CreateSessionRequest",CreateSessionRequest);
