// node-opcua-service-session/_generated_/_auto_generated_UserNameIdentityToken.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:28.472Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/UserNameIdentityToken_schema").UserNameIdentityToken_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * A token representing a user identified by a user name and password
 * 
 * @class UserNameIdentityToken
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.policyId] {String} The policy id specified in a user token policy for the endpoint being used.
 * @param  [options.userName] {String} The user name
 * @param  [options.password] {ByteString} The password encrypted with the server certificate.
 * @param  [options.encryptionAlgorithm] {String} The algorithm used to encrypt the password.
 */
function UserNameIdentityToken(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        return ;
    }

    /**
      * The policy id specified in a user token policy for the endpoint being used.
      * @property policyId
      * @type {String}
      */
    self.policyId = initialize_field(schema.fields[0], options.policyId);

    /**
      * The user name
      * @property userName
      * @type {String}
      */
    self.userName = initialize_field(schema.fields[1], options.userName);

    /**
      * The password encrypted with the server certificate.
      * @property password
      * @type {ByteString}
      */
    self.password = initialize_field(schema.fields[2], options.password);

    /**
      * The algorithm used to encrypt the password.
      * @property encryptionAlgorithm
      * @type {String}
      */
    self.encryptionAlgorithm = initialize_field(schema.fields[3], options.encryptionAlgorithm);

   // Object.preventExtensions(self);
}
util.inherits(UserNameIdentityToken,BaseUAObject);
UserNameIdentityToken.prototype.encodingDefaultBinary = makeExpandedNodeId(324,0);
UserNameIdentityToken.prototype.encodingDefaultXml = makeExpandedNodeId(323,0);
UserNameIdentityToken.prototype._schema = schema;

var encode_String = _defaultTypeMap.String.encode;
var decode_String = _defaultTypeMap.String.decode;
var encode_ByteString = _defaultTypeMap.ByteString.encode;
var decode_ByteString = _defaultTypeMap.ByteString.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
UserNameIdentityToken.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_String(this.policyId,stream);
    encode_String(this.userName,stream);
    encode_ByteString(this.password,stream);
    encode_String(this.encryptionAlgorithm,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
UserNameIdentityToken.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.policyId = decode_String(stream,options);
    this.userName = decode_String(stream,options);
    this.password = decode_ByteString(stream,options);
    this.encryptionAlgorithm = decode_String(stream,options);
};
UserNameIdentityToken.possibleFields = [
  "policyId",
         "userName",
         "password",
         "encryptionAlgorithm"
];


exports.UserNameIdentityToken = UserNameIdentityToken;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("UserNameIdentityToken",UserNameIdentityToken);
