// node-opcua-service-filter/_generated_/_auto_generated_EventFilter.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:13.731Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/EventFilter_schema").EventFilter_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const SimpleAttributeOperand = require("./_auto_generated_SimpleAttributeOperand").SimpleAttributeOperand;
const ContentFilter = require("./_auto_generated_ContentFilter").ContentFilter;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;
const MonitoringFilter = require("./_auto_generated_MonitoringFilter").MonitoringFilter;

/**
 * 
 * @class EventFilter
 * @constructor
 * @extends MonitoringFilter
 * @param  options {Object}
 */
function EventFilter(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    MonitoringFilter.call(this,options);
    if (options === null) { 
        MonitoringFilter.call(this,options);
        self.selectClauses =  null; /* null array */
        self.whereClause =  null; /* new ContentFilter(null); */
        return ;
    }

    /**
      * 
      * @property selectClauses
      * @type {SimpleAttributeOperand[]}
      */
    self.selectClauses = [];
    if (options.selectClauses) {
        assert(_.isArray(options.selectClauses));
        self.selectClauses = options.selectClauses.map(function(e){ return new SimpleAttributeOperand(e); } );
    }

    /**
      * 
      * @property whereClause
      * @type {ContentFilter}
      */
    self.whereClause =  new ContentFilter( options.whereClause);

   // Object.preventExtensions(self);
}
util.inherits(EventFilter,MonitoringFilter);
EventFilter.prototype.encodingDefaultBinary = makeExpandedNodeId(727,0);
EventFilter.prototype.encodingDefaultXml = makeExpandedNodeId(726,0);
EventFilter.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
EventFilter.prototype.encode = function(stream,options) {
    // call base class implementation first
    MonitoringFilter.prototype.encode.call(this,stream,options);
    encodeArray(this.selectClauses,stream,function(obj,stream){ obj.encode(stream,options); }); 
   this.whereClause.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
EventFilter.prototype.decode = function(stream) {
    // call base class implementation first
    MonitoringFilter.prototype.decode.call(this,stream);
    this.selectClauses = decodeArray(stream, function(stream) { 
       const obj = new SimpleAttributeOperand(null);
       obj.decode(stream);
       return obj; 
    });
    this.whereClause.decode(stream);
};
EventFilter.possibleFields = [
  "selectClauses",
         "whereClause"
];


exports.EventFilter = EventFilter;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("EventFilter",EventFilter);
