// node-opcua-service-filter/_generated_/_auto_generated_ElementOperand.js
// --------- This code has been automatically generated !!! 2018-07-30T21:12:13.609Z
"use strict";
/**
 * @module opcua.address_space.types
 */
const assert = require("node-opcua-assert").assert;
const util = require("util");
const _  = require("underscore");
const makeNodeId = require("node-opcua-nodeid").makeNodeId;
const schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
const extract_all_fields                       = schema_helpers.extract_all_fields;
const resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
const initialize_field                         = schema_helpers.initialize_field;
const initialize_field_array                   = schema_helpers.initialize_field_array;
const check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
const _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
const ec = require("node-opcua-basic-types");
const encodeArray = ec.encodeArray;
const decodeArray = ec.decodeArray;
const makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
const generate_new_id = require("node-opcua-factory").generate_new_id;
const _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
const schema = require("../schemas/ElementOperand_schema").ElementOperand_Schema;
const getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
const BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;
const FilterOperand = require("./_auto_generated_FilterOperand").FilterOperand;

/**
 * 
 * @class ElementOperand
 * @constructor
 * @extends FilterOperand
 * @param  options {Object}
 */
function ElementOperand(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    const self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    FilterOperand.call(this,options);
    if (options === null) { 
        FilterOperand.call(this,options);
        return ;
    }

    /**
      * Index into the element array
      * @property index
      * @type {UInt32}
      */
    self.index = initialize_field(schema.fields[0], options.index);

   // Object.preventExtensions(self);
}
util.inherits(ElementOperand,FilterOperand);
ElementOperand.prototype.encodingDefaultBinary = makeExpandedNodeId(594,0);
ElementOperand.prototype.encodingDefaultXml = makeExpandedNodeId(593,0);
ElementOperand.prototype._schema = schema;

const encode_UInt32 = _defaultTypeMap.UInt32.encode;
const decode_UInt32 = _defaultTypeMap.UInt32.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ElementOperand.prototype.encode = function(stream,options) {
    // call base class implementation first
    FilterOperand.prototype.encode.call(this,stream,options);
    encode_UInt32(this.index,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 */
ElementOperand.prototype.decode = function(stream) {
    // call base class implementation first
    FilterOperand.prototype.decode.call(this,stream);
    this.index = decode_UInt32(stream);
};
ElementOperand.possibleFields = [
  "index"
];


exports.ElementOperand = ElementOperand;
const register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ElementOperand",ElementOperand);
