// node-opcua-service-query/_generated_/_auto_generated_QueryFirstRequest.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:37.401Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/QueryFirstRequest_schema").QueryFirstRequest_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var RequestHeader = getFactory("RequestHeader");
var ViewDescription = getFactory("ViewDescription");
var NodeTypeDescription = require("./_auto_generated_NodeTypeDescription").NodeTypeDescription;
var ContentFilter = getFactory("ContentFilter");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class QueryFirstRequest
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.requestHeader] {RequestHeader} A standard header included in all requests sent to a server.
 * @param  [options.view] {ViewDescription} 
 * @param  [options.nodeTypes] {NodeTypeDescription[]} 
 * @param  [options.filter] {ContentFilter} 
 * @param  [options.maxDataSetsToReturn] {Counter} 
 * @param  [options.maxReferencesToReturn] {Counter} 
 */
function QueryFirstRequest(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.requestHeader =  null; /* new RequestHeader(null); */
        self.view =  null; /* new ViewDescription(null); */
        self.nodeTypes =  null; /* null array */
        self.filter =  null; /* new ContentFilter(null); */
        return ;
    }

    /**
      * A standard header included in all requests sent to a server.
      * @property requestHeader
      * @type {RequestHeader}
      */
    self.requestHeader =  new RequestHeader( options.requestHeader);

    /**
      * 
      * @property view
      * @type {ViewDescription}
      */
    self.view =  new ViewDescription( options.view);

    /**
      * 
      * @property nodeTypes
      * @type {NodeTypeDescription[]}
      */
    self.nodeTypes = [];
    if (options.nodeTypes) {
        assert(_.isArray(options.nodeTypes));
        self.nodeTypes = options.nodeTypes.map(function(e){ return new NodeTypeDescription(e); } );
    }

    /**
      * 
      * @property filter
      * @type {ContentFilter}
      */
    self.filter =  new ContentFilter( options.filter);

    /**
      * 
      * @property maxDataSetsToReturn
      * @type {Counter}
      */
    self.maxDataSetsToReturn = initialize_field(schema.fields[4], options.maxDataSetsToReturn);

    /**
      * 
      * @property maxReferencesToReturn
      * @type {Counter}
      */
    self.maxReferencesToReturn = initialize_field(schema.fields[5], options.maxReferencesToReturn);

   // Object.preventExtensions(self);
}
util.inherits(QueryFirstRequest,BaseUAObject);
QueryFirstRequest.prototype.encodingDefaultBinary = makeExpandedNodeId(615,0);
QueryFirstRequest.prototype.encodingDefaultXml = makeExpandedNodeId(614,0);
QueryFirstRequest.prototype._schema = schema;

var encode_Counter = _defaultTypeMap.Counter.encode;
var decode_Counter = _defaultTypeMap.Counter.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
QueryFirstRequest.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.requestHeader.encode(stream,options);
   this.view.encode(stream,options);
    encodeArray(this.nodeTypes,stream,function(obj,stream){ obj.encode(stream,options); }); 
   this.filter.encode(stream,options);
    encode_Counter(this.maxDataSetsToReturn,stream);
    encode_Counter(this.maxReferencesToReturn,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
QueryFirstRequest.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.requestHeader.decode(stream,options);
    this.view.decode(stream,options);
    this.nodeTypes = decodeArray(stream, function(stream) { 
       var obj = new NodeTypeDescription(null);
       obj.decode(stream,options);
       return obj; 
    });
    this.filter.decode(stream,options);
    this.maxDataSetsToReturn = decode_Counter(stream,options);
    this.maxReferencesToReturn = decode_Counter(stream,options);
};
QueryFirstRequest.possibleFields = [
  "requestHeader",
         "view",
         "nodeTypes",
         "filter",
         "maxDataSetsToReturn",
         "maxReferencesToReturn"
];


exports.QueryFirstRequest = QueryFirstRequest;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("QueryFirstRequest",QueryFirstRequest);
