// node-opcua-service-query/_generated_/_auto_generated_QueryDataSet.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:37.407Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/QueryDataSet_schema").QueryDataSet_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var Variant = getFactory("Variant");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * Data related to a Node returned in a Query response.
 * 
 * @class QueryDataSet
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.nodeId] {ExpandedNodeId} 
 * @param  [options.typeDefinitionNode] {ExpandedNodeId} 
 * @param  [options.values] {Variant[]} 
 */
function QueryDataSet(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.values =  null; /* null array */
        return ;
    }

    /**
      * 
      * @property nodeId
      * @type {ExpandedNodeId}
      */
    self.nodeId = initialize_field(schema.fields[0], options.nodeId);

    /**
      * 
      * @property typeDefinitionNode
      * @type {ExpandedNodeId}
      */
    self.typeDefinitionNode = initialize_field(schema.fields[1], options.typeDefinitionNode);

    /**
      * 
      * @property values
      * @type {Variant[]}
      */
    self.values = [];
    if (options.values) {
        assert(_.isArray(options.values));
        self.values = options.values.map(function(e){ return new Variant(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(QueryDataSet,BaseUAObject);
QueryDataSet.prototype.encodingDefaultBinary = makeExpandedNodeId(579,0);
QueryDataSet.prototype.encodingDefaultXml = makeExpandedNodeId(578,0);
QueryDataSet.prototype._schema = schema;

var encode_ExpandedNodeId = _defaultTypeMap.ExpandedNodeId.encode;
var decode_ExpandedNodeId = _defaultTypeMap.ExpandedNodeId.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
QueryDataSet.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_ExpandedNodeId(this.nodeId,stream);
    encode_ExpandedNodeId(this.typeDefinitionNode,stream);
    encodeArray(this.values,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
QueryDataSet.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.nodeId = decode_ExpandedNodeId(stream,options);
    this.typeDefinitionNode = decode_ExpandedNodeId(stream,options);
    this.values = decodeArray(stream, function(stream) { 
       var obj = new Variant(null);
       obj.decode(stream,options);
       return obj; 
    });
};
QueryDataSet.possibleFields = [
  "nodeId",
         "typeDefinitionNode",
         "values"
];


exports.QueryDataSet = QueryDataSet;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("QueryDataSet",QueryDataSet);
