// node-opcua-service-query/_generated_/_auto_generated_QueryNextResponse.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:37.636Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/QueryNextResponse_schema").QueryNextResponse_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var ResponseHeader = getFactory("ResponseHeader");
var QueryDataSet = require("./_auto_generated_QueryDataSet").QueryDataSet;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class QueryNextResponse
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.responseHeader] {ResponseHeader} 
 * @param  [options.queryDataSet] {QueryDataSet[]} 
 * @param  [options.revisedContinuationPoint] {ContinuationPoint} 
 */
function QueryNextResponse(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.responseHeader =  null; /* new ResponseHeader(null); */
        self.queryDataSet =  null; /* null array */
        return ;
    }

    /**
      * 
      * @property responseHeader
      * @type {ResponseHeader}
      */
    self.responseHeader =  new ResponseHeader( options.responseHeader);

    /**
      * 
      * @property queryDataSet
      * @type {QueryDataSet[]}
      */
    self.queryDataSet = [];
    if (options.queryDataSet) {
        assert(_.isArray(options.queryDataSet));
        self.queryDataSet = options.queryDataSet.map(function(e){ return new QueryDataSet(e); } );
    }

    /**
      * 
      * @property revisedContinuationPoint
      * @type {ContinuationPoint}
      */
    self.revisedContinuationPoint = initialize_field(schema.fields[2], options.revisedContinuationPoint);

   // Object.preventExtensions(self);
}
util.inherits(QueryNextResponse,BaseUAObject);
QueryNextResponse.prototype.encodingDefaultBinary = makeExpandedNodeId(624,0);
QueryNextResponse.prototype.encodingDefaultXml = makeExpandedNodeId(623,0);
QueryNextResponse.prototype._schema = schema;

var encode_ContinuationPoint = _defaultTypeMap.ContinuationPoint.encode;
var decode_ContinuationPoint = _defaultTypeMap.ContinuationPoint.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
QueryNextResponse.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
   this.responseHeader.encode(stream,options);
    encodeArray(this.queryDataSet,stream,function(obj,stream){ obj.encode(stream,options); }); 
    encode_ContinuationPoint(this.revisedContinuationPoint,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
QueryNextResponse.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.responseHeader.decode(stream,options);
    this.queryDataSet = decodeArray(stream, function(stream) { 
       var obj = new QueryDataSet(null);
       obj.decode(stream,options);
       return obj; 
    });
    this.revisedContinuationPoint = decode_ContinuationPoint(stream,options);
};
QueryNextResponse.possibleFields = [
  "responseHeader",
         "queryDataSet",
         "revisedContinuationPoint"
];


exports.QueryNextResponse = QueryNextResponse;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("QueryNextResponse",QueryNextResponse);
