// node-opcua-service-query/_generated_/_auto_generated_NodeTypeDescription.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:37.294Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/NodeTypeDescription_schema").NodeTypeDescription_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var QueryDataDescription = require("./_auto_generated_QueryDataDescription").QueryDataDescription;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class NodeTypeDescription
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.typeDefinitionNode] {ExpandedNodeId} NodeId of the originating TypeDefinitionNode of the instances for which data is to be returned.
 * @param  [options.includeSubtypes] {Boolean} A flag that indicates whether the Server should include instances of subtypes of the TypeDefinitionNode in the list of instances of the Node type.
 * @param  [options.dataToReturn] {QueryDataDescription[]} Specifies an Attribute or Reference from the originating typeDefinitionNode along a given relativePath for which to return data. This structure is defined in-line with the following indented items.
 */
function NodeTypeDescription(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.dataToReturn =  null; /* null array */
        return ;
    }

    /**
      * NodeId of the originating TypeDefinitionNode of the instances for which data is to be returned.
      * @property typeDefinitionNode
      * @type {ExpandedNodeId}
      */
    self.typeDefinitionNode = initialize_field(schema.fields[0], options.typeDefinitionNode);

    /**
      * A flag that indicates whether the Server should include instances of subtypes of the TypeDefinitionNode in the list of instances of the Node type.
      * @property includeSubtypes
      * @type {Boolean}
      */
    self.includeSubtypes = initialize_field(schema.fields[1], options.includeSubtypes);

    /**
      * Specifies an Attribute or Reference from the originating typeDefinitionNode along a given relativePath for which to return data. This structure is defined in-line with the following indented items.
      * @property dataToReturn
      * @type {QueryDataDescription[]}
      */
    self.dataToReturn = [];
    if (options.dataToReturn) {
        assert(_.isArray(options.dataToReturn));
        self.dataToReturn = options.dataToReturn.map(function(e){ return new QueryDataDescription(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(NodeTypeDescription,BaseUAObject);
NodeTypeDescription.prototype.encodingDefaultBinary = makeExpandedNodeId(575,0);
NodeTypeDescription.prototype.encodingDefaultXml = makeExpandedNodeId(574,0);
NodeTypeDescription.prototype._schema = schema;

var encode_ExpandedNodeId = _defaultTypeMap.ExpandedNodeId.encode;
var decode_ExpandedNodeId = _defaultTypeMap.ExpandedNodeId.decode;
var encode_Boolean = _defaultTypeMap.Boolean.encode;
var decode_Boolean = _defaultTypeMap.Boolean.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
NodeTypeDescription.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_ExpandedNodeId(this.typeDefinitionNode,stream);
    encode_Boolean(this.includeSubtypes,stream);
    encodeArray(this.dataToReturn,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
NodeTypeDescription.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.typeDefinitionNode = decode_ExpandedNodeId(stream,options);
    this.includeSubtypes = decode_Boolean(stream,options);
    this.dataToReturn = decodeArray(stream, function(stream) { 
       var obj = new QueryDataDescription(null);
       obj.decode(stream,options);
       return obj; 
    });
};
NodeTypeDescription.possibleFields = [
  "typeDefinitionNode",
         "includeSubtypes",
         "dataToReturn"
];


exports.NodeTypeDescription = NodeTypeDescription;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("NodeTypeDescription",NodeTypeDescription);
