// node-opcua-service-endpoints/_generated_/_auto_generated_ApplicationDescription.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:18.560Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ApplicationDescription_schema").ApplicationDescription_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var LocalizedText = getFactory("LocalizedText");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class ApplicationDescription
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.applicationUri] {String} 
 * @param  [options.productUri] {String} 
 * @param  [options.applicationName] {LocalizedText} 
 * @param  [options.applicationType] {ApplicationType} 
 * @param  [options.gatewayServerUri] {String} 
 * @param  [options.discoveryProfileUri] {String} 
 * @param  [options.discoveryUrls] {String[]} 
 */
function ApplicationDescription(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.applicationName =  null; /* new LocalizedText(null); */
        return ;
    }

    /**
      * 
      * @property applicationUri
      * @type {String}
      */
    self.applicationUri = initialize_field(schema.fields[0], options.applicationUri);

    /**
      * 
      * @property productUri
      * @type {String}
      */
    self.productUri = initialize_field(schema.fields[1], options.productUri);

    /**
      * 
      * @property applicationName
      * @type {LocalizedText}
      */
    self.applicationName =  new LocalizedText( options.applicationName);

    /**
      * 
      * @property applicationType
      * @type {ApplicationType}
      */
    self.setApplicationType(initialize_field(schema.fields[3], options.applicationType));

    /**
      * 
      * @property gatewayServerUri
      * @type {String}
      */
    self.gatewayServerUri = initialize_field(schema.fields[4], options.gatewayServerUri);

    /**
      * 
      * @property discoveryProfileUri
      * @type {String}
      */
    self.discoveryProfileUri = initialize_field(schema.fields[5], options.discoveryProfileUri);

    /**
      * 
      * @property discoveryUrls
      * @type {String[]}
      */
    self.discoveryUrls = initialize_field_array(schema.fields[6], options.discoveryUrls);

   // Object.preventExtensions(self);
}
util.inherits(ApplicationDescription,BaseUAObject);

//## Define Enumeration setters
ApplicationDescription.prototype.setApplicationType = function(value) {
   var coercedValue = _enumerations.ApplicationType.typedEnum.get(value);
   /* istanbul ignore next */
   if (coercedValue === undefined || coercedValue === null) {
      throw new Error("value cannot be coerced to ApplicationType: " + value);
   }
   this.applicationType = coercedValue;
};
ApplicationDescription.prototype.encodingDefaultBinary = makeExpandedNodeId(310,0);
ApplicationDescription.prototype.encodingDefaultXml = makeExpandedNodeId(309,0);
ApplicationDescription.prototype._schema = schema;

var encode_String = _defaultTypeMap.String.encode;
var decode_String = _defaultTypeMap.String.decode;
var encode_ApplicationType = _enumerations.ApplicationType.encode;
var decode_ApplicationType = _enumerations.ApplicationType.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ApplicationDescription.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_String(this.applicationUri,stream);
    encode_String(this.productUri,stream);
   this.applicationName.encode(stream,options);
    encode_ApplicationType(this.applicationType,stream);
    encode_String(this.gatewayServerUri,stream);
    encode_String(this.discoveryProfileUri,stream);
    encodeArray(this.discoveryUrls, stream, encode_String);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
ApplicationDescription.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.applicationUri = decode_String(stream,options);
    this.productUri = decode_String(stream,options);
    this.applicationName.decode(stream,options);
    this.applicationType = decode_ApplicationType(stream,options);
    this.gatewayServerUri = decode_String(stream,options);
    this.discoveryProfileUri = decode_String(stream,options);
    this.discoveryUrls = decodeArray(stream, decode_String);
};
ApplicationDescription.possibleFields = [
  "applicationUri",
         "productUri",
         "applicationName",
         "applicationType",
         "gatewayServerUri",
         "discoveryProfileUri",
         "discoveryUrls"
];


exports.ApplicationDescription = ApplicationDescription;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ApplicationDescription",ApplicationDescription);
