// node-opcua-service-history/_generated_/_auto_generated_HistoryReadValueId.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:31.051Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/HistoryReadValueId_schema").HistoryReadValueId_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var QualifiedName = getFactory("QualifiedName");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class HistoryReadValueId
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.nodeId] {NodeId} 
 * @param  [options.indexRange] {String} 
 * @param  [options.dataEncoding] {QualifiedName} 
 * @param  [options.continuationPoint = null] {ByteString} 
 */
function HistoryReadValueId(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.dataEncoding =  null; /* new QualifiedName(null); */
        return ;
    }

    /**
      * 
      * @property nodeId
      * @type {NodeId}
      */
    self.nodeId = initialize_field(schema.fields[0], options.nodeId);

    /**
      * 
      * @property indexRange
      * @type {String}
      */
    self.indexRange = initialize_field(schema.fields[1], options.indexRange);

    /**
      * 
      * @property dataEncoding
      * @type {QualifiedName}
      */
    self.dataEncoding =  new QualifiedName( options.dataEncoding);

    /**
      * 
      * @property continuationPoint
      * @type {ByteString}
      * @default  null
      */
    self.continuationPoint = initialize_field(schema.fields[3], options.continuationPoint);

   // Object.preventExtensions(self);
}
util.inherits(HistoryReadValueId,BaseUAObject);
HistoryReadValueId.prototype.encodingDefaultBinary = makeExpandedNodeId(637,0);
HistoryReadValueId.prototype.encodingDefaultXml = makeExpandedNodeId(636,0);
HistoryReadValueId.prototype._schema = schema;

var encode_NodeId = _defaultTypeMap.NodeId.encode;
var decode_NodeId = _defaultTypeMap.NodeId.decode;
var encode_String = _defaultTypeMap.String.encode;
var decode_String = _defaultTypeMap.String.decode;
var encode_ByteString = _defaultTypeMap.ByteString.encode;
var decode_ByteString = _defaultTypeMap.ByteString.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
HistoryReadValueId.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encode_NodeId(this.nodeId,stream);
    encode_String(this.indexRange,stream);
   this.dataEncoding.encode(stream,options);
    encode_ByteString(this.continuationPoint,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
HistoryReadValueId.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.nodeId = decode_NodeId(stream,options);
    this.indexRange = decode_String(stream,options);
    this.dataEncoding.decode(stream,options);
    this.continuationPoint = decode_ByteString(stream,options);
};
HistoryReadValueId.possibleFields = [
  "nodeId",
         "indexRange",
         "dataEncoding",
         "continuationPoint"
];


exports.HistoryReadValueId = HistoryReadValueId;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("HistoryReadValueId",HistoryReadValueId);
