// node-opcua-service-history/_generated_/_auto_generated_ReadEventDetails.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:31.297Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ReadEventDetails_schema").ReadEventDetails_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var EventFilter = getFactory("EventFilter");
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;
var HistoryReadDetails = require("./_auto_generated_HistoryReadDetails").HistoryReadDetails;

/**
 * 
 * @class ReadEventDetails
 * @constructor
 * @extends HistoryReadDetails
 * @param  options {Object}
 * @param  [options.numValuesPerNode] {UInt32} 
 * @param  [options.startTime] {DateTime} 
 * @param  [options.endTime] {DateTime} 
 * @param  [options.filter] {EventFilter} 
 */
function ReadEventDetails(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    HistoryReadDetails.call(this,options);
    if (options === null) { 
        HistoryReadDetails.call(this,options);
        self.filter =  null; /* new EventFilter(null); */
        return ;
    }

    /**
      * 
      * @property numValuesPerNode
      * @type {UInt32}
      */
    self.numValuesPerNode = initialize_field(schema.fields[0], options.numValuesPerNode);

    /**
      * 
      * @property startTime
      * @type {DateTime}
      */
    self.startTime = initialize_field(schema.fields[1], options.startTime);

    /**
      * 
      * @property endTime
      * @type {DateTime}
      */
    self.endTime = initialize_field(schema.fields[2], options.endTime);

    /**
      * 
      * @property filter
      * @type {EventFilter}
      */
    self.filter =  new EventFilter( options.filter);

   // Object.preventExtensions(self);
}
util.inherits(ReadEventDetails,HistoryReadDetails);
ReadEventDetails.prototype.encodingDefaultBinary = makeExpandedNodeId(646,0);
ReadEventDetails.prototype.encodingDefaultXml = makeExpandedNodeId(645,0);
ReadEventDetails.prototype._schema = schema;

var encode_UInt32 = _defaultTypeMap.UInt32.encode;
var decode_UInt32 = _defaultTypeMap.UInt32.decode;
var encode_DateTime = _defaultTypeMap.DateTime.encode;
var decode_DateTime = _defaultTypeMap.DateTime.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ReadEventDetails.prototype.encode = function(stream,options) {
    // call base class implementation first
    HistoryReadDetails.prototype.encode.call(this,stream,options);
    encode_UInt32(this.numValuesPerNode,stream);
    encode_DateTime(this.startTime,stream);
    encode_DateTime(this.endTime,stream);
   this.filter.encode(stream,options);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
ReadEventDetails.prototype.decode = function(stream,options) {
    // call base class implementation first
    HistoryReadDetails.prototype.decode.call(this,stream,options);
    this.numValuesPerNode = decode_UInt32(stream,options);
    this.startTime = decode_DateTime(stream,options);
    this.endTime = decode_DateTime(stream,options);
    this.filter.decode(stream,options);
};
ReadEventDetails.possibleFields = [
  "numValuesPerNode",
         "startTime",
         "endTime",
         "filter"
];


exports.ReadEventDetails = ReadEventDetails;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ReadEventDetails",ReadEventDetails);
