// node-opcua-service-history/_generated_/_auto_generated_HistoryModifiedData.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:31.159Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/HistoryModifiedData_schema").HistoryModifiedData_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var DataValue = getFactory("DataValue");
var ModificationInfo = require("./_auto_generated_ModificationInfo").ModificationInfo;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;

/**
 * 
 * @class HistoryModifiedData
 * @constructor
 * @extends BaseUAObject
 * @param  options {Object}
 * @param  [options.dataValues] {DataValue[]} An array of values of history data for the Node. The size of the array depends on the requested data parameters.
 * @param  [options.modificationInfos] {ModificationInfo[]} 
 */
function HistoryModifiedData(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    BaseUAObject.call(this,options);
    if (options === null) { 
        BaseUAObject.call(this,options);
        self.dataValues =  null; /* null array */
        self.modificationInfos =  null; /* null array */
        return ;
    }

    /**
      * An array of values of history data for the Node. The size of the array depends on the requested data parameters.
      * @property dataValues
      * @type {DataValue[]}
      */
    self.dataValues = [];
    if (options.dataValues) {
        assert(_.isArray(options.dataValues));
        self.dataValues = options.dataValues.map(function(e){ return new DataValue(e); } );
    }

    /**
      * 
      * @property modificationInfos
      * @type {ModificationInfo[]}
      */
    self.modificationInfos = [];
    if (options.modificationInfos) {
        assert(_.isArray(options.modificationInfos));
        self.modificationInfos = options.modificationInfos.map(function(e){ return new ModificationInfo(e); } );
    }

   // Object.preventExtensions(self);
}
util.inherits(HistoryModifiedData,BaseUAObject);
HistoryModifiedData.prototype.encodingDefaultBinary = makeExpandedNodeId(11227,0);
HistoryModifiedData.prototype.encodingDefaultXml = makeExpandedNodeId(11219,0);
HistoryModifiedData.prototype._schema = schema;

/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
HistoryModifiedData.prototype.encode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.encode.call(this,stream,options);
    encodeArray(this.dataValues,stream,function(obj,stream){ obj.encode(stream,options); }); 
    encodeArray(this.modificationInfos,stream,function(obj,stream){ obj.encode(stream,options); }); 
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
HistoryModifiedData.prototype.decode = function(stream,options) {
    // call base class implementation first
    BaseUAObject.prototype.decode.call(this,stream,options);
    this.dataValues = decodeArray(stream, function(stream) { 
       var obj = new DataValue(null);
       obj.decode(stream,options);
       return obj; 
    });
    this.modificationInfos = decodeArray(stream, function(stream) { 
       var obj = new ModificationInfo(null);
       obj.decode(stream,options);
       return obj; 
    });
};
HistoryModifiedData.possibleFields = [
  "dataValues",
         "modificationInfos"
];


exports.HistoryModifiedData = HistoryModifiedData;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("HistoryModifiedData",HistoryModifiedData);
