// node-opcua-service-history/_generated_/_auto_generated_ReadAtTimeDetails.js
// --------- This code has been automatically generated !!! 2018-02-08T07:25:31.187Z
"use strict";
/**
 * @module opcua.address_space.types
 */
var assert = require("node-opcua-assert");
var util = require("util");
var _  = require("underscore");
var makeNodeId = require("node-opcua-nodeid").makeNodeId;
var schema_helpers =  require("node-opcua-factory/src/factories_schema_helpers");
var extract_all_fields                       = schema_helpers.extract_all_fields;
var resolve_schema_field_types               = schema_helpers.resolve_schema_field_types;
var initialize_field                         = schema_helpers.initialize_field;
var initialize_field_array                   = schema_helpers.initialize_field_array;
var check_options_correctness_against_schema = schema_helpers.check_options_correctness_against_schema;
var _defaultTypeMap = require("node-opcua-factory/src/factories_builtin_types")._defaultTypeMap;
var ec = require("node-opcua-basic-types");
var encodeArray = ec.encodeArray;
var decodeArray = ec.decodeArray;
var makeExpandedNodeId = require("node-opcua-nodeid/src/expanded_nodeid").makeExpandedNodeId;
var generate_new_id = require("node-opcua-factory").generate_new_id;
var _enumerations = require("node-opcua-factory/src/factories_enumerations")._private._enumerations;
var schema = require("../schemas/ReadAtTimeDetails_schema").ReadAtTimeDetails_Schema;
var getFactory = require("node-opcua-factory/src/factories_factories").getFactory;
var BaseUAObject = require("node-opcua-factory/src/factories_baseobject").BaseUAObject;
var HistoryReadDetails = require("./_auto_generated_HistoryReadDetails").HistoryReadDetails;

/**
 * 
 * @class ReadAtTimeDetails
 * @constructor
 * @extends HistoryReadDetails
 * @param  options {Object}
 * @param  [options.reqTimes] {DateTime[]} 
 * @param  [options.useSimpleBounds] {Boolean} 
 */
function ReadAtTimeDetails(options)
{
    options = options || {};
    /* istanbul ignore next */
    if (schema_helpers.doDebug) { check_options_correctness_against_schema(this,schema,options); }
    var self = this;
    assert(this instanceof BaseUAObject); //  ' keyword "new" is required for constructor call')
    resolve_schema_field_types(schema);

    HistoryReadDetails.call(this,options);
    if (options === null) { 
        HistoryReadDetails.call(this,options);
        return ;
    }

    /**
      * 
      * @property reqTimes
      * @type {DateTime[]}
      */
    self.reqTimes = initialize_field_array(schema.fields[0], options.reqTimes);

    /**
      * 
      * @property useSimpleBounds
      * @type {Boolean}
      */
    self.useSimpleBounds = initialize_field(schema.fields[1], options.useSimpleBounds);

   // Object.preventExtensions(self);
}
util.inherits(ReadAtTimeDetails,HistoryReadDetails);
ReadAtTimeDetails.prototype.encodingDefaultBinary = makeExpandedNodeId(655,0);
ReadAtTimeDetails.prototype.encodingDefaultXml = makeExpandedNodeId(654,0);
ReadAtTimeDetails.prototype._schema = schema;

var encode_DateTime = _defaultTypeMap.DateTime.encode;
var decode_DateTime = _defaultTypeMap.DateTime.decode;
var encode_Boolean = _defaultTypeMap.Boolean.encode;
var decode_Boolean = _defaultTypeMap.Boolean.decode;
/**
 * encode the object into a binary stream
 * @method encode
 *
 * @param stream {BinaryStream} 
 */
ReadAtTimeDetails.prototype.encode = function(stream,options) {
    // call base class implementation first
    HistoryReadDetails.prototype.encode.call(this,stream,options);
    encodeArray(this.reqTimes, stream, encode_DateTime);
    encode_Boolean(this.useSimpleBounds,stream);
};
/**
 * decode the object from a binary stream
 * @method decode
 *
 * @param stream {BinaryStream} 
 * @param [option] {object} 
 */
ReadAtTimeDetails.prototype.decode = function(stream,options) {
    // call base class implementation first
    HistoryReadDetails.prototype.decode.call(this,stream,options);
    this.reqTimes = decodeArray(stream, decode_DateTime);
    this.useSimpleBounds = decode_Boolean(stream,options);
};
ReadAtTimeDetails.possibleFields = [
  "reqTimes",
         "useSimpleBounds"
];


exports.ReadAtTimeDetails = ReadAtTimeDetails;
var register_class_definition = require("node-opcua-factory/src/factories_factories").register_class_definition;
register_class_definition("ReadAtTimeDetails",ReadAtTimeDetails);
